{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Workflow used for: \n",
    "  1. Apply selected PGS scores into imputation data\n",
    "  1. Standardize raw scores and calculate Odds Ratios between case/controls\n",
    "  1. Generate visualization plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import all necessary libraries\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tkr\n",
    "import seaborn as sns\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 1) **Apply** selected PGS scores into imputation data using **pgs-calc** (https://github.com/lukfor/pgs-calc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "basedir = \"/labs/tassimes/rodrigoguarischi/projects/sea/apply_grs\"\n",
    "\n",
    "# Change working directory\n",
    "os.chdir(basedir)\n",
    "\n",
    "imputed_genotypes = \"/labs/tassimes/rodrigoguarischi/projects/sea/imputed_data/michigan_hrc/*.vcf.gz\"\n",
    "# imputed_genotypes = \"/labs/tassimes/rodrigoguarischi/projects/sea/imputed_data/topmed/liftover_hg19/*no_chr_prefix.vcf.gz\"\n",
    "\n",
    "ref_weights=[\"wGRS49\", \"PGS000349\", \"PGS000018\", \"PGS000667\", \"PGS000889\"]\n",
    "ref_weights_folder=\"./pgs_reference_weights/\"\n",
    "\n",
    "# Create full path to files\n",
    "ref_weights_paths = \",\".join([ref_weights_folder + pgs_name + \".txt.gz\" for pgs_name in ref_weights])\n",
    "\n",
    "for min_r2 in [0, 0.3, 0.5, 0.8]:\n",
    "    \n",
    "    print(\"Calculating scores for {0} at min R2 >= {1}\".format( \", \".join(ref_weights), min_r2 ) )\n",
    "    \n",
    "    output_scores_filename = \"_\".join(ref_weights) + \"_r\" + str(min_r2).replace(\".\",\"\") + \".scores.txt\"\n",
    "    html_report_filename = \"_\".join(ref_weights) + \"_r\" + str(min_r2).replace(\".\",\"\") + \".html\"\n",
    "    \n",
    "    !./pgs-calc/pgs-calc apply \\\n",
    "        --ref {ref_weights_paths} \\\n",
    "        --minR2 {min_r2} \\\n",
    "        --threads 22 \\\n",
    "        --no-ansi \\\n",
    "        # Run pgs-calc using GENOTYPE information, instead of DOSAGE (the default)\n",
    "        # --genotypes=GT \\ \n",
    "        --report-html={html_report_filename} \\\n",
    "        --out {output_scores_filename} \\\n",
    "        {imputed_genotypes}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 2) **Standardize** raw scores and calculate **Odds Ratios** between case/controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "race   sex   \n",
      "Black  Female     92\n",
      "       Male      412\n",
      "White  Female    128\n",
      "       Male      436\n",
      "Name: sex, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGaP_SubjID</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>cf</th>\n",
       "      <th>cr</th>\n",
       "      <th>rltotal</th>\n",
       "      <th>rlmean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seaid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>445969</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>445970</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>445971</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>445972</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15.8</td>\n",
       "      <td>5.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>445973</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         dbGaP_SubjID     sex   race  age  agex2   bmi    tf   tr    af    ar  \\\n",
       "seaid                                                                           \n",
       "pd10016        445969  Female  White   29    841  19.9   6.7  0.0  10.7   0.0   \n",
       "pd10018        445970    Male  White   30    900  28.3  27.3  0.0  22.0   0.0   \n",
       "pd10023        445971    Male  White   27    729  21.7  27.3  0.0  55.0   0.0   \n",
       "pd10028        445972    Male  Black   31    961    29  16.0  0.0  36.8  14.8   \n",
       "pd10031        445973    Male  White   28    784  21.5  20.0  0.0  16.7   0.0   \n",
       "\n",
       "          cf   cr  rltotal    rlmean  \n",
       "seaid                                 \n",
       "pd10016  3.3  0.0      0.0  0.000000  \n",
       "pd10018  3.4  0.6      0.6  0.200000  \n",
       "pd10023  4.0  0.0      0.0  0.000000  \n",
       "pd10028  1.7  1.0     15.8  5.266667  \n",
       "pd10031  0.0  0.0      0.0  0.000000  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read phenotypes and recode sex and race attributes\n",
    "sea_phenotypes = pd.read_table( \"./raw_files/SEA/SEA_Phase2_Subject_Phenotypes.txt\", index_col=\"seaid\" )\n",
    "\n",
    "# Recode Sex and Race\n",
    "sea_phenotypes = sea_phenotypes.replace( {\n",
    "    \"sex\": { 1:\"Male\", 2:\"Female\" },\n",
    "    \"race\": { 1:\"White\", 2:\"Black\" }\n",
    "    } )\n",
    "\n",
    "# Print counts by race and sex and first lines from dataframe \n",
    "print( sea_phenotypes.groupby([\"race\",\"sex\"])[\"sex\"].count() )\n",
    "sea_phenotypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGaP_SubjID</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>...</th>\n",
       "      <th>wGRS49_r05</th>\n",
       "      <th>PGS000349_r05</th>\n",
       "      <th>PGS000018_r05</th>\n",
       "      <th>PGS000667_r05</th>\n",
       "      <th>PGS000889_r05</th>\n",
       "      <th>wGRS49_r08</th>\n",
       "      <th>PGS000349_r08</th>\n",
       "      <th>PGS000018_r08</th>\n",
       "      <th>PGS000667_r08</th>\n",
       "      <th>PGS000889_r08</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>445969</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014892</td>\n",
       "      <td>0.876993</td>\n",
       "      <td>0.710600</td>\n",
       "      <td>2.385210</td>\n",
       "      <td>1.331384</td>\n",
       "      <td>0.500438</td>\n",
       "      <td>1.211101</td>\n",
       "      <td>0.464124</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.397157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>445970</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.119400</td>\n",
       "      <td>1.052222</td>\n",
       "      <td>1.198771</td>\n",
       "      <td>1.649050</td>\n",
       "      <td>0.948200</td>\n",
       "      <td>0.174197</td>\n",
       "      <td>0.624084</td>\n",
       "      <td>0.458905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.373110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>445971</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.137079</td>\n",
       "      <td>0.729801</td>\n",
       "      <td>0.746956</td>\n",
       "      <td>1.443510</td>\n",
       "      <td>1.508553</td>\n",
       "      <td>0.213983</td>\n",
       "      <td>0.418310</td>\n",
       "      <td>0.430099</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.814898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>445972</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.301765</td>\n",
       "      <td>0.705801</td>\n",
       "      <td>1.184374</td>\n",
       "      <td>16.642935</td>\n",
       "      <td>0.264086</td>\n",
       "      <td>-0.009379</td>\n",
       "      <td>0.304889</td>\n",
       "      <td>0.743781</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.341113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>445973</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.193940</td>\n",
       "      <td>0.980570</td>\n",
       "      <td>0.744978</td>\n",
       "      <td>3.189740</td>\n",
       "      <td>-0.079091</td>\n",
       "      <td>0.330856</td>\n",
       "      <td>0.491178</td>\n",
       "      <td>0.349139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.201303</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         dbGaP_SubjID     sex   race  age  agex2   bmi    tf   tr    af    ar  \\\n",
       "pd10016        445969  Female  White   29    841  19.9   6.7  0.0  10.7   0.0   \n",
       "pd10018        445970    Male  White   30    900  28.3  27.3  0.0  22.0   0.0   \n",
       "pd10023        445971    Male  White   27    729  21.7  27.3  0.0  55.0   0.0   \n",
       "pd10028        445972    Male  Black   31    961    29  16.0  0.0  36.8  14.8   \n",
       "pd10031        445973    Male  White   28    784  21.5  20.0  0.0  16.7   0.0   \n",
       "\n",
       "         ...  wGRS49_r05  PGS000349_r05  PGS000018_r05  PGS000667_r05  \\\n",
       "pd10016  ...   -0.014892       0.876993       0.710600       2.385210   \n",
       "pd10018  ...   -0.119400       1.052222       1.198771       1.649050   \n",
       "pd10023  ...    0.137079       0.729801       0.746956       1.443510   \n",
       "pd10028  ...   -0.301765       0.705801       1.184374      16.642935   \n",
       "pd10031  ...   -0.193940       0.980570       0.744978       3.189740   \n",
       "\n",
       "         PGS000889_r05  wGRS49_r08  PGS000349_r08  PGS000018_r08  \\\n",
       "pd10016       1.331384    0.500438       1.211101       0.464124   \n",
       "pd10018       0.948200    0.174197       0.624084       0.458905   \n",
       "pd10023       1.508553    0.213983       0.418310       0.430099   \n",
       "pd10028       0.264086   -0.009379       0.304889       0.743781   \n",
       "pd10031      -0.079091    0.330856       0.491178       0.349139   \n",
       "\n",
       "         PGS000667_r08  PGS000889_r08  \n",
       "pd10016            0.0       0.397157  \n",
       "pd10018            0.0       0.373110  \n",
       "pd10023            0.0       0.814898  \n",
       "pd10028            0.0      -0.341113  \n",
       "pd10031            0.0       0.201303  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read files with raw scores of multiple PRSs\n",
    "\n",
    "import os, glob\n",
    "\n",
    "li = []\n",
    "li.append(sea_phenotypes)\n",
    "\n",
    "# for current_score_file in glob.glob( \"./apply_grs/results_genotypes/*.scores.txt\" ):\n",
    "# for current_score_file in glob.glob( \"./apply_grs/results_dosage/*.scores.txt\" ):\n",
    "# for current_score_file in glob.glob( \"./apply_grs/michigan_hrc_20220419/*.scores.txt\" ):\n",
    "for current_score_file in glob.glob( \"./apply_grs/topmed_20220419/*.scores.txt\" ):\n",
    "    pgs_results = pd.read_table( current_score_file, sep = \",\" )\n",
    "    pgs_results[\"sample\"] = pgs_results[\"sample\"].str.split(\"_\", expand = True)[0]\n",
    "    pgs_results = pgs_results.set_index(\"sample\")\n",
    "    min_r2_used = os.path.basename( current_score_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "    pgs_results = pgs_results.add_suffix(\"_\" + min_r2_used)\n",
    "\n",
    "    li.append( pgs_results )\n",
    "    \n",
    "sea_merged = pd.concat( li, axis=1)\n",
    "sea_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign case/control classes to samples using top quartile rule splitting by sex\n",
    "\n",
    "# Phenotype of interest\n",
    "phenotype = \"cr\"\n",
    "\n",
    "# Subset cohort\n",
    "sea_merged_whites = sea_merged[ sea_merged[\"race\"] == \"White\" ]\n",
    "\n",
    "# Calculate thresholds for each sex\n",
    "male_threshold = sea_merged_whites[ sea_merged_whites[\"sex\"]==\"Male\" ][phenotype].quantile(0.75)\n",
    "female_threshold = sea_merged_whites[sea_merged_whites[\"sex\"]==\"Female\"][phenotype].quantile(0.75)\n",
    "print( \"Thresholds to be used for case/control definition: male = {0} and female = {1}\".format( male_threshold, female_threshold) )\n",
    "\n",
    "if( male_threshold == 0 or female_threshold == 0 ):\n",
    "    sys.exit(\"Male ({0}) and/or Female ({1}) thresholds equals to zero!\".format(male_threshold, female_threshold))\n",
    "\n",
    "# Add a new column called Case with all values equals to False\n",
    "sea_merged_whites = sea_merged_whites.assign( Case=False )\n",
    "\n",
    "# Identify subjects above threshold to assign them to group \"Case\"\n",
    "for i in sea_merged_whites.index:\n",
    "    if( sea_merged_whites.loc[i, \"sex\"] == \"Male\" and sea_merged_whites.loc[i, phenotype] > male_threshold ):\n",
    "        sea_merged_whites.loc[i, \"Case\"] = True\n",
    "    elif( sea_merged_whites.loc[i, \"sex\"] == \"Female\" and sea_merged_whites.loc[i, phenotype] > female_threshold ):\n",
    "        sea_merged_whites.loc[i, \"Case\"] = True\n",
    "\n",
    "sea_merged_whites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sea_merged_whites.head()\n",
    "import numpy as np\n",
    "from sklearn import preprocessing\n",
    "import statsmodels.api as sm\n",
    "\n",
    "standardize_scores = []\n",
    "logit_models_dict = {}\n",
    "models_summaries = pd.DataFrame()\n",
    "\n",
    "# Recode dependent variable to 0 and 1 to fit glm\n",
    "sea_merged_whites[\"Case_recoded\"] = sea_merged_whites[\"Case\"]\n",
    "sea_merged_whites[\"Case_recoded\"].replace(True, 1, inplace=True)\n",
    "sea_merged_whites[\"Case_recoded\"].replace(False, 0, inplace=True)\n",
    "\n",
    "# Normalize all scores with mean = 0 and SD = 1\n",
    "for grs in [\"wGRS49\",\"PGS000349\",\"PGS000018\", \"PGS000667\", \"PGS000889\"]:\n",
    "    for threshold in [\"r0\", \"r03\", \"r05\", \"r08\"]:\n",
    "        raw_score_name = grs + \"_\" + threshold\n",
    "        standardize_score_name = grs + \"_\" + threshold + \"_standardize\"\n",
    "        standardize_scores.append(standardize_score_name)\n",
    "        \n",
    "        # Standardize raw scores using method scale\n",
    "        sea_merged_whites[ standardize_score_name ] = preprocessing.scale( sea_merged_whites[ raw_score_name ] )\n",
    "        \n",
    "        # Fit a logistic model using standardize scores and save it on dictionary \n",
    "        logit_models_dict[ standardize_score_name ] = sm.formula.glm(\n",
    "            \"Case_recoded ~ \" + standardize_score_name + \" + age + sex\",\n",
    "            family=sm.families.Binomial(),\n",
    "            data=sea_merged_whites).fit()\n",
    "\n",
    "        model_summary_series = pd.Series(\n",
    "            data={\n",
    "                \"beta\": np.exp( logit_models_dict[standardize_score_name].params )[ standardize_score_name ],\n",
    "                \"score_pvalue\": logit_models_dict[standardize_score_name].pvalues[standardize_score_name],\n",
    "                \"conf_interval_lower\": np.exp( logit_models_dict[standardize_score_name].conf_int()[0][standardize_score_name] ),\n",
    "                \"conf_interval_upper\": np.exp( logit_models_dict[standardize_score_name].conf_int()[1][standardize_score_name] )\n",
    "            },\n",
    "            name=standardize_score_name\n",
    "        )\n",
    "        \n",
    "        models_summaries = pd.concat( [models_summaries, model_summary_series], axis=1)\n",
    "\n",
    "# Print summary statistics to confirm that all standadize scores have mean = 0 and SD = 1\n",
    "# print( \"Normalized scores:\\n{0}\\n\".format(sea_merged_whites[ standardize_scores ].describe().transpose()) )\n",
    "# print( \"Normalized scores:\" )\n",
    "# print( sea_merged_whites[ standardize_scores ].describe().transpose() )\n",
    "      \n",
    "# Transpose models_summaries object and print results for all models tested\n",
    "models_summaries = models_summaries.transpose()\n",
    "# print(models_summaries, end=\"\")\n",
    "# print( \"Models summaries:\\n{0}\\n\".format(models_summaries.transpose()) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_name = \"wGRS49_r03_standardize\"\n",
    "# print( logit_models_dict[model_name].summary() )\n",
    "print( logit_models_dict[model_name].pvalues[model_name] )\n",
    "# print( np.exp( logit_models_dict[model_name].params )[ model_name ] )\n",
    "# print( np.exp( logit_models_dict[model_name].conf_int().loc[[ model_name ]] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot boxplots and histograms and boxplots of raw scores spliting by sex\n",
    "\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "score_list = [\"wGRS49_r0\", \"wGRS49_r03\", \"wGRS49_r05\", \"wGRS49_r08\"]\n",
    "# score_list = [\"PGS000349_r0\", \"PGS000349_r03\", \"PGS000349_r05\", \"PGS000349_r08\"]\n",
    "# score_list = [\"PGS000018_r0\", \"PGS000018_r03\", \"PGS000018_r05\", \"PGS000018_r08\"]\n",
    "\n",
    "fig, axs = plt.subplots(nrows=3, ncols=4, figsize=(5*len(score_list),12))\n",
    "for i in range(0, len(score_list)):\n",
    "    sns.boxplot(\n",
    "        x = \"Case\",\n",
    "        y = score_list[i],\n",
    "        data = sea_merged_whites,\n",
    "        palette = reversed(sns.color_palette(n_colors=2)),\n",
    "        width=0.4,\n",
    "        fliersize=2,\n",
    "        ax=axs[0,i],\n",
    "    )\n",
    "    sns.histplot(\n",
    "        x = score_list[i],\n",
    "        data = sea_merged_whites[ sea_merged_whites[\"sex\"] == \"Male\" ],\n",
    "        hue=\"Case\",\n",
    "        hue_order=[True, False],\n",
    "        kde=True,\n",
    "        ax=axs[1,i]\n",
    "        ).set(title='Males (n={0})'.format( sum( sea_merged_whites[\"sex\"] == \"Male\" ) ) )\n",
    "    sns.histplot(\n",
    "        x = score_list[i],\n",
    "        data = sea_merged_whites[ sea_merged_whites[\"sex\"] == \"Female\" ],\n",
    "        hue=\"Case\",\n",
    "        hue_order=[True, False],\n",
    "        kde=True,\n",
    "        ax=axs[2,i]\n",
    "    ).set(title='Females (n={0})'.format( sum( sea_merged_whites[\"sex\"] == \"Female\" ) ) )\n",
    "\n",
    "plt.subplots_adjust(top=1.25)\n",
    "\n",
    "# # Plot only boxplots\n",
    "\n",
    "# fig, axs = plt.subplots(ncols=len(score_list), figsize=(5*len(score_list),4))\n",
    "# for i in range(0, len(score_list)):\n",
    "#     sns.boxplot(\n",
    "#         x = \"Case\",\n",
    "#         y = score_list[i],\n",
    "#         data = sea_merged_whites,\n",
    "#         hue=\"sex\",\n",
    "#         hue_order=['Male','Female'],\n",
    "#         width=0.4,\n",
    "#         fliersize=2,\n",
    "#         ax=axs[i],\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test variables\n",
    "sns.boxplot(\n",
    "    x = \"Case\",\n",
    "    y = \"age\",\n",
    "    data = sea_merged_whites,\n",
    "    width=0.4,\n",
    "    fliersize=2\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export dataset to CSV files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export full dataset\n",
    "sea_merged_whites.to_csv(\"exports/sea_whites_phenotypes_scores_full.csv\")\n",
    "\n",
    "# Export only subset of columns Tim asked\n",
    "select_columns = [\"dbGaP_SubjID\",\"sex\",\"race\",\"age\",\"agex2\",\"bmi\",\"cr\",\"wGRS49_r05\",\"PGS000349_r05\",\"PGS000018_r05\",\"wGRS49_r03\",\"PGS000349_r03\",\"PGS000018_r03\",\"wGRS49_r0\",\"PGS000349_r0\",\"PGS000018_r0\",\"wGRS49_r08\",\"PGS000349_r08\",\"PGS000018_r08\", \"Case\"]\n",
    "sea_merged_whites[select_columns].to_csv(\"exports/sea_whites_phenotypes_scores_column_subset.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "sns.scatterplot(\n",
    "    x = \"wGRS49_r03_scaled\",\n",
    "    y = \"Case\",\n",
    "    data = sea_merged_whites\n",
    "    # ,\n",
    "    # alpha=0.4,\n",
    "    # s=6\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test logistic regression in <code>R</code>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Read CSV file with raw scores\n",
    "input_table <- read.csv(\"exports/sea_whites_phenotypes_scores_full.csv\", row.names=1)\n",
    "\n",
    "# Normalize all scores with mean = 0 and SD = 1\n",
    "for ( grs in c(\"wGRS49\",\"PGS000349\",\"PGS000018\", \"PGS000667\", \"PGS000889\") ){\n",
    "    for ( threshold in c(\"r0\", \"r03\", \"r05\", \"r08\") ){\n",
    "        current_score <- paste0( grs, \"_\", threshold);\n",
    "        score_values <- input_table[[ current_score ]]\n",
    "        score_mean <- mean( score_values )\n",
    "        score_sd <- sd( score_values )\n",
    "        input_table[ paste0(current_score, \"_scaled\") ] <- (score_values-score_mean)/score_sd\n",
    "        # input_table[ paste0(current_score, \"_scaled\") ] <- scale( score_values )\n",
    "    }\n",
    "}\n",
    "\n",
    "# Test Standardization (check values of min, mean, max and sd)\n",
    "normalization_summary <- function(current_scaled_score){\n",
    "  return (\n",
    "    list(\n",
    "      \"min\"=min(current_scaled_score),\n",
    "      \"mean\"=mean(current_scaled_score),\n",
    "      \"max\"=max(current_scaled_score),\n",
    "      \"sd\"=sd(current_scaled_score)\n",
    "    )\n",
    "  )\n",
    "}\n",
    "t( sapply( input_table[ , grepl(\"_scaled\", names(input_table)) ], normalization_summary ) )\n",
    "\n",
    "# Test number of case/controls for each sex and age\n",
    "xtabs( ~ Case + sex, data=input_table)\n",
    "xtabs( ~ Case + age, data=input_table)\n",
    "\n",
    "# Apply logistic regression to scaled scores\n",
    "models <- list()\n",
    "for ( score_scaled_name in grep(\"_scaled\", names(input_table), value = T) ){\n",
    "  \n",
    "  # Skip PGSs with 0 SNPs evaluated (scores equals to NaN)  \n",
    "  if( is.na( median( input_table[, score_scaled_name ] ) ) ) next;\n",
    "  \n",
    "  input_table_subset <- input_table[, c(score_scaled_name, \"age\", \"sex\", \"Case\") ]\n",
    "  colnames( input_table_subset ) <- c(\"score_scaled\", \"age\", \"sex\", \"Case\")\n",
    "  models[[ score_scaled_name ]] <- glm( Case ~ score_scaled + age + sex, data = input_table_subset, family = \"binomial\")\n",
    "}\n",
    "\n",
    "# Print Odds Ratios for each score at 4 levels of imputation Rsqrt thresholding (0, 0.3, 0.5 and 0.8)\n",
    "t( sapply( models, function(current_model) exp( current_model$coefficients ) ) )\n",
    "\n",
    "# Get p-values\n",
    "t( sapply( models, function(current_model) coef(summary(current_model))[,4] ) )\n",
    "\n",
    "summary( models$wGRS49_r0_scaled )\n",
    "\n",
    "---\n",
    "apply_logit <- function(current_scaled_score, input_table) {\n",
    "    logit_formula <- as.formula( paste0(\"Case ~ \" , current_scaled_score ,\" + age + sex\") )\n",
    "    logistic_model <- glm( logit_formula, data = input_table, family = \"binomial\")\n",
    "    return( logistic_model )\n",
    "} \n",
    "scaled_scores <- grep(\"_scaled\", names(input_table), value = T)\n",
    "models <- lapply( scaled_scores, function(current_scaled_score) apply_logit(current_scaled_score, input_table) )\n",
    "names(models) <- scaled_scores\n",
    "\n",
    "----\n",
    "\n",
    "summary( models$wGRS49_r03_scaled )\n",
    "str( models$wGRS49_r03_scaled )\n",
    "\n",
    "table( ifelse( models$wGRS49_r03_scaled$fitted.values > 0.5, \"assigned_case\", \"assigned_control\") )\n",
    "\n",
    "summary( models[[2]] )\n",
    "\n",
    "create_pdf_plot <- function(model, input_data) {\n",
    "\n",
    "    library(ggplot2)\n",
    "    library(cowplot)\n",
    "\n",
    "    predicted_data <- data.frame(\n",
    "        probability_of_hd = model$fitted.values,\n",
    "        cr = input_table$Case\n",
    "        )\n",
    "\n",
    "    predicted_data <- predicted_data[ order(predicted_data$probability_of_hd, decreasing=FALSE), ]\n",
    "    predicted_data$rank <- 1:nrow(predicted_data)\n",
    "\n",
    "    ggplot( data=predicted_data, aes(x=rank, y=probability_of_hd) ) +\n",
    "        geom_point(aes(color=cr), alpha=1, shape=4, stroke=2) +\n",
    "        xlab(\"Index\") +\n",
    "        ylab(\"Predicted probability of getting heart disease\")\n",
    "    ggsave(\"heart_disease_probabilities.pdf\")\n",
    "}\n",
    "\n",
    "create_pdf_plot(models[[1]], input_table)\n",
    "\n",
    "-----\n",
    "\n",
    "# Test model based on multiple scores\n",
    "logistic_model <- glm( Case ~ ., data = input_table[,c(scaled_scores, \"age\", \"sex\", \"Case\")], family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "-----\n",
    "\n",
    "summary( models[, \"wGRS49_r0_scaled\"] )\n",
    "\n",
    "t( sapply( scaled_scores, function(x) apply_logit(x, input_table) ) )\n",
    "\n",
    "t( sapply( grep(\"_scaled\", names(input_table), value = T), function(x) apply_logit(x, input_table) ) )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "sapply( grep(\"_scaled\", names(input_table), value = T), function(x) logit_formula <- print0(\"Case ~ \" , x ,\" + age + sex\"; print(logit_formula) )\n",
    "\n",
    "sapply( 1:5, function(x) x*2)\n",
    "\n",
    "logistic_model <- glm(Case ~ wGRS49_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "logistic_model <- glm(Case ~ PGS000349_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "logistic_model <- glm(Case ~ PGS000018_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "input_table_subset <- input_table[ , c( grep(\"_scaled\", names(input_table), value = T), \"age\", \"sex\", \"Case\") ]\n",
    "\n",
    "logistic_model <- glm(Case~., data = input_table_subset, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "// input_table[ paste0(current_score, \"_scaled\") ] <- scale( input_table[[ current_score ]], center = 0, scale = 1)\n",
    "\n",
    "range( input_table$PGS000018_r03_scaled )\n",
    "input_table$wGRS49_r03_scaled <- scale(input_table$wGRS49_r03, center = 0, scale = 1)\n",
    "input_table$PGS000018_r03_scaled <- scale(input_table$PGS000018_r03, center = 0, scale = 1)\n",
    "\n",
    "logistic_model <- glm(Case ~ wGRS49_r03_scaled, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "\n",
    "# Predict test data based on model\n",
    "predict_reg <- predict(logistic_model, input_table, type = \"response\")\n",
    "range(predict_reg)\n",
    "predict_reg <- ifelse( predict_reg > 0.5, 1, 0)\n",
    "\n",
    "range(predict_reg)\n",
    "logistic_model$\n",
    "\n",
    "predict_reg <- <- predict(logistic_model, logistic_model, type = \"response\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# table = df.groupby(level=\"Cancer\").sum().values\n",
    "\n",
    "table = [[840,51663],[32,5053]]\n",
    "print(table)\n",
    "oddsratio, pvalue = stats.fisher_exact(table)\n",
    "print(\"OddsR: \", oddsratio, \"p-Value:\", pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sea_merged_whites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "for score in [\"wGRS49_r03\", \"wGRS49_r05\", \"wGRS49_r08\", \"PGS000018_r03\", \"PGS000018_r05\", \"PGS000018_r08\"]:\n",
    "    for traits in [\"tf_quartile4\", \"tr_quartile4\", \"af_quartile4\", \"ar_quartile4\", \"cf_quartile4\", \"cr_quartile4\"]:\n",
    "        # score = \"PGS000018_r08\"\n",
    "        cases = sea_merged_whites[traits]\n",
    "        print( score , \"\\t\", traits, \"\\t\", stats.ttest_ind( sea_merged_whites[cases][score], sea_merged_whites[np.invert(cases)][score] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "\n",
    "current_score_file = \"./apply_grs/wGRS49_PGS000349_PGS000018_r0.scores.txt\"\n",
    "\n",
    "pgs_results = pd.read_table( current_score_file, sep = \",\" )\n",
    "pgs_results[\"sample\"] = pgs_results[\"sample\"].str.split(\"_\", expand = True)[0]\n",
    "pgs_results = pgs_results.set_index(\"sample\")\n",
    "min_r2_used = os.path.basename( current_score_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "pgs_results = pgs_results.add_suffix(\"_\" + min_r2_used)\n",
    "pgs_results.head()\n",
    "    # score_name = os.path.basename( current_score_file ).split(\".\")[0]\n",
    "    # pgs_results.rename( columns={ pgs_results.columns[0]:score_name }, inplace=True )\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "kernelspec": {
   "display_name": "Python 3.6.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
