{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Workflow used for:\n",
    "  1. Load metadata about GRSs used\n",
    "  1. Read phenotypic information and consolidate it with raw PGS scores from **apply_grs.ipynb**\n",
    "  1. Standardize raw scores and calculate Odds Ratios between case/controls\n",
    "  1. Generate visualization plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import all necessary libraries\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tkr\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os, glob\n",
    "from sklearn import preprocessing\n",
    "import statsmodels.api as sm\n",
    "import json\n",
    "from xlsxwriter.utility import xl_rowcol_to_cell"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 1)** Load metadata about GRSs used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/apply_grs/pgs_reference_weights/\")\n",
    "\n",
    "# Load all key/values present on the header of PGS files into a dictionary\n",
    "pgs_metadata = {}\n",
    "for current_score_file in glob.glob( \"*.txt.gz\" ):\n",
    "    score_name = current_score_file.replace(\".txt.gz\",\"\")\n",
    "    pgs_metadata[score_name] = {}\n",
    "    with gzip.open( current_score_file,'rt') as f:\n",
    "        for line in f:\n",
    "            if( ( line.startswith(\"#\") ) & (not line.startswith(\"##\") ) ):\n",
    "                line = line.strip()\n",
    "                line = line.replace(\"#\", \"\")\n",
    "                key, value = line.split(\"=\")\n",
    "                pgs_metadata[score_name][key] = value\n",
    "    if( str( pgs_metadata[score_name][\"pgs_id\"] ).startswith(\"PGS\") ):\n",
    "        pgs_metadata[score_name][\"pgs_catalog_hyperlink\"] = \"https://www.pgscatalog.org/score/{0}/\".format(pgs_metadata[score_name][\"pgs_id\"])\n",
    "    else:\n",
    "        pgs_metadata[score_name][\"pgs_catalog_hyperlink\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 2)** Read phenotypic information and consolidate with raw PGS scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "race   sex   \n",
      "black  female     92\n",
      "       male      412\n",
      "white  female    128\n",
      "       male      436\n",
      "Name: sex, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGaP SubjID</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>cf</th>\n",
       "      <th>cr</th>\n",
       "      <th>rltotal</th>\n",
       "      <th>rlmean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seaid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>445969</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>445970</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>445971</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>445972</td>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15.8</td>\n",
       "      <td>5.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>445973</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         dbGaP SubjID     sex   race  age  agex2   bmi    tf   tr    af    ar  \\\n",
       "seaid                                                                           \n",
       "pd10016        445969  female  white   29    841  19.9   6.7  0.0  10.7   0.0   \n",
       "pd10018        445970    male  white   30    900  28.3  27.3  0.0  22.0   0.0   \n",
       "pd10023        445971    male  white   27    729  21.7  27.3  0.0  55.0   0.0   \n",
       "pd10028        445972    male  black   31    961    29  16.0  0.0  36.8  14.8   \n",
       "pd10031        445973    male  white   28    784  21.5  20.0  0.0  16.7   0.0   \n",
       "\n",
       "          cf   cr  rltotal    rlmean  \n",
       "seaid                                 \n",
       "pd10016  3.3  0.0      0.0  0.000000  \n",
       "pd10018  3.4  0.6      0.6  0.200000  \n",
       "pd10023  4.0  0.0      0.0  0.000000  \n",
       "pd10028  1.7  1.0     15.8  5.266667  \n",
       "pd10031  0.0  0.0      0.0  0.000000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/\")\n",
    "\n",
    "# Read phenotypes and recode sex and race attributes\n",
    "sea_phenotypes = pd.read_table(\n",
    "    \"data_preparation_to_imputation/86679/NHLBI/SEA_Herrington/phs000349v1/p1/phenotype/phs000349.v1.pht002191.v1.p1.c1.SEA_Phase2_Subject_Phenotypes.GRU.txt\",\n",
    "    index_col=\"seaid\",\n",
    "    comment=\"#\")\n",
    "\n",
    "# Recode Sex and Race\n",
    "sea_phenotypes = sea_phenotypes.replace( {\n",
    "    \"sex\": { 1:\"male\", 2:\"female\" },\n",
    "    \"race\": { 1:\"white\", 2:\"black\" }\n",
    "    } )\n",
    "\n",
    "# Print counts by race and sex and first lines from dataframe \n",
    "print( sea_phenotypes.groupby([\"race\",\"sex\"])[\"sex\"].count() )\n",
    "sea_phenotypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGaP SubjID</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>...</th>\n",
       "      <th>topmed_r03_PGS000013</th>\n",
       "      <th>topmed_r03_PGS001357</th>\n",
       "      <th>topmed_r03_PGS002114</th>\n",
       "      <th>topmed_r03_PGS001105</th>\n",
       "      <th>topmed_r03_PGS000349</th>\n",
       "      <th>topmed_r03_PGS001917</th>\n",
       "      <th>topmed_r03_PGS001818</th>\n",
       "      <th>topmed_r03_PGS000957</th>\n",
       "      <th>topmed_r03_PGS000889</th>\n",
       "      <th>topmed_r03_PGS000018</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>445969</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.195099</td>\n",
       "      <td>-0.032170</td>\n",
       "      <td>2.662745</td>\n",
       "      <td>4.381828</td>\n",
       "      <td>0.899946</td>\n",
       "      <td>0.011109</td>\n",
       "      <td>-0.147684</td>\n",
       "      <td>0.027173</td>\n",
       "      <td>0.945526</td>\n",
       "      <td>0.634510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>445970</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.181852</td>\n",
       "      <td>-0.232240</td>\n",
       "      <td>1.028924</td>\n",
       "      <td>1.048539</td>\n",
       "      <td>1.140902</td>\n",
       "      <td>-0.013622</td>\n",
       "      <td>0.644547</td>\n",
       "      <td>0.874353</td>\n",
       "      <td>0.870190</td>\n",
       "      <td>1.261631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>445971</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.151067</td>\n",
       "      <td>-0.191068</td>\n",
       "      <td>3.098229</td>\n",
       "      <td>1.721006</td>\n",
       "      <td>0.694613</td>\n",
       "      <td>0.001444</td>\n",
       "      <td>0.470915</td>\n",
       "      <td>0.735025</td>\n",
       "      <td>1.365498</td>\n",
       "      <td>0.623044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>445972</td>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016966</td>\n",
       "      <td>-0.251807</td>\n",
       "      <td>3.165720</td>\n",
       "      <td>4.340293</td>\n",
       "      <td>0.627491</td>\n",
       "      <td>0.029261</td>\n",
       "      <td>0.090178</td>\n",
       "      <td>1.278612</td>\n",
       "      <td>-0.102817</td>\n",
       "      <td>1.027068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>445973</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.047663</td>\n",
       "      <td>-0.204522</td>\n",
       "      <td>3.074076</td>\n",
       "      <td>2.298869</td>\n",
       "      <td>1.016425</td>\n",
       "      <td>0.006443</td>\n",
       "      <td>-0.178284</td>\n",
       "      <td>0.897139</td>\n",
       "      <td>-0.176216</td>\n",
       "      <td>0.740625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 214 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         dbGaP SubjID     sex   race  age  agex2   bmi    tf   tr    af    ar  \\\n",
       "pd10016        445969  female  white   29    841  19.9   6.7  0.0  10.7   0.0   \n",
       "pd10018        445970    male  white   30    900  28.3  27.3  0.0  22.0   0.0   \n",
       "pd10023        445971    male  white   27    729  21.7  27.3  0.0  55.0   0.0   \n",
       "pd10028        445972    male  black   31    961    29  16.0  0.0  36.8  14.8   \n",
       "pd10031        445973    male  white   28    784  21.5  20.0  0.0  16.7   0.0   \n",
       "\n",
       "         ...  topmed_r03_PGS000013  topmed_r03_PGS001357  \\\n",
       "pd10016  ...              0.195099             -0.032170   \n",
       "pd10018  ...              0.181852             -0.232240   \n",
       "pd10023  ...              0.151067             -0.191068   \n",
       "pd10028  ...             -0.016966             -0.251807   \n",
       "pd10031  ...              0.047663             -0.204522   \n",
       "\n",
       "         topmed_r03_PGS002114  topmed_r03_PGS001105  topmed_r03_PGS000349  \\\n",
       "pd10016              2.662745              4.381828              0.899946   \n",
       "pd10018              1.028924              1.048539              1.140902   \n",
       "pd10023              3.098229              1.721006              0.694613   \n",
       "pd10028              3.165720              4.340293              0.627491   \n",
       "pd10031              3.074076              2.298869              1.016425   \n",
       "\n",
       "         topmed_r03_PGS001917  topmed_r03_PGS001818  topmed_r03_PGS000957  \\\n",
       "pd10016              0.011109             -0.147684              0.027173   \n",
       "pd10018             -0.013622              0.644547              0.874353   \n",
       "pd10023              0.001444              0.470915              0.735025   \n",
       "pd10028              0.029261              0.090178              1.278612   \n",
       "pd10031              0.006443             -0.178284              0.897139   \n",
       "\n",
       "         topmed_r03_PGS000889  topmed_r03_PGS000018  \n",
       "pd10016              0.945526              0.634510  \n",
       "pd10018              0.870190              1.261631  \n",
       "pd10023              1.365498              0.623044  \n",
       "pd10028             -0.102817              1.027068  \n",
       "pd10031             -0.176216              0.740625  \n",
       "\n",
       "[5 rows x 214 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read files with raw scores of multiple GRSs\n",
    "\n",
    "li = []\n",
    "li.append(sea_phenotypes)\n",
    "\n",
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/apply_grs/raw_scores_20220504/\")\n",
    "\n",
    "# Loop over info files and save info as a dictionary   \n",
    "scores_info_dict = {}\n",
    "for current_info_file in glob.glob( \"*.info.txt\" ):\n",
    "\n",
    "    # Parse filename to get info about GRSs.\n",
    "    # Format should be: <hrc|topmed>_<whites|blacks>_<8-digits_date>_multiGRS_<minr2_used>.info.txt\n",
    "    min_r2_used = os.path.basename( current_info_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "    reference_panel_name = os.path.basename( current_info_file ).split(\"_\")[0]\n",
    "    race = os.path.basename( current_info_file ).split(\"_\")[1]\n",
    "    \n",
    "    f = open( current_info_file )\n",
    "    data = json.load(f)\n",
    "\n",
    "    # Load metrics on a dictionary of dictionaries using panel, r2 and PGS name and each race as keys\n",
    "    for score_result in data:\n",
    "        key = \"_\".join( (reference_panel_name, min_r2_used, score_result[\"name\"]) )\n",
    "        \n",
    "        # If key doesn't exist yet in the dictionary, create it\n",
    "        if key not in scores_info_dict:\n",
    "            scores_info_dict[key] = {}\n",
    "            \n",
    "        scores_info_dict[key][race] = score_result\n",
    "\n",
    "    f.close()\n",
    "\n",
    "# Loop over scores files, load all into memory and save it to list object 'li'    \n",
    "for current_score_file in glob.glob( \"*.scores.txt\" ):\n",
    "        \n",
    "    grs_results = pd.read_table( current_score_file, sep = \",\" )\n",
    "    grs_results[\"sample\"] = grs_results[\"sample\"].str.split(\"_\", expand = True)[0]\n",
    "    grs_results = grs_results.set_index(\"sample\")\n",
    "    min_r2_used = os.path.basename( current_score_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "    reference_panel_name = os.path.basename( current_score_file ).split(\"_\")[0]\n",
    "        \n",
    "    # Add prefix to column names. Names should match pattern <hrc|topmed>_<threshold>_<scoreid>\n",
    "    grs_results = grs_results.add_prefix(reference_panel_name + \"_\" + min_r2_used + \"_\")\n",
    "\n",
    "    # Test if this set of scores already exists in li. \n",
    "    # If the other race was already loaded, append subjects to the dataframe. Otherwise, append dataframe to li\n",
    "    new_score_set = True\n",
    "    for i in range( len(li) ):\n",
    "        if( set( grs_results.columns ) == set( li[i].columns ) ):\n",
    "            li[i] = pd.concat( [li[i], grs_results], axis=0 )\n",
    "            new_score_set = False\n",
    "\n",
    "    if( new_score_set ):\n",
    "        li.append( grs_results )\n",
    "\n",
    "# Consolidate data into a dataframe and print first lines\n",
    "sea_dataset_full = pd.concat( li, axis=1 )\n",
    "sea_dataset_full.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thresholds to be used for case/control definition:\n",
      " - white_male = 1.6000 \n",
      " - white_female = 0.4250 \n",
      " - black_male = 1.0000 \n",
      " - black_female = 0.0000 (WARNING: This group has Q3 equals zero. Considering only non-zeros as CASE group!!)\n",
      "race   sex     Case \n",
      "black  female  False     73\n",
      "               True      19\n",
      "       male    False    311\n",
      "               True     101\n",
      "white  female  False     96\n",
      "               True      32\n",
      "       male    False    328\n",
      "               True     108\n",
      "Name: Case, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGaP SubjID</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>...</th>\n",
       "      <th>topmed_r03_PGS001357</th>\n",
       "      <th>topmed_r03_PGS002114</th>\n",
       "      <th>topmed_r03_PGS001105</th>\n",
       "      <th>topmed_r03_PGS000349</th>\n",
       "      <th>topmed_r03_PGS001917</th>\n",
       "      <th>topmed_r03_PGS001818</th>\n",
       "      <th>topmed_r03_PGS000957</th>\n",
       "      <th>topmed_r03_PGS000889</th>\n",
       "      <th>topmed_r03_PGS000018</th>\n",
       "      <th>Case</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>445969</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.032170</td>\n",
       "      <td>2.662745</td>\n",
       "      <td>4.381828</td>\n",
       "      <td>0.899946</td>\n",
       "      <td>0.011109</td>\n",
       "      <td>-0.147684</td>\n",
       "      <td>0.027173</td>\n",
       "      <td>0.945526</td>\n",
       "      <td>0.634510</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>445970</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.232240</td>\n",
       "      <td>1.028924</td>\n",
       "      <td>1.048539</td>\n",
       "      <td>1.140902</td>\n",
       "      <td>-0.013622</td>\n",
       "      <td>0.644547</td>\n",
       "      <td>0.874353</td>\n",
       "      <td>0.870190</td>\n",
       "      <td>1.261631</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>445971</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.191068</td>\n",
       "      <td>3.098229</td>\n",
       "      <td>1.721006</td>\n",
       "      <td>0.694613</td>\n",
       "      <td>0.001444</td>\n",
       "      <td>0.470915</td>\n",
       "      <td>0.735025</td>\n",
       "      <td>1.365498</td>\n",
       "      <td>0.623044</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>445972</td>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.251807</td>\n",
       "      <td>3.165720</td>\n",
       "      <td>4.340293</td>\n",
       "      <td>0.627491</td>\n",
       "      <td>0.029261</td>\n",
       "      <td>0.090178</td>\n",
       "      <td>1.278612</td>\n",
       "      <td>-0.102817</td>\n",
       "      <td>1.027068</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>445973</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.204522</td>\n",
       "      <td>3.074076</td>\n",
       "      <td>2.298869</td>\n",
       "      <td>1.016425</td>\n",
       "      <td>0.006443</td>\n",
       "      <td>-0.178284</td>\n",
       "      <td>0.897139</td>\n",
       "      <td>-0.176216</td>\n",
       "      <td>0.740625</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 215 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         dbGaP SubjID     sex   race  age  agex2   bmi    tf   tr    af    ar  \\\n",
       "pd10016        445969  female  white   29    841  19.9   6.7  0.0  10.7   0.0   \n",
       "pd10018        445970    male  white   30    900  28.3  27.3  0.0  22.0   0.0   \n",
       "pd10023        445971    male  white   27    729  21.7  27.3  0.0  55.0   0.0   \n",
       "pd10028        445972    male  black   31    961    29  16.0  0.0  36.8  14.8   \n",
       "pd10031        445973    male  white   28    784  21.5  20.0  0.0  16.7   0.0   \n",
       "\n",
       "         ...  topmed_r03_PGS001357  topmed_r03_PGS002114  \\\n",
       "pd10016  ...             -0.032170              2.662745   \n",
       "pd10018  ...             -0.232240              1.028924   \n",
       "pd10023  ...             -0.191068              3.098229   \n",
       "pd10028  ...             -0.251807              3.165720   \n",
       "pd10031  ...             -0.204522              3.074076   \n",
       "\n",
       "         topmed_r03_PGS001105  topmed_r03_PGS000349  topmed_r03_PGS001917  \\\n",
       "pd10016              4.381828              0.899946              0.011109   \n",
       "pd10018              1.048539              1.140902             -0.013622   \n",
       "pd10023              1.721006              0.694613              0.001444   \n",
       "pd10028              4.340293              0.627491              0.029261   \n",
       "pd10031              2.298869              1.016425              0.006443   \n",
       "\n",
       "         topmed_r03_PGS001818  topmed_r03_PGS000957  topmed_r03_PGS000889  \\\n",
       "pd10016             -0.147684              0.027173              0.945526   \n",
       "pd10018              0.644547              0.874353              0.870190   \n",
       "pd10023              0.470915              0.735025              1.365498   \n",
       "pd10028              0.090178              1.278612             -0.102817   \n",
       "pd10031             -0.178284              0.897139             -0.176216   \n",
       "\n",
       "         topmed_r03_PGS000018   Case  \n",
       "pd10016              0.634510  False  \n",
       "pd10018              1.261631  False  \n",
       "pd10023              0.623044  False  \n",
       "pd10028              1.027068  False  \n",
       "pd10031              0.740625  False  \n",
       "\n",
       "[5 rows x 215 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign case/control classes to samples using top quartile rule splitting by sex\n",
    "\n",
    "# Phenotype of interest\n",
    "phenotype = \"cr\"\n",
    "\n",
    "# Calculate thresholds for each race and sex\n",
    "print( \"Thresholds to be used for case/control definition:\" )\n",
    "thresholds = {}\n",
    "for race in [\"white\", \"black\"]:\n",
    "    for sex in [\"male\", \"female\"]:\n",
    "        \n",
    "        # Define Q3 as threshold and save it in thresholds dictionary\n",
    "        key = race + \"_\" + sex\n",
    "        thresholds[ key ] = sea_dataset_full[ (( sea_dataset_full[\"race\"]==race ) & ( sea_dataset_full[\"sex\"]==sex ))  ][phenotype].quantile(0.75)\n",
    "        \n",
    "        # Print values and warnings, if needed\n",
    "        note = \"\"\n",
    "        if( thresholds[ key ] == 0 ):\n",
    "            note = \"(WARNING: This group has Q3 equals zero. Considering only non-zeros as CASE group!!)\"\n",
    "        print( \" - {0} = {1:.4f} {2}\".format( key, thresholds[key], note ) )\n",
    "\n",
    "# Add a new column called Case with all values equals to False\n",
    "sea_dataset_full = sea_dataset_full.assign( Case=False )\n",
    "\n",
    "# Identify subjects above threshold to assign them to group \"Case\"\n",
    "for i in sea_dataset_full.index:\n",
    "\n",
    "    sex = sea_dataset_full.loc[ i, \"sex\"]\n",
    "    race = sea_dataset_full.loc[ i, \"race\"]\n",
    "    key = race + \"_\" + sex\n",
    "    \n",
    "    if( sea_dataset_full.loc[i, phenotype] > thresholds[ key ] ):\n",
    "        sea_dataset_full.loc[i, \"Case\"] = True\n",
    "\n",
    "# Print summary of Case/Controls for each subgroup\n",
    "print( sea_dataset_full.groupby([\"race\",\"sex\",\"Case\"])[\"Case\"].count() )\n",
    "sea_dataset_full.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 3)** **Standardize** raw scores and calculate **Odds Ratios** between case/controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_ref</th>\n",
       "      <th>race</th>\n",
       "      <th>reference_panel</th>\n",
       "      <th>threshold</th>\n",
       "      <th>GRS</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>score_pvalue</th>\n",
       "      <th>conf_interval_lower</th>\n",
       "      <th>conf_interval_upper</th>\n",
       "      <th>name</th>\n",
       "      <th>...</th>\n",
       "      <th>trait_reported</th>\n",
       "      <th>trait_mapped</th>\n",
       "      <th>trait_efo</th>\n",
       "      <th>weight_type</th>\n",
       "      <th>genome_build</th>\n",
       "      <th>variants_number</th>\n",
       "      <th>pgp_id</th>\n",
       "      <th>citation</th>\n",
       "      <th>pgs_catalog_hyperlink</th>\n",
       "      <th>license</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>white_hrc_r03_PGS000013</th>\n",
       "      <td>hrc_r03_PGS000013</td>\n",
       "      <td>white</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r03</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>1.43122</td>\n",
       "      <td>0.000648759</td>\n",
       "      <td>1.16472</td>\n",
       "      <td>1.75871</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>...</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>6630150</td>\n",
       "      <td>PGP000006</td>\n",
       "      <td>Khera AV et al. Nat Genet (2018). doi:10.1038/...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS000013/</td>\n",
       "      <td>Freely available to the academic community for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>white_topmed_r0_PGS000013</th>\n",
       "      <td>topmed_r0_PGS000013</td>\n",
       "      <td>white</td>\n",
       "      <td>topmed</td>\n",
       "      <td>r0</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>1.41677</td>\n",
       "      <td>0.000907223</td>\n",
       "      <td>1.15324</td>\n",
       "      <td>1.74052</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>...</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>6630150</td>\n",
       "      <td>PGP000006</td>\n",
       "      <td>Khera AV et al. Nat Genet (2018). doi:10.1038/...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS000013/</td>\n",
       "      <td>Freely available to the academic community for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>white_topmed_r03_PGS000013</th>\n",
       "      <td>topmed_r03_PGS000013</td>\n",
       "      <td>white</td>\n",
       "      <td>topmed</td>\n",
       "      <td>r03</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>1.40907</td>\n",
       "      <td>0.00108849</td>\n",
       "      <td>1.14702</td>\n",
       "      <td>1.73098</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>...</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>6630150</td>\n",
       "      <td>PGP000006</td>\n",
       "      <td>Khera AV et al. Nat Genet (2018). doi:10.1038/...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS000013/</td>\n",
       "      <td>Freely available to the academic community for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_wGRS49</th>\n",
       "      <td>hrc_r08_wGRS49</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>wGRS49</td>\n",
       "      <td>1.40481</td>\n",
       "      <td>0.00165781</td>\n",
       "      <td>1.13668</td>\n",
       "      <td>1.73619</td>\n",
       "      <td>wGRS49</td>\n",
       "      <td>...</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>beta</td>\n",
       "      <td>hg19</td>\n",
       "      <td>49</td>\n",
       "      <td>wGRS49</td>\n",
       "      <td>Circ Cardiovasc Genet. 2015;8:803-811. DOI: 10...</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>white_hrc_r0_PGS000013</th>\n",
       "      <td>hrc_r0_PGS000013</td>\n",
       "      <td>white</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r0</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>1.40474</td>\n",
       "      <td>0.00113998</td>\n",
       "      <td>1.14467</td>\n",
       "      <td>1.72389</td>\n",
       "      <td>PGS000013</td>\n",
       "      <td>...</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>6630150</td>\n",
       "      <td>PGP000006</td>\n",
       "      <td>Khera AV et al. Nat Genet (2018). doi:10.1038/...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS000013/</td>\n",
       "      <td>Freely available to the academic community for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS002009</th>\n",
       "      <td>hrc_r08_PGS002009</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS002009</td>\n",
       "      <td>0.839041</td>\n",
       "      <td>0.100641</td>\n",
       "      <td>0.680446</td>\n",
       "      <td>1.0346</td>\n",
       "      <td>PGS002009</td>\n",
       "      <td>...</td>\n",
       "      <td>Systolic blood pressure, automated reading</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>EFO_0006335</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>68449</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS002009/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r03_PGS002037</th>\n",
       "      <td>hrc_r03_PGS002037</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r03</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>0.838776</td>\n",
       "      <td>0.0937109</td>\n",
       "      <td>0.682909</td>\n",
       "      <td>1.03022</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>...</td>\n",
       "      <td>Tobacco use disorder</td>\n",
       "      <td>nicotine dependence</td>\n",
       "      <td>EFO_0003768</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>847691</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS002037/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS001900</th>\n",
       "      <td>hrc_r08_PGS001900</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS001900</td>\n",
       "      <td>0.823963</td>\n",
       "      <td>0.0713973</td>\n",
       "      <td>0.667562</td>\n",
       "      <td>1.01701</td>\n",
       "      <td>PGS001900</td>\n",
       "      <td>...</td>\n",
       "      <td>Diastolic blood pressure, automated reading</td>\n",
       "      <td>diastolic blood pressure</td>\n",
       "      <td>EFO_0006336</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>66335</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS001900/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r05_PGS002037</th>\n",
       "      <td>hrc_r05_PGS002037</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r05</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>0.818149</td>\n",
       "      <td>0.0560459</td>\n",
       "      <td>0.665912</td>\n",
       "      <td>1.00519</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>...</td>\n",
       "      <td>Tobacco use disorder</td>\n",
       "      <td>nicotine dependence</td>\n",
       "      <td>EFO_0003768</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>847691</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS002037/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS002037</th>\n",
       "      <td>hrc_r08_PGS002037</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>0.71258</td>\n",
       "      <td>0.00158226</td>\n",
       "      <td>0.577471</td>\n",
       "      <td>0.879299</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>...</td>\n",
       "      <td>Tobacco use disorder</td>\n",
       "      <td>nicotine dependence</td>\n",
       "      <td>EFO_0003768</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>847691</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS002037/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>400 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       model_ref   race reference_panel  \\\n",
       "white_hrc_r03_PGS000013        hrc_r03_PGS000013  white             hrc   \n",
       "white_topmed_r0_PGS000013    topmed_r0_PGS000013  white          topmed   \n",
       "white_topmed_r03_PGS000013  topmed_r03_PGS000013  white          topmed   \n",
       "black_hrc_r08_wGRS49              hrc_r08_wGRS49  black             hrc   \n",
       "white_hrc_r0_PGS000013          hrc_r0_PGS000013  white             hrc   \n",
       "...                                          ...    ...             ...   \n",
       "black_hrc_r08_PGS002009        hrc_r08_PGS002009  black             hrc   \n",
       "black_hrc_r03_PGS002037        hrc_r03_PGS002037  black             hrc   \n",
       "black_hrc_r08_PGS001900        hrc_r08_PGS001900  black             hrc   \n",
       "black_hrc_r05_PGS002037        hrc_r05_PGS002037  black             hrc   \n",
       "black_hrc_r08_PGS002037        hrc_r08_PGS002037  black             hrc   \n",
       "\n",
       "                           threshold        GRS odds_ratio score_pvalue  \\\n",
       "white_hrc_r03_PGS000013          r03  PGS000013    1.43122  0.000648759   \n",
       "white_topmed_r0_PGS000013         r0  PGS000013    1.41677  0.000907223   \n",
       "white_topmed_r03_PGS000013       r03  PGS000013    1.40907   0.00108849   \n",
       "black_hrc_r08_wGRS49             r08     wGRS49    1.40481   0.00165781   \n",
       "white_hrc_r0_PGS000013            r0  PGS000013    1.40474   0.00113998   \n",
       "...                              ...        ...        ...          ...   \n",
       "black_hrc_r08_PGS002009          r08  PGS002009   0.839041     0.100641   \n",
       "black_hrc_r03_PGS002037          r03  PGS002037   0.838776    0.0937109   \n",
       "black_hrc_r08_PGS001900          r08  PGS001900   0.823963    0.0713973   \n",
       "black_hrc_r05_PGS002037          r05  PGS002037   0.818149    0.0560459   \n",
       "black_hrc_r08_PGS002037          r08  PGS002037    0.71258   0.00158226   \n",
       "\n",
       "                           conf_interval_lower conf_interval_upper       name  \\\n",
       "white_hrc_r03_PGS000013                1.16472             1.75871  PGS000013   \n",
       "white_topmed_r0_PGS000013              1.15324             1.74052  PGS000013   \n",
       "white_topmed_r03_PGS000013             1.14702             1.73098  PGS000013   \n",
       "black_hrc_r08_wGRS49                   1.13668             1.73619     wGRS49   \n",
       "white_hrc_r0_PGS000013                 1.14467             1.72389  PGS000013   \n",
       "...                                        ...                 ...        ...   \n",
       "black_hrc_r08_PGS002009               0.680446              1.0346  PGS002009   \n",
       "black_hrc_r03_PGS002037               0.682909             1.03022  PGS002037   \n",
       "black_hrc_r08_PGS001900               0.667562             1.01701  PGS001900   \n",
       "black_hrc_r05_PGS002037               0.665912             1.00519  PGS002037   \n",
       "black_hrc_r08_PGS002037               0.577471            0.879299  PGS002037   \n",
       "\n",
       "                            ...                               trait_reported  \\\n",
       "white_hrc_r03_PGS000013     ...                      Coronary artery disease   \n",
       "white_topmed_r0_PGS000013   ...                      Coronary artery disease   \n",
       "white_topmed_r03_PGS000013  ...                      Coronary artery disease   \n",
       "black_hrc_r08_wGRS49        ...                      Coronary artery disease   \n",
       "white_hrc_r0_PGS000013      ...                      Coronary artery disease   \n",
       "...                         ...                                          ...   \n",
       "black_hrc_r08_PGS002009     ...   Systolic blood pressure, automated reading   \n",
       "black_hrc_r03_PGS002037     ...                         Tobacco use disorder   \n",
       "black_hrc_r08_PGS001900     ...  Diastolic blood pressure, automated reading   \n",
       "black_hrc_r05_PGS002037     ...                         Tobacco use disorder   \n",
       "black_hrc_r08_PGS002037     ...                         Tobacco use disorder   \n",
       "\n",
       "                                        trait_mapped    trait_efo weight_type  \\\n",
       "white_hrc_r03_PGS000013      coronary artery disease  EFO_0001645          NR   \n",
       "white_topmed_r0_PGS000013    coronary artery disease  EFO_0001645          NR   \n",
       "white_topmed_r03_PGS000013   coronary artery disease  EFO_0001645          NR   \n",
       "black_hrc_r08_wGRS49         coronary artery disease  EFO_0001645        beta   \n",
       "white_hrc_r0_PGS000013       coronary artery disease  EFO_0001645          NR   \n",
       "...                                              ...          ...         ...   \n",
       "black_hrc_r08_PGS002009      systolic blood pressure  EFO_0006335        beta   \n",
       "black_hrc_r03_PGS002037          nicotine dependence  EFO_0003768        beta   \n",
       "black_hrc_r08_PGS001900     diastolic blood pressure  EFO_0006336        beta   \n",
       "black_hrc_r05_PGS002037          nicotine dependence  EFO_0003768        beta   \n",
       "black_hrc_r08_PGS002037          nicotine dependence  EFO_0003768        beta   \n",
       "\n",
       "                           genome_build variants_number     pgp_id  \\\n",
       "white_hrc_r03_PGS000013            hg19         6630150  PGP000006   \n",
       "white_topmed_r0_PGS000013          hg19         6630150  PGP000006   \n",
       "white_topmed_r03_PGS000013         hg19         6630150  PGP000006   \n",
       "black_hrc_r08_wGRS49               hg19              49     wGRS49   \n",
       "white_hrc_r0_PGS000013             hg19         6630150  PGP000006   \n",
       "...                                 ...             ...        ...   \n",
       "black_hrc_r08_PGS002009          GRCh37           68449  PGP000263   \n",
       "black_hrc_r03_PGS002037          GRCh37          847691  PGP000263   \n",
       "black_hrc_r08_PGS001900          GRCh37           66335  PGP000263   \n",
       "black_hrc_r05_PGS002037          GRCh37          847691  PGP000263   \n",
       "black_hrc_r08_PGS002037          GRCh37          847691  PGP000263   \n",
       "\n",
       "                                                                     citation  \\\n",
       "white_hrc_r03_PGS000013     Khera AV et al. Nat Genet (2018). doi:10.1038/...   \n",
       "white_topmed_r0_PGS000013   Khera AV et al. Nat Genet (2018). doi:10.1038/...   \n",
       "white_topmed_r03_PGS000013  Khera AV et al. Nat Genet (2018). doi:10.1038/...   \n",
       "black_hrc_r08_wGRS49        Circ Cardiovasc Genet. 2015;8:803-811. DOI: 10...   \n",
       "white_hrc_r0_PGS000013      Khera AV et al. Nat Genet (2018). doi:10.1038/...   \n",
       "...                                                                       ...   \n",
       "black_hrc_r08_PGS002009     Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "black_hrc_r03_PGS002037     Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "black_hrc_r08_PGS001900     Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "black_hrc_r05_PGS002037     Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "black_hrc_r08_PGS002037     Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "\n",
       "                                                  pgs_catalog_hyperlink  \\\n",
       "white_hrc_r03_PGS000013     https://www.pgscatalog.org/score/PGS000013/   \n",
       "white_topmed_r0_PGS000013   https://www.pgscatalog.org/score/PGS000013/   \n",
       "white_topmed_r03_PGS000013  https://www.pgscatalog.org/score/PGS000013/   \n",
       "black_hrc_r08_wGRS49                                                      \n",
       "white_hrc_r0_PGS000013      https://www.pgscatalog.org/score/PGS000013/   \n",
       "...                                                                 ...   \n",
       "black_hrc_r08_PGS002009     https://www.pgscatalog.org/score/PGS002009/   \n",
       "black_hrc_r03_PGS002037     https://www.pgscatalog.org/score/PGS002037/   \n",
       "black_hrc_r08_PGS001900     https://www.pgscatalog.org/score/PGS001900/   \n",
       "black_hrc_r05_PGS002037     https://www.pgscatalog.org/score/PGS002037/   \n",
       "black_hrc_r08_PGS002037     https://www.pgscatalog.org/score/PGS002037/   \n",
       "\n",
       "                                                                      license  \n",
       "white_hrc_r03_PGS000013     Freely available to the academic community for...  \n",
       "white_topmed_r0_PGS000013   Freely available to the academic community for...  \n",
       "white_topmed_r03_PGS000013  Freely available to the academic community for...  \n",
       "black_hrc_r08_wGRS49                                                      NaN  \n",
       "white_hrc_r0_PGS000013      Freely available to the academic community for...  \n",
       "...                                                                       ...  \n",
       "black_hrc_r08_PGS002009                                                   NaN  \n",
       "black_hrc_r03_PGS002037                                                   NaN  \n",
       "black_hrc_r08_PGS001900                                                   NaN  \n",
       "black_hrc_r05_PGS002037                                                   NaN  \n",
       "black_hrc_r08_PGS002037                                                   NaN  \n",
       "\n",
       "[400 rows x 34 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize raw scores and calculate OR\n",
    "\n",
    "li = []\n",
    "standardize_scores = []\n",
    "logit_models_dict = {}\n",
    "models_summaries = pd.DataFrame()\n",
    "\n",
    "# Get list of panels, thresholds and GRSs used from info dictonary\n",
    "reference_panels = set( [ key.split('_')[0] for key in scores_info_dict.keys() ] )\n",
    "thresholds = set( [ key.split('_')[1] for key in scores_info_dict.keys() ] )\n",
    "grss = set( [ key.split('_')[2] for key in scores_info_dict.keys() ] )\n",
    "\n",
    "# Normalize all scores with mean = 0 and SD = 1\n",
    "for race in [\"white\", \"black\"]:\n",
    "    \n",
    "    # Subset cohort between whites and blacks to run logistic regression individually\n",
    "    sea_dataset_subset = sea_dataset_full[ sea_dataset_full[\"race\"] == race ]\n",
    "\n",
    "    # Create a dependent variable named \"Case_recoded\" based on column \"Case\", conding it as 0 and 1 to fit glm\n",
    "    sea_dataset_subset = sea_dataset_subset.assign( Case_recoded=sea_dataset_subset[\"Case\"].replace(True, 1).replace(False, 0) )\n",
    "    \n",
    "    for reference_panel_id in reference_panels:\n",
    "        \n",
    "        for threshold in thresholds:\n",
    "\n",
    "            for grs in grss:\n",
    "\n",
    "                raw_score_name = \"_\".join( (reference_panel_id, threshold, grs) )\n",
    "                standardize_score_name = \"_\".join( (race, reference_panel_id, threshold, grs) )\n",
    "                standardize_scores.append(standardize_score_name)\n",
    "        \n",
    "                # Standardize raw scores using method scale\n",
    "                sea_dataset_subset[ standardize_score_name ] = preprocessing.scale( sea_dataset_subset[ raw_score_name ] )\n",
    "        \n",
    "                # Fit a logistic model using standardize scores and save it on dictionary \n",
    "                logit_models_dict[ standardize_score_name ] = sm.formula.glm(\n",
    "                    \"Case_recoded ~ \" + standardize_score_name + \" + age + sex\",\n",
    "                    family=sm.families.Binomial(),\n",
    "                    data=sea_dataset_subset).fit()\n",
    "\n",
    "                # Create dictonary with scores performances\n",
    "                model_summary_series = {\n",
    "                        \"model_ref\": raw_score_name,\n",
    "                        \"race\": race,\n",
    "                        \"reference_panel\": reference_panel_id,\n",
    "                        \"threshold\": threshold,\n",
    "                        \"GRS\": grs,\n",
    "                        \"odds_ratio\": np.exp( logit_models_dict[ standardize_score_name ].params )[ standardize_score_name ],\n",
    "                        \"score_pvalue\": logit_models_dict[ standardize_score_name ].pvalues[ standardize_score_name ],\n",
    "                        \"conf_interval_lower\": np.exp( logit_models_dict[ standardize_score_name ].conf_int()[0][ standardize_score_name ] ),\n",
    "                        \"conf_interval_upper\": np.exp( logit_models_dict[ standardize_score_name ].conf_int()[1][ standardize_score_name ] )\n",
    "                    }\n",
    "                # Add additional info from info files (coverage, etc)\n",
    "                model_summary_series.update( scores_info_dict[raw_score_name][race + \"s\"] )\n",
    "                \n",
    "                # Add metadata about the GRSs to the output\n",
    "                model_summary_series.update( pgs_metadata[grs] )\n",
    "                \n",
    "                # Transform dictionary on a pandas Series and name it as <standardize_score_name>\n",
    "                model_summary_series = pd.Series(\n",
    "                    data=model_summary_series,\n",
    "                    name=standardize_score_name\n",
    "                )\n",
    "                                \n",
    "                li.append(model_summary_series)\n",
    "                     \n",
    "# Concatenate results for all models tested and transpose object to make visualization easier \n",
    "models_summaries = pd.concat( li, axis=1 ).transpose().sort_values(\"odds_ratio\", ascending=False)\n",
    "models_summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export dataframe with summaries on a xls file on the exports folder\n",
    "\n",
    "writer = pd.ExcelWriter(\n",
    "    path = \"/labs/tassimes/rodrigoguarischi/projects/sea/exports/SEA_models_summary.xls\",\n",
    "    engine = 'xlsxwriter'\n",
    "    )\n",
    "\n",
    "models_summaries.to_excel(writer, sheet_name='models_summary')\n",
    "\n",
    "# Get access to the workbook and sheet\n",
    "workbook = writer.book\n",
    "worksheet = writer.sheets['models_summary']\n",
    "\n",
    "# Automatically add filters to columns\n",
    "worksheet.autofilter( \"A1:AH{0}\".format(len( models_summaries.index ) + 1) )\n",
    "\n",
    "# Add a percent format with 2 decimal points\n",
    "percent_fmt = workbook.add_format({'num_format': '0.00%', 'align': 'left'})\n",
    "\n",
    "# Add a left-alignment format\n",
    "left_alignment_fmt = workbook.add_format({'align': 'left'})\n",
    "\n",
    "# Set all columns to left alignment\n",
    "worksheet.set_column('A:AH', 15, left_alignment_fmt )\n",
    "\n",
    "# Format the columns by width and include number formats\n",
    "worksheet.set_column('A:B', 30, left_alignment_fmt)\n",
    "worksheet.set_column('Z:AA', 40, left_alignment_fmt)\n",
    "worksheet.set_column('AG:AG', 70, left_alignment_fmt)\n",
    "worksheet.set_column('AH:AH', 40, left_alignment_fmt)\n",
    "\n",
    "# Format coverage column as percent\n",
    "worksheet.set_column('O:O', 15, percent_fmt)\n",
    "\n",
    "for model_index in range(0, len(models_summaries)):\n",
    "\n",
    "    row_number = model_index + 1    \n",
    "    model_index_name = models_summaries.index[model_index]\n",
    "    \n",
    "    # models_summaries.loc[model_index, \"pgs_catalog_hyperlink\"]\n",
    "    hyperlink = models_summaries.loc[model_index_name,\"pgs_catalog_hyperlink\"]\n",
    "\n",
    "    # Determine where we will place the formula\n",
    "    cell_location = xl_rowcol_to_cell( row_number, 33)\n",
    "    \n",
    "    # print( \"\\t\".join( [str( row_number ), model_index_name, str(hyperlink), cell_location] ) )\n",
    "    worksheet.write_url(cell_location, hyperlink)\n",
    "        \n",
    "# Save modifications\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export dataframe with summaries on a xls file on the exports folder\n",
    "\n",
    "models_summaries.to_excel( \n",
    "  \"/labs/tassimes/rodrigoguarischi/projects/sea/exports/SEA_models_summary.xls\",\n",
    "  sheet_name=\"models_summary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_name = \"White_hrc_r03_wGRS49\"\n",
    "print( logit_models_dict[model_name].summary() )\n",
    "# print( logit_models_dict[model_name].pvalues[model_name] )\n",
    "# print( np.exp( logit_models_dict[model_name].params )[ model_name ] )\n",
    "# print( np.exp( logit_models_dict[model_name].conf_int().loc[[ model_name ]] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot boxplots and histograms and boxplots of raw scores spliting by sex\n",
    "\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "score_list = [\"wGRS49_r0\", \"wGRS49_r03\", \"wGRS49_r05\", \"wGRS49_r08\"]\n",
    "# score_list = [\"PGS000349_r0\", \"PGS000349_r03\", \"PGS000349_r05\", \"PGS000349_r08\"]\n",
    "# score_list = [\"PGS000018_r0\", \"PGS000018_r03\", \"PGS000018_r05\", \"PGS000018_r08\"]\n",
    "\n",
    "fig, axs = plt.subplots(nrows=3, ncols=4, figsize=(5*len(score_list),12))\n",
    "for i in range(0, len(score_list)):\n",
    "    sns.boxplot(\n",
    "        x = \"Case\",\n",
    "        y = score_list[i],\n",
    "        data = sea_merged_whites,\n",
    "        palette = reversed(sns.color_palette(n_colors=2)),\n",
    "        width=0.4,\n",
    "        fliersize=2,\n",
    "        ax=axs[0,i],\n",
    "    )\n",
    "    sns.histplot(\n",
    "        x = score_list[i],\n",
    "        data = sea_merged_whites[ sea_merged_whites[\"sex\"] == \"Male\" ],\n",
    "        hue=\"Case\",\n",
    "        hue_order=[True, False],\n",
    "        kde=True,\n",
    "        ax=axs[1,i]\n",
    "        ).set(title='Males (n={0})'.format( sum( sea_merged_whites[\"sex\"] == \"Male\" ) ) )\n",
    "    sns.histplot(\n",
    "        x = score_list[i],\n",
    "        data = sea_merged_whites[ sea_merged_whites[\"sex\"] == \"Female\" ],\n",
    "        hue=\"Case\",\n",
    "        hue_order=[True, False],\n",
    "        kde=True,\n",
    "        ax=axs[2,i]\n",
    "    ).set(title='Females (n={0})'.format( sum( sea_merged_whites[\"sex\"] == \"Female\" ) ) )\n",
    "\n",
    "plt.subplots_adjust(top=1.25)\n",
    "\n",
    "# # Plot only boxplots\n",
    "\n",
    "# fig, axs = plt.subplots(ncols=len(score_list), figsize=(5*len(score_list),4))\n",
    "# for i in range(0, len(score_list)):\n",
    "#     sns.boxplot(\n",
    "#         x = \"Case\",\n",
    "#         y = score_list[i],\n",
    "#         data = sea_merged_whites,\n",
    "#         hue=\"sex\",\n",
    "#         hue_order=['Male','Female'],\n",
    "#         width=0.4,\n",
    "#         fliersize=2,\n",
    "#         ax=axs[i],\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test variables\n",
    "sns.boxplot(\n",
    "    x = \"Case\",\n",
    "    y = \"age\",\n",
    "    data = sea_merged_whites,\n",
    "    width=0.4,\n",
    "    fliersize=2\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export dataset to CSV files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export full dataset\n",
    "sea_merged_whites.to_csv(\"exports/sea_whites_phenotypes_scores_full.csv\")\n",
    "\n",
    "# Export only subset of columns Tim asked\n",
    "select_columns = [\"dbGaP_SubjID\",\"sex\",\"race\",\"age\",\"agex2\",\"bmi\",\"cr\",\"wGRS49_r05\",\"PGS000349_r05\",\"PGS000018_r05\",\"wGRS49_r03\",\"PGS000349_r03\",\"PGS000018_r03\",\"wGRS49_r0\",\"PGS000349_r0\",\"PGS000018_r0\",\"wGRS49_r08\",\"PGS000349_r08\",\"PGS000018_r08\", \"Case\"]\n",
    "sea_merged_whites[select_columns].to_csv(\"exports/sea_whites_phenotypes_scores_column_subset.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "sns.scatterplot(\n",
    "    x = \"wGRS49_r03_scaled\",\n",
    "    y = \"Case\",\n",
    "    data = sea_merged_whites\n",
    "    # ,\n",
    "    # alpha=0.4,\n",
    "    # s=6\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test logistic regression in <code>R</code>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Read CSV file with raw scores\n",
    "input_table <- read.csv(\"exports/sea_whites_phenotypes_scores_full.csv\", row.names=1)\n",
    "\n",
    "# Normalize all scores with mean = 0 and SD = 1\n",
    "for ( grs in c(\"wGRS49\",\"PGS000349\",\"PGS000018\", \"PGS000667\", \"PGS000889\") ){\n",
    "    for ( threshold in c(\"r0\", \"r03\", \"r05\", \"r08\") ){\n",
    "        current_score <- paste0( grs, \"_\", threshold);\n",
    "        score_values <- input_table[[ current_score ]]\n",
    "        score_mean <- mean( score_values )\n",
    "        score_sd <- sd( score_values )\n",
    "        input_table[ paste0(current_score, \"_scaled\") ] <- (score_values-score_mean)/score_sd\n",
    "        # input_table[ paste0(current_score, \"_scaled\") ] <- scale( score_values )\n",
    "    }\n",
    "}\n",
    "\n",
    "# Test Standardization (check values of min, mean, max and sd)\n",
    "normalization_summary <- function(current_scaled_score){\n",
    "  return (\n",
    "    list(\n",
    "      \"min\"=min(current_scaled_score),\n",
    "      \"mean\"=mean(current_scaled_score),\n",
    "      \"max\"=max(current_scaled_score),\n",
    "      \"sd\"=sd(current_scaled_score)\n",
    "    )\n",
    "  )\n",
    "}\n",
    "t( sapply( input_table[ , grepl(\"_scaled\", names(input_table)) ], normalization_summary ) )\n",
    "\n",
    "# Test number of case/controls for each sex and age\n",
    "xtabs( ~ Case + sex, data=input_table)\n",
    "xtabs( ~ Case + age, data=input_table)\n",
    "\n",
    "# Apply logistic regression to scaled scores\n",
    "models <- list()\n",
    "for ( score_scaled_name in grep(\"_scaled\", names(input_table), value = T) ){\n",
    "  \n",
    "  # Skip PGSs with 0 SNPs evaluated (scores equals to NaN)  \n",
    "  if( is.na( median( input_table[, score_scaled_name ] ) ) ) next;\n",
    "  \n",
    "  input_table_subset <- input_table[, c(score_scaled_name, \"age\", \"sex\", \"Case\") ]\n",
    "  colnames( input_table_subset ) <- c(\"score_scaled\", \"age\", \"sex\", \"Case\")\n",
    "  models[[ score_scaled_name ]] <- glm( Case ~ score_scaled + age + sex, data = input_table_subset, family = \"binomial\")\n",
    "}\n",
    "\n",
    "# Print Odds Ratios for each score at 4 levels of imputation Rsqrt thresholding (0, 0.3, 0.5 and 0.8)\n",
    "t( sapply( models, function(current_model) exp( current_model$coefficients ) ) )\n",
    "\n",
    "# Get p-values\n",
    "t( sapply( models, function(current_model) coef(summary(current_model))[,4] ) )\n",
    "\n",
    "summary( models$wGRS49_r0_scaled )\n",
    "\n",
    "---\n",
    "apply_logit <- function(current_scaled_score, input_table) {\n",
    "    logit_formula <- as.formula( paste0(\"Case ~ \" , current_scaled_score ,\" + age + sex\") )\n",
    "    logistic_model <- glm( logit_formula, data = input_table, family = \"binomial\")\n",
    "    return( logistic_model )\n",
    "} \n",
    "scaled_scores <- grep(\"_scaled\", names(input_table), value = T)\n",
    "models <- lapply( scaled_scores, function(current_scaled_score) apply_logit(current_scaled_score, input_table) )\n",
    "names(models) <- scaled_scores\n",
    "\n",
    "----\n",
    "\n",
    "summary( models$wGRS49_r03_scaled )\n",
    "str( models$wGRS49_r03_scaled )\n",
    "\n",
    "table( ifelse( models$wGRS49_r03_scaled$fitted.values > 0.5, \"assigned_case\", \"assigned_control\") )\n",
    "\n",
    "summary( models[[2]] )\n",
    "\n",
    "create_pdf_plot <- function(model, input_data) {\n",
    "\n",
    "    library(ggplot2)\n",
    "    library(cowplot)\n",
    "\n",
    "    predicted_data <- data.frame(\n",
    "        probability_of_hd = model$fitted.values,\n",
    "        cr = input_table$Case\n",
    "        )\n",
    "\n",
    "    predicted_data <- predicted_data[ order(predicted_data$probability_of_hd, decreasing=FALSE), ]\n",
    "    predicted_data$rank <- 1:nrow(predicted_data)\n",
    "\n",
    "    ggplot( data=predicted_data, aes(x=rank, y=probability_of_hd) ) +\n",
    "        geom_point(aes(color=cr), alpha=1, shape=4, stroke=2) +\n",
    "        xlab(\"Index\") +\n",
    "        ylab(\"Predicted probability of getting heart disease\")\n",
    "    ggsave(\"heart_disease_probabilities.pdf\")\n",
    "}\n",
    "\n",
    "create_pdf_plot(models[[1]], input_table)\n",
    "\n",
    "-----\n",
    "\n",
    "# Test model based on multiple scores\n",
    "logistic_model <- glm( Case ~ ., data = input_table[,c(scaled_scores, \"age\", \"sex\", \"Case\")], family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "-----\n",
    "\n",
    "summary( models[, \"wGRS49_r0_scaled\"] )\n",
    "\n",
    "t( sapply( scaled_scores, function(x) apply_logit(x, input_table) ) )\n",
    "\n",
    "t( sapply( grep(\"_scaled\", names(input_table), value = T), function(x) apply_logit(x, input_table) ) )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "sapply( grep(\"_scaled\", names(input_table), value = T), function(x) logit_formula <- print0(\"Case ~ \" , x ,\" + age + sex\"; print(logit_formula) )\n",
    "\n",
    "sapply( 1:5, function(x) x*2)\n",
    "\n",
    "logistic_model <- glm(Case ~ wGRS49_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "logistic_model <- glm(Case ~ PGS000349_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "logistic_model <- glm(Case ~ PGS000018_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "input_table_subset <- input_table[ , c( grep(\"_scaled\", names(input_table), value = T), \"age\", \"sex\", \"Case\") ]\n",
    "\n",
    "logistic_model <- glm(Case~., data = input_table_subset, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "// input_table[ paste0(current_score, \"_scaled\") ] <- scale( input_table[[ current_score ]], center = 0, scale = 1)\n",
    "\n",
    "range( input_table$PGS000018_r03_scaled )\n",
    "input_table$wGRS49_r03_scaled <- scale(input_table$wGRS49_r03, center = 0, scale = 1)\n",
    "input_table$PGS000018_r03_scaled <- scale(input_table$PGS000018_r03, center = 0, scale = 1)\n",
    "\n",
    "logistic_model <- glm(Case ~ wGRS49_r03_scaled, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "\n",
    "# Predict test data based on model\n",
    "predict_reg <- predict(logistic_model, input_table, type = \"response\")\n",
    "range(predict_reg)\n",
    "predict_reg <- ifelse( predict_reg > 0.5, 1, 0)\n",
    "\n",
    "range(predict_reg)\n",
    "logistic_model$\n",
    "\n",
    "predict_reg <- <- predict(logistic_model, logistic_model, type = \"response\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# table = df.groupby(level=\"Cancer\").sum().values\n",
    "\n",
    "table = [[840,51663],[32,5053]]\n",
    "print(table)\n",
    "oddsratio, pvalue = stats.fisher_exact(table)\n",
    "print(\"OddsR: \", oddsratio, \"p-Value:\", pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sea_merged_whites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "for score in [\"wGRS49_r03\", \"wGRS49_r05\", \"wGRS49_r08\", \"PGS000018_r03\", \"PGS000018_r05\", \"PGS000018_r08\"]:\n",
    "    for traits in [\"tf_quartile4\", \"tr_quartile4\", \"af_quartile4\", \"ar_quartile4\", \"cf_quartile4\", \"cr_quartile4\"]:\n",
    "        # score = \"PGS000018_r08\"\n",
    "        cases = sea_merged_whites[traits]\n",
    "        print( score , \"\\t\", traits, \"\\t\", stats.ttest_ind( sea_merged_whites[cases][score], sea_merged_whites[np.invert(cases)][score] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "\n",
    "current_score_file = \"./apply_grs/wGRS49_PGS000349_PGS000018_r0.scores.txt\"\n",
    "\n",
    "pgs_results = pd.read_table( current_score_file, sep = \",\" )\n",
    "pgs_results[\"sample\"] = pgs_results[\"sample\"].str.split(\"_\", expand = True)[0]\n",
    "pgs_results = pgs_results.set_index(\"sample\")\n",
    "min_r2_used = os.path.basename( current_score_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "pgs_results = pgs_results.add_suffix(\"_\" + min_r2_used)\n",
    "pgs_results.head()\n",
    "    # score_name = os.path.basename( current_score_file ).split(\".\")[0]\n",
    "    # pgs_results.rename( columns={ pgs_results.columns[0]:score_name }, inplace=True )\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "kernelspec": {
   "display_name": "Python 3.6.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
