{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Workflow used for:\n",
    "  1. Load metadata about GRSs used\n",
    "  1. Read phenotypic information and consolidate it with raw PGS scores from **apply_grs.ipynb**\n",
    "  1. Standardize raw scores and calculate Odds Ratios between case/controls\n",
    "  1. Generate visualization plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import all necessary libraries\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tkr\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os, glob\n",
    "from sklearn import preprocessing\n",
    "import statsmodels.api as sm\n",
    "import json\n",
    "from xlsxwriter.utility import xl_rowcol_to_cell\n",
    "import glob\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 1)** Load metadata about GRSs used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/apply_grs/pgs_reference_weights/\")\n",
    "\n",
    "# Load all key/values present on the header of PGS files into a dictionary\n",
    "pgs_metadata = {}\n",
    "for current_score_file in glob.glob( \"*.txt.gz\" ):\n",
    "    score_name = current_score_file.replace(\".txt.gz\",\"\")\n",
    "    pgs_metadata[score_name] = {}\n",
    "    with gzip.open( current_score_file,'rt') as f:\n",
    "        for line in f:\n",
    "            if( ( line.startswith(\"#\") ) & (not line.startswith(\"##\") ) ):\n",
    "                line = line.strip()\n",
    "                line = line.replace(\"#\", \"\")\n",
    "                key, value = line.split(\"=\")\n",
    "                pgs_metadata[score_name][key] = value\n",
    "    if( str( pgs_metadata[score_name][\"pgs_id\"] ).startswith(\"PGS\") ):\n",
    "        pgs_metadata[score_name][\"pgs_catalog_hyperlink\"] = \"https://www.pgscatalog.org/score/{0}/\".format(pgs_metadata[score_name][\"pgs_id\"])\n",
    "    else:\n",
    "        pgs_metadata[score_name][\"pgs_catalog_hyperlink\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 2)** Read phenotypic information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "race   sex   \n",
      "black  female     92\n",
      "       male      412\n",
      "white  female    128\n",
      "       male      436\n",
      "Name: sex, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>cf</th>\n",
       "      <th>cr</th>\n",
       "      <th>rltotal</th>\n",
       "      <th>rlmean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seaid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15.8</td>\n",
       "      <td>5.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12871</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>25</td>\n",
       "      <td>625</td>\n",
       "      <td>26</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12872</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>26</td>\n",
       "      <td>676</td>\n",
       "      <td>18.9</td>\n",
       "      <td>53.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12873</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>17</td>\n",
       "      <td>289</td>\n",
       "      <td>21.9</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12875</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>23</td>\n",
       "      <td>529</td>\n",
       "      <td>22.7</td>\n",
       "      <td>17.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12876</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>15</td>\n",
       "      <td>225</td>\n",
       "      <td>30.6</td>\n",
       "      <td>9.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1068 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sex   race  age  agex2   bmi    tf   tr    af    ar   cf   cr  \\\n",
       "seaid                                                                       \n",
       "pd10016  female  white   29    841  19.9   6.7  0.0  10.7   0.0  3.3  0.0   \n",
       "pd10018    male  white   30    900  28.3  27.3  0.0  22.0   0.0  3.4  0.6   \n",
       "pd10023    male  white   27    729  21.7  27.3  0.0  55.0   0.0  4.0  0.0   \n",
       "pd10028    male  black   31    961    29  16.0  0.0  36.8  14.8  1.7  1.0   \n",
       "pd10031    male  white   28    784  21.5  20.0  0.0  16.7   0.0  0.0  0.0   \n",
       "...         ...    ...  ...    ...   ...   ...  ...   ...   ...  ...  ...   \n",
       "pd12871    male  white   25    625    26  15.0  0.0   9.0   0.0  0.0  0.0   \n",
       "pd12872    male  black   26    676  18.9  53.3  0.0  55.0   0.0  0.0  0.0   \n",
       "pd12873    male  black   17    289  21.9  18.0  0.0  19.0   0.0  3.0  0.0   \n",
       "pd12875    male  black   23    529  22.7  17.7  0.0  32.3   0.0  2.3  0.0   \n",
       "pd12876  female  white   15    225  30.6   9.7  0.0   2.3   0.0  0.0  0.0   \n",
       "\n",
       "         rltotal    rlmean  \n",
       "seaid                       \n",
       "pd10016      0.0  0.000000  \n",
       "pd10018      0.6  0.200000  \n",
       "pd10023      0.0  0.000000  \n",
       "pd10028     15.8  5.266667  \n",
       "pd10031      0.0  0.000000  \n",
       "...          ...       ...  \n",
       "pd12871      0.0  0.000000  \n",
       "pd12872      0.0  0.000000  \n",
       "pd12873      0.0  0.000000  \n",
       "pd12875      0.0  0.000000  \n",
       "pd12876      0.0  0.000000  \n",
       "\n",
       "[1068 rows x 13 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/\")\n",
    "\n",
    "# Read phenotypes and recode sex and race attributes\n",
    "sea_phenotypes = pd.read_table(\n",
    "    \"data_preparation_to_imputation/86679/NHLBI/SEA_Herrington/phs000349v1/p1/phenotype/phs000349.v1.pht002191.v1.p1.c1.SEA_Phase2_Subject_Phenotypes.GRU.txt\",\n",
    "    index_col=\"seaid\",\n",
    "    comment=\"#\")\n",
    "\n",
    "# Recode Sex and Race\n",
    "sea_phenotypes = sea_phenotypes.replace( {\n",
    "    \"sex\": { 1:\"male\", 2:\"female\" },\n",
    "    \"race\": { 1:\"white\", 2:\"black\" }\n",
    "    } )\n",
    "\n",
    "# Drop unnecessary column\n",
    "sea_phenotypes = sea_phenotypes.drop(\"dbGaP SubjID\", axis = 1)\n",
    "\n",
    "# Print counts by race and sex and first lines from dataframe \n",
    "print( sea_phenotypes.groupby([\"race\",\"sex\"])[\"sex\"].count() )\n",
    "sea_phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 3) Load **PCA results** to dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>cf</th>\n",
       "      <th>...</th>\n",
       "      <th>PC11</th>\n",
       "      <th>PC12</th>\n",
       "      <th>PC13</th>\n",
       "      <th>PC14</th>\n",
       "      <th>PC15</th>\n",
       "      <th>PC16</th>\n",
       "      <th>PC17</th>\n",
       "      <th>PC18</th>\n",
       "      <th>PC19</th>\n",
       "      <th>PC20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.006825</td>\n",
       "      <td>0.018891</td>\n",
       "      <td>-0.003455</td>\n",
       "      <td>-0.007874</td>\n",
       "      <td>0.026135</td>\n",
       "      <td>0.019208</td>\n",
       "      <td>0.004665</td>\n",
       "      <td>0.008936</td>\n",
       "      <td>-0.040799</td>\n",
       "      <td>0.000123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.008184</td>\n",
       "      <td>0.002899</td>\n",
       "      <td>-0.001910</td>\n",
       "      <td>0.026848</td>\n",
       "      <td>0.019029</td>\n",
       "      <td>0.011589</td>\n",
       "      <td>-0.001970</td>\n",
       "      <td>-0.029758</td>\n",
       "      <td>-0.008889</td>\n",
       "      <td>0.037244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.021547</td>\n",
       "      <td>-0.017545</td>\n",
       "      <td>-0.011910</td>\n",
       "      <td>0.006369</td>\n",
       "      <td>0.021369</td>\n",
       "      <td>-0.028954</td>\n",
       "      <td>0.028462</td>\n",
       "      <td>0.003554</td>\n",
       "      <td>-0.003874</td>\n",
       "      <td>0.003383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.059962</td>\n",
       "      <td>0.025734</td>\n",
       "      <td>0.003848</td>\n",
       "      <td>-0.019902</td>\n",
       "      <td>-0.055411</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.046521</td>\n",
       "      <td>0.005009</td>\n",
       "      <td>-0.061828</td>\n",
       "      <td>-0.018428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003675</td>\n",
       "      <td>-0.002468</td>\n",
       "      <td>0.035139</td>\n",
       "      <td>-0.068820</td>\n",
       "      <td>0.086355</td>\n",
       "      <td>-0.012242</td>\n",
       "      <td>-0.072977</td>\n",
       "      <td>-0.018035</td>\n",
       "      <td>-0.090649</td>\n",
       "      <td>0.143556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12871</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>25</td>\n",
       "      <td>625</td>\n",
       "      <td>26</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.013646</td>\n",
       "      <td>-0.009487</td>\n",
       "      <td>0.000942</td>\n",
       "      <td>0.008001</td>\n",
       "      <td>0.004313</td>\n",
       "      <td>-0.014603</td>\n",
       "      <td>0.003558</td>\n",
       "      <td>0.020290</td>\n",
       "      <td>-0.006042</td>\n",
       "      <td>0.016711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12872</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>26</td>\n",
       "      <td>676</td>\n",
       "      <td>18.9</td>\n",
       "      <td>53.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.012626</td>\n",
       "      <td>0.019843</td>\n",
       "      <td>0.022616</td>\n",
       "      <td>0.054914</td>\n",
       "      <td>-0.068902</td>\n",
       "      <td>-0.012065</td>\n",
       "      <td>0.007603</td>\n",
       "      <td>-0.044239</td>\n",
       "      <td>-0.019055</td>\n",
       "      <td>-0.036954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12873</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>17</td>\n",
       "      <td>289</td>\n",
       "      <td>21.9</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014341</td>\n",
       "      <td>0.017928</td>\n",
       "      <td>0.073089</td>\n",
       "      <td>-0.008301</td>\n",
       "      <td>0.052788</td>\n",
       "      <td>0.054721</td>\n",
       "      <td>0.008493</td>\n",
       "      <td>-0.021289</td>\n",
       "      <td>-0.013251</td>\n",
       "      <td>-0.009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12875</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>23</td>\n",
       "      <td>529</td>\n",
       "      <td>22.7</td>\n",
       "      <td>17.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.057942</td>\n",
       "      <td>0.152120</td>\n",
       "      <td>-0.119545</td>\n",
       "      <td>0.073026</td>\n",
       "      <td>0.134121</td>\n",
       "      <td>0.099255</td>\n",
       "      <td>-0.018629</td>\n",
       "      <td>0.017820</td>\n",
       "      <td>-0.017232</td>\n",
       "      <td>0.075466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12876</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>15</td>\n",
       "      <td>225</td>\n",
       "      <td>30.6</td>\n",
       "      <td>9.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.032392</td>\n",
       "      <td>0.046445</td>\n",
       "      <td>-0.007091</td>\n",
       "      <td>-0.007472</td>\n",
       "      <td>0.000555</td>\n",
       "      <td>0.003258</td>\n",
       "      <td>0.013948</td>\n",
       "      <td>0.033616</td>\n",
       "      <td>0.037905</td>\n",
       "      <td>-0.030626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1068 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sex   race  age  agex2   bmi    tf   tr    af    ar   cf  ...  \\\n",
       "pd10016  female  white   29    841  19.9   6.7  0.0  10.7   0.0  3.3  ...   \n",
       "pd10018    male  white   30    900  28.3  27.3  0.0  22.0   0.0  3.4  ...   \n",
       "pd10023    male  white   27    729  21.7  27.3  0.0  55.0   0.0  4.0  ...   \n",
       "pd10028    male  black   31    961    29  16.0  0.0  36.8  14.8  1.7  ...   \n",
       "pd10031    male  white   28    784  21.5  20.0  0.0  16.7   0.0  0.0  ...   \n",
       "...         ...    ...  ...    ...   ...   ...  ...   ...   ...  ...  ...   \n",
       "pd12871    male  white   25    625    26  15.0  0.0   9.0   0.0  0.0  ...   \n",
       "pd12872    male  black   26    676  18.9  53.3  0.0  55.0   0.0  0.0  ...   \n",
       "pd12873    male  black   17    289  21.9  18.0  0.0  19.0   0.0  3.0  ...   \n",
       "pd12875    male  black   23    529  22.7  17.7  0.0  32.3   0.0  2.3  ...   \n",
       "pd12876  female  white   15    225  30.6   9.7  0.0   2.3   0.0  0.0  ...   \n",
       "\n",
       "             PC11      PC12      PC13      PC14      PC15      PC16      PC17  \\\n",
       "pd10016 -0.006825  0.018891 -0.003455 -0.007874  0.026135  0.019208  0.004665   \n",
       "pd10018 -0.008184  0.002899 -0.001910  0.026848  0.019029  0.011589 -0.001970   \n",
       "pd10023  0.021547 -0.017545 -0.011910  0.006369  0.021369 -0.028954  0.028462   \n",
       "pd10028 -0.059962  0.025734  0.003848 -0.019902 -0.055411  0.004292  0.046521   \n",
       "pd10031 -0.003675 -0.002468  0.035139 -0.068820  0.086355 -0.012242 -0.072977   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "pd12871 -0.013646 -0.009487  0.000942  0.008001  0.004313 -0.014603  0.003558   \n",
       "pd12872 -0.012626  0.019843  0.022616  0.054914 -0.068902 -0.012065  0.007603   \n",
       "pd12873  0.014341  0.017928  0.073089 -0.008301  0.052788  0.054721  0.008493   \n",
       "pd12875 -0.057942  0.152120 -0.119545  0.073026  0.134121  0.099255 -0.018629   \n",
       "pd12876 -0.032392  0.046445 -0.007091 -0.007472  0.000555  0.003258  0.013948   \n",
       "\n",
       "             PC18      PC19      PC20  \n",
       "pd10016  0.008936 -0.040799  0.000123  \n",
       "pd10018 -0.029758 -0.008889  0.037244  \n",
       "pd10023  0.003554 -0.003874  0.003383  \n",
       "pd10028  0.005009 -0.061828 -0.018428  \n",
       "pd10031 -0.018035 -0.090649  0.143556  \n",
       "...           ...       ...       ...  \n",
       "pd12871  0.020290 -0.006042  0.016711  \n",
       "pd12872 -0.044239 -0.019055 -0.036954  \n",
       "pd12873 -0.021289 -0.013251 -0.009285  \n",
       "pd12875  0.017820 -0.017232  0.075466  \n",
       "pd12876  0.033616  0.037905 -0.030626  \n",
       "\n",
       "[1068 rows x 33 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/pca_analysis/\")\n",
    "\n",
    "pca_results = {}\n",
    "\n",
    "pca_results[\"whites\"] = pd.read_table( \"pca_whites.eigenvec\", sep = \"\\t\" )\n",
    "pca_results[\"whites\"] = pca_results[\"whites\"].set_index(\"IID\")\n",
    "\n",
    "pca_results[\"blacks\"] = pd.read_table( \"pca_blacks.eigenvec\", sep = \"\\t\" )\n",
    "pca_results[\"blacks\"] = pca_results[\"blacks\"].set_index(\"IID\")\n",
    "\n",
    "pca_results[\"all\"] = pd.concat( [pca_results[\"whites\"], pca_results[\"blacks\"]], axis=0 ).drop(\"#FID\", axis = 1)\n",
    "\n",
    "sea_dataset_full = pd.concat( [sea_phenotypes, pca_results[\"all\"]], axis = 1 )\n",
    "sea_dataset_full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 4) Consolidate raw **PGS scores**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>cf</th>\n",
       "      <th>...</th>\n",
       "      <th>hrc_r03_PGS001105</th>\n",
       "      <th>hrc_r03_PGS000349</th>\n",
       "      <th>hrc_r03_PGS001917</th>\n",
       "      <th>hrc_r03_PGS001818</th>\n",
       "      <th>hrc_r03_TEMprsCatherine</th>\n",
       "      <th>hrc_r03_HDL20201014Shoa</th>\n",
       "      <th>hrc_r03_PGS000957</th>\n",
       "      <th>hrc_r03_logTG20201014Shoa</th>\n",
       "      <th>hrc_r03_PGS000889</th>\n",
       "      <th>hrc_r03_PGS000018</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3</td>\n",
       "      <td>...</td>\n",
       "      <td>3.841305</td>\n",
       "      <td>0.873616</td>\n",
       "      <td>0.005746</td>\n",
       "      <td>-0.101349</td>\n",
       "      <td>16.285275</td>\n",
       "      <td>0.168135</td>\n",
       "      <td>-0.201919</td>\n",
       "      <td>-2.372354</td>\n",
       "      <td>0.921467</td>\n",
       "      <td>0.711844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.416213</td>\n",
       "      <td>0.917900</td>\n",
       "      <td>-0.006679</td>\n",
       "      <td>0.706644</td>\n",
       "      <td>-4.972381</td>\n",
       "      <td>0.430336</td>\n",
       "      <td>0.826526</td>\n",
       "      <td>-0.133204</td>\n",
       "      <td>1.041347</td>\n",
       "      <td>1.533552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.079553</td>\n",
       "      <td>0.556460</td>\n",
       "      <td>0.011725</td>\n",
       "      <td>0.684267</td>\n",
       "      <td>18.696551</td>\n",
       "      <td>0.173749</td>\n",
       "      <td>0.751608</td>\n",
       "      <td>-0.524123</td>\n",
       "      <td>1.782580</td>\n",
       "      <td>0.996043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2.323256</td>\n",
       "      <td>0.562552</td>\n",
       "      <td>0.020173</td>\n",
       "      <td>0.003794</td>\n",
       "      <td>18.868259</td>\n",
       "      <td>-0.199715</td>\n",
       "      <td>0.788126</td>\n",
       "      <td>5.545249</td>\n",
       "      <td>0.709041</td>\n",
       "      <td>1.059956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.167806</td>\n",
       "      <td>0.691087</td>\n",
       "      <td>-0.010445</td>\n",
       "      <td>-0.145423</td>\n",
       "      <td>13.379956</td>\n",
       "      <td>0.244205</td>\n",
       "      <td>0.582374</td>\n",
       "      <td>0.165059</td>\n",
       "      <td>-0.173128</td>\n",
       "      <td>0.845935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12871</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>25</td>\n",
       "      <td>625</td>\n",
       "      <td>26</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.000878</td>\n",
       "      <td>0.350081</td>\n",
       "      <td>0.016940</td>\n",
       "      <td>-0.224118</td>\n",
       "      <td>25.951367</td>\n",
       "      <td>0.291442</td>\n",
       "      <td>0.519859</td>\n",
       "      <td>-1.189634</td>\n",
       "      <td>1.192794</td>\n",
       "      <td>0.805912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12872</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>26</td>\n",
       "      <td>676</td>\n",
       "      <td>18.9</td>\n",
       "      <td>53.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.970732</td>\n",
       "      <td>1.308783</td>\n",
       "      <td>0.027034</td>\n",
       "      <td>0.586234</td>\n",
       "      <td>31.842752</td>\n",
       "      <td>-0.102002</td>\n",
       "      <td>0.845895</td>\n",
       "      <td>6.761289</td>\n",
       "      <td>0.627752</td>\n",
       "      <td>1.013249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12873</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>17</td>\n",
       "      <td>289</td>\n",
       "      <td>21.9</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.690796</td>\n",
       "      <td>0.956890</td>\n",
       "      <td>0.024645</td>\n",
       "      <td>0.041306</td>\n",
       "      <td>16.919972</td>\n",
       "      <td>-0.009641</td>\n",
       "      <td>0.574422</td>\n",
       "      <td>5.649729</td>\n",
       "      <td>1.058504</td>\n",
       "      <td>1.243416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12875</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>23</td>\n",
       "      <td>529</td>\n",
       "      <td>22.7</td>\n",
       "      <td>17.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>...</td>\n",
       "      <td>2.639480</td>\n",
       "      <td>-0.035250</td>\n",
       "      <td>0.011434</td>\n",
       "      <td>-0.065502</td>\n",
       "      <td>38.277627</td>\n",
       "      <td>-0.139036</td>\n",
       "      <td>0.946164</td>\n",
       "      <td>3.113457</td>\n",
       "      <td>-0.735947</td>\n",
       "      <td>0.465990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12876</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>15</td>\n",
       "      <td>225</td>\n",
       "      <td>30.6</td>\n",
       "      <td>9.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.857105</td>\n",
       "      <td>0.185440</td>\n",
       "      <td>0.034240</td>\n",
       "      <td>-0.038732</td>\n",
       "      <td>36.688912</td>\n",
       "      <td>0.544742</td>\n",
       "      <td>0.868555</td>\n",
       "      <td>-0.544476</td>\n",
       "      <td>1.370564</td>\n",
       "      <td>0.659264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1068 rows × 289 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sex   race  age  agex2   bmi    tf   tr    af    ar   cf  ...  \\\n",
       "pd10016  female  white   29    841  19.9   6.7  0.0  10.7   0.0  3.3  ...   \n",
       "pd10018    male  white   30    900  28.3  27.3  0.0  22.0   0.0  3.4  ...   \n",
       "pd10023    male  white   27    729  21.7  27.3  0.0  55.0   0.0  4.0  ...   \n",
       "pd10028    male  black   31    961    29  16.0  0.0  36.8  14.8  1.7  ...   \n",
       "pd10031    male  white   28    784  21.5  20.0  0.0  16.7   0.0  0.0  ...   \n",
       "...         ...    ...  ...    ...   ...   ...  ...   ...   ...  ...  ...   \n",
       "pd12871    male  white   25    625    26  15.0  0.0   9.0   0.0  0.0  ...   \n",
       "pd12872    male  black   26    676  18.9  53.3  0.0  55.0   0.0  0.0  ...   \n",
       "pd12873    male  black   17    289  21.9  18.0  0.0  19.0   0.0  3.0  ...   \n",
       "pd12875    male  black   23    529  22.7  17.7  0.0  32.3   0.0  2.3  ...   \n",
       "pd12876  female  white   15    225  30.6   9.7  0.0   2.3   0.0  0.0  ...   \n",
       "\n",
       "         hrc_r03_PGS001105  hrc_r03_PGS000349  hrc_r03_PGS001917  \\\n",
       "pd10016           3.841305           0.873616           0.005746   \n",
       "pd10018          -0.416213           0.917900          -0.006679   \n",
       "pd10023           1.079553           0.556460           0.011725   \n",
       "pd10028           2.323256           0.562552           0.020173   \n",
       "pd10031           1.167806           0.691087          -0.010445   \n",
       "...                    ...                ...                ...   \n",
       "pd12871           3.000878           0.350081           0.016940   \n",
       "pd12872           1.970732           1.308783           0.027034   \n",
       "pd12873           2.690796           0.956890           0.024645   \n",
       "pd12875           2.639480          -0.035250           0.011434   \n",
       "pd12876           2.857105           0.185440           0.034240   \n",
       "\n",
       "         hrc_r03_PGS001818  hrc_r03_TEMprsCatherine  hrc_r03_HDL20201014Shoa  \\\n",
       "pd10016          -0.101349                16.285275                 0.168135   \n",
       "pd10018           0.706644                -4.972381                 0.430336   \n",
       "pd10023           0.684267                18.696551                 0.173749   \n",
       "pd10028           0.003794                18.868259                -0.199715   \n",
       "pd10031          -0.145423                13.379956                 0.244205   \n",
       "...                    ...                      ...                      ...   \n",
       "pd12871          -0.224118                25.951367                 0.291442   \n",
       "pd12872           0.586234                31.842752                -0.102002   \n",
       "pd12873           0.041306                16.919972                -0.009641   \n",
       "pd12875          -0.065502                38.277627                -0.139036   \n",
       "pd12876          -0.038732                36.688912                 0.544742   \n",
       "\n",
       "         hrc_r03_PGS000957  hrc_r03_logTG20201014Shoa  hrc_r03_PGS000889  \\\n",
       "pd10016          -0.201919                  -2.372354           0.921467   \n",
       "pd10018           0.826526                  -0.133204           1.041347   \n",
       "pd10023           0.751608                  -0.524123           1.782580   \n",
       "pd10028           0.788126                   5.545249           0.709041   \n",
       "pd10031           0.582374                   0.165059          -0.173128   \n",
       "...                    ...                        ...                ...   \n",
       "pd12871           0.519859                  -1.189634           1.192794   \n",
       "pd12872           0.845895                   6.761289           0.627752   \n",
       "pd12873           0.574422                   5.649729           1.058504   \n",
       "pd12875           0.946164                   3.113457          -0.735947   \n",
       "pd12876           0.868555                  -0.544476           1.370564   \n",
       "\n",
       "         hrc_r03_PGS000018  \n",
       "pd10016           0.711844  \n",
       "pd10018           1.533552  \n",
       "pd10023           0.996043  \n",
       "pd10028           1.059956  \n",
       "pd10031           0.845935  \n",
       "...                    ...  \n",
       "pd12871           0.805912  \n",
       "pd12872           1.013249  \n",
       "pd12873           1.243416  \n",
       "pd12875           0.465990  \n",
       "pd12876           0.659264  \n",
       "\n",
       "[1068 rows x 289 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read files with raw scores of multiple GRSs\n",
    "\n",
    "li = []\n",
    "li.append(sea_dataset_full)\n",
    "\n",
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/apply_grs/raw_scores_20220525_wrong_name/\")\n",
    "\n",
    "# Loop over info files and save info as a dictionary   \n",
    "scores_info_dict = {}\n",
    "for current_info_file in glob.glob( \"*.info.txt\" ):\n",
    "\n",
    "    # Parse filename to get info about GRSs.\n",
    "    # Format should be: <hrc|topmed>_<whites|blacks>_<8-digits_date>_multiGRS_<minr2_used>.info.txt\n",
    "    min_r2_used = os.path.basename( current_info_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "    reference_panel_name = os.path.basename( current_info_file ).split(\"_\")[0]\n",
    "    race = os.path.basename( current_info_file ).split(\"_\")[1]\n",
    "    \n",
    "    f = open( current_info_file )\n",
    "    data = json.load(f)\n",
    "\n",
    "    # Load metrics on a dictionary of dictionaries using panel, r2 and PGS name and each race as keys\n",
    "    for score_result in data:\n",
    "        key = \"_\".join( (reference_panel_name, min_r2_used, score_result[\"name\"]) )\n",
    "        \n",
    "        # If key doesn't exist yet in the dictionary, create it\n",
    "        if key not in scores_info_dict:\n",
    "            scores_info_dict[key] = {}\n",
    "            \n",
    "        scores_info_dict[key][race] = score_result\n",
    "\n",
    "    f.close()\n",
    "\n",
    "# Loop over scores files, load all into memory and save it to list object 'li'    \n",
    "for current_score_file in glob.glob( \"*.scores.txt\" ):\n",
    "        \n",
    "    grs_results = pd.read_table( current_score_file, sep = \",\" )\n",
    "    grs_results[\"sample\"] = grs_results[\"sample\"].str.split(\"_\", expand = True)[0]\n",
    "    grs_results = grs_results.set_index(\"sample\")\n",
    "    min_r2_used = os.path.basename( current_score_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "    reference_panel_name = os.path.basename( current_score_file ).split(\"_\")[0]\n",
    "        \n",
    "    # Add prefix to column names. Names should match pattern <hrc|topmed>_<threshold>_<scoreid>\n",
    "    grs_results = grs_results.add_prefix(reference_panel_name + \"_\" + min_r2_used + \"_\")\n",
    "\n",
    "    # Test if this set of scores already exists in li. \n",
    "    # If the other race was already loaded, append subjects to the dataframe. Otherwise, append dataframe to li\n",
    "    new_score_set = True\n",
    "    for i in range( len(li) ):\n",
    "        if( set( grs_results.columns ) == set( li[i].columns ) ):\n",
    "            li[i] = pd.concat( [li[i], grs_results], axis=0 )\n",
    "            new_score_set = False\n",
    "\n",
    "    if( new_score_set ):\n",
    "        li.append( grs_results )\n",
    "\n",
    "# Consolidate data into a dataframe and print first lines\n",
    "sea_dataset_full = pd.concat( li, axis=1 )\n",
    "sea_dataset_full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 5) Assign **case/control** to cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thresholds to be used for case/control definition:\n",
      " - white_male = 1.6000 \n",
      " - white_female = 0.4250 \n",
      " - black_male = 1.0000 \n",
      " - black_female = 0.0000 (WARNING: This group has Q3 equals zero. Considering only non-zeros as CASE group!!)\n",
      "race   sex     Case \n",
      "black  female  False     73\n",
      "               True      19\n",
      "       male    False    311\n",
      "               True     101\n",
      "white  female  False     96\n",
      "               True      32\n",
      "       male    False    328\n",
      "               True     108\n",
      "Name: Case, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>age</th>\n",
       "      <th>agex2</th>\n",
       "      <th>bmi</th>\n",
       "      <th>tf</th>\n",
       "      <th>tr</th>\n",
       "      <th>af</th>\n",
       "      <th>ar</th>\n",
       "      <th>cf</th>\n",
       "      <th>...</th>\n",
       "      <th>hrc_r03_PGS000349</th>\n",
       "      <th>hrc_r03_PGS001917</th>\n",
       "      <th>hrc_r03_PGS001818</th>\n",
       "      <th>hrc_r03_TEMprsCatherine</th>\n",
       "      <th>hrc_r03_HDL20201014Shoa</th>\n",
       "      <th>hrc_r03_PGS000957</th>\n",
       "      <th>hrc_r03_logTG20201014Shoa</th>\n",
       "      <th>hrc_r03_PGS000889</th>\n",
       "      <th>hrc_r03_PGS000018</th>\n",
       "      <th>Case</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pd10016</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>29</td>\n",
       "      <td>841</td>\n",
       "      <td>19.9</td>\n",
       "      <td>6.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.873616</td>\n",
       "      <td>0.005746</td>\n",
       "      <td>-0.101349</td>\n",
       "      <td>16.285275</td>\n",
       "      <td>0.168135</td>\n",
       "      <td>-0.201919</td>\n",
       "      <td>-2.372354</td>\n",
       "      <td>0.921467</td>\n",
       "      <td>0.711844</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10018</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>30</td>\n",
       "      <td>900</td>\n",
       "      <td>28.3</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.917900</td>\n",
       "      <td>-0.006679</td>\n",
       "      <td>0.706644</td>\n",
       "      <td>-4.972381</td>\n",
       "      <td>0.430336</td>\n",
       "      <td>0.826526</td>\n",
       "      <td>-0.133204</td>\n",
       "      <td>1.041347</td>\n",
       "      <td>1.533552</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10023</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>27</td>\n",
       "      <td>729</td>\n",
       "      <td>21.7</td>\n",
       "      <td>27.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.556460</td>\n",
       "      <td>0.011725</td>\n",
       "      <td>0.684267</td>\n",
       "      <td>18.696551</td>\n",
       "      <td>0.173749</td>\n",
       "      <td>0.751608</td>\n",
       "      <td>-0.524123</td>\n",
       "      <td>1.782580</td>\n",
       "      <td>0.996043</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10028</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>31</td>\n",
       "      <td>961</td>\n",
       "      <td>29</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>14.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>...</td>\n",
       "      <td>0.562552</td>\n",
       "      <td>0.020173</td>\n",
       "      <td>0.003794</td>\n",
       "      <td>18.868259</td>\n",
       "      <td>-0.199715</td>\n",
       "      <td>0.788126</td>\n",
       "      <td>5.545249</td>\n",
       "      <td>0.709041</td>\n",
       "      <td>1.059956</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd10031</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>28</td>\n",
       "      <td>784</td>\n",
       "      <td>21.5</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.691087</td>\n",
       "      <td>-0.010445</td>\n",
       "      <td>-0.145423</td>\n",
       "      <td>13.379956</td>\n",
       "      <td>0.244205</td>\n",
       "      <td>0.582374</td>\n",
       "      <td>0.165059</td>\n",
       "      <td>-0.173128</td>\n",
       "      <td>0.845935</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12871</th>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>25</td>\n",
       "      <td>625</td>\n",
       "      <td>26</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.350081</td>\n",
       "      <td>0.016940</td>\n",
       "      <td>-0.224118</td>\n",
       "      <td>25.951367</td>\n",
       "      <td>0.291442</td>\n",
       "      <td>0.519859</td>\n",
       "      <td>-1.189634</td>\n",
       "      <td>1.192794</td>\n",
       "      <td>0.805912</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12872</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>26</td>\n",
       "      <td>676</td>\n",
       "      <td>18.9</td>\n",
       "      <td>53.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.308783</td>\n",
       "      <td>0.027034</td>\n",
       "      <td>0.586234</td>\n",
       "      <td>31.842752</td>\n",
       "      <td>-0.102002</td>\n",
       "      <td>0.845895</td>\n",
       "      <td>6.761289</td>\n",
       "      <td>0.627752</td>\n",
       "      <td>1.013249</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12873</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>17</td>\n",
       "      <td>289</td>\n",
       "      <td>21.9</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.956890</td>\n",
       "      <td>0.024645</td>\n",
       "      <td>0.041306</td>\n",
       "      <td>16.919972</td>\n",
       "      <td>-0.009641</td>\n",
       "      <td>0.574422</td>\n",
       "      <td>5.649729</td>\n",
       "      <td>1.058504</td>\n",
       "      <td>1.243416</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12875</th>\n",
       "      <td>male</td>\n",
       "      <td>black</td>\n",
       "      <td>23</td>\n",
       "      <td>529</td>\n",
       "      <td>22.7</td>\n",
       "      <td>17.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.035250</td>\n",
       "      <td>0.011434</td>\n",
       "      <td>-0.065502</td>\n",
       "      <td>38.277627</td>\n",
       "      <td>-0.139036</td>\n",
       "      <td>0.946164</td>\n",
       "      <td>3.113457</td>\n",
       "      <td>-0.735947</td>\n",
       "      <td>0.465990</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd12876</th>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>15</td>\n",
       "      <td>225</td>\n",
       "      <td>30.6</td>\n",
       "      <td>9.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.185440</td>\n",
       "      <td>0.034240</td>\n",
       "      <td>-0.038732</td>\n",
       "      <td>36.688912</td>\n",
       "      <td>0.544742</td>\n",
       "      <td>0.868555</td>\n",
       "      <td>-0.544476</td>\n",
       "      <td>1.370564</td>\n",
       "      <td>0.659264</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1068 rows × 290 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sex   race  age  agex2   bmi    tf   tr    af    ar   cf  ...  \\\n",
       "pd10016  female  white   29    841  19.9   6.7  0.0  10.7   0.0  3.3  ...   \n",
       "pd10018    male  white   30    900  28.3  27.3  0.0  22.0   0.0  3.4  ...   \n",
       "pd10023    male  white   27    729  21.7  27.3  0.0  55.0   0.0  4.0  ...   \n",
       "pd10028    male  black   31    961    29  16.0  0.0  36.8  14.8  1.7  ...   \n",
       "pd10031    male  white   28    784  21.5  20.0  0.0  16.7   0.0  0.0  ...   \n",
       "...         ...    ...  ...    ...   ...   ...  ...   ...   ...  ...  ...   \n",
       "pd12871    male  white   25    625    26  15.0  0.0   9.0   0.0  0.0  ...   \n",
       "pd12872    male  black   26    676  18.9  53.3  0.0  55.0   0.0  0.0  ...   \n",
       "pd12873    male  black   17    289  21.9  18.0  0.0  19.0   0.0  3.0  ...   \n",
       "pd12875    male  black   23    529  22.7  17.7  0.0  32.3   0.0  2.3  ...   \n",
       "pd12876  female  white   15    225  30.6   9.7  0.0   2.3   0.0  0.0  ...   \n",
       "\n",
       "         hrc_r03_PGS000349  hrc_r03_PGS001917  hrc_r03_PGS001818  \\\n",
       "pd10016           0.873616           0.005746          -0.101349   \n",
       "pd10018           0.917900          -0.006679           0.706644   \n",
       "pd10023           0.556460           0.011725           0.684267   \n",
       "pd10028           0.562552           0.020173           0.003794   \n",
       "pd10031           0.691087          -0.010445          -0.145423   \n",
       "...                    ...                ...                ...   \n",
       "pd12871           0.350081           0.016940          -0.224118   \n",
       "pd12872           1.308783           0.027034           0.586234   \n",
       "pd12873           0.956890           0.024645           0.041306   \n",
       "pd12875          -0.035250           0.011434          -0.065502   \n",
       "pd12876           0.185440           0.034240          -0.038732   \n",
       "\n",
       "         hrc_r03_TEMprsCatherine  hrc_r03_HDL20201014Shoa  hrc_r03_PGS000957  \\\n",
       "pd10016                16.285275                 0.168135          -0.201919   \n",
       "pd10018                -4.972381                 0.430336           0.826526   \n",
       "pd10023                18.696551                 0.173749           0.751608   \n",
       "pd10028                18.868259                -0.199715           0.788126   \n",
       "pd10031                13.379956                 0.244205           0.582374   \n",
       "...                          ...                      ...                ...   \n",
       "pd12871                25.951367                 0.291442           0.519859   \n",
       "pd12872                31.842752                -0.102002           0.845895   \n",
       "pd12873                16.919972                -0.009641           0.574422   \n",
       "pd12875                38.277627                -0.139036           0.946164   \n",
       "pd12876                36.688912                 0.544742           0.868555   \n",
       "\n",
       "         hrc_r03_logTG20201014Shoa  hrc_r03_PGS000889  hrc_r03_PGS000018  \\\n",
       "pd10016                  -2.372354           0.921467           0.711844   \n",
       "pd10018                  -0.133204           1.041347           1.533552   \n",
       "pd10023                  -0.524123           1.782580           0.996043   \n",
       "pd10028                   5.545249           0.709041           1.059956   \n",
       "pd10031                   0.165059          -0.173128           0.845935   \n",
       "...                            ...                ...                ...   \n",
       "pd12871                  -1.189634           1.192794           0.805912   \n",
       "pd12872                   6.761289           0.627752           1.013249   \n",
       "pd12873                   5.649729           1.058504           1.243416   \n",
       "pd12875                   3.113457          -0.735947           0.465990   \n",
       "pd12876                  -0.544476           1.370564           0.659264   \n",
       "\n",
       "          Case  \n",
       "pd10016  False  \n",
       "pd10018  False  \n",
       "pd10023  False  \n",
       "pd10028  False  \n",
       "pd10031  False  \n",
       "...        ...  \n",
       "pd12871  False  \n",
       "pd12872  False  \n",
       "pd12873  False  \n",
       "pd12875  False  \n",
       "pd12876  False  \n",
       "\n",
       "[1068 rows x 290 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign case/control classes to samples using top quartile rule splitting by sex\n",
    "\n",
    "# Phenotype of interest\n",
    "phenotype = \"cr\"\n",
    "\n",
    "# Calculate thresholds for each race and sex\n",
    "print( \"Thresholds to be used for case/control definition:\" )\n",
    "thresholds = {}\n",
    "for race in [\"white\", \"black\"]:\n",
    "    for sex in [\"male\", \"female\"]:\n",
    "        \n",
    "        # Define Q3 as threshold and save it in thresholds dictionary\n",
    "        key = race + \"_\" + sex\n",
    "        thresholds[ key ] = sea_dataset_full[ (( sea_dataset_full[\"race\"]==race ) & ( sea_dataset_full[\"sex\"]==sex ))  ][phenotype].quantile(0.75)\n",
    "        \n",
    "        # Print values and warnings, if needed\n",
    "        note = \"\"\n",
    "        if( thresholds[ key ] == 0 ):\n",
    "            note = \"(WARNING: This group has Q3 equals zero. Considering only non-zeros as CASE group!!)\"\n",
    "        print( \" - {0} = {1:.4f} {2}\".format( key, thresholds[key], note ) )\n",
    "\n",
    "# Add a new column called Case with all values equals to False\n",
    "sea_dataset_full = sea_dataset_full.assign( Case=False )\n",
    "\n",
    "# Identify subjects above threshold to assign them to group \"Case\"\n",
    "for i in sea_dataset_full.index:\n",
    "\n",
    "    sex = sea_dataset_full.loc[ i, \"sex\"]\n",
    "    race = sea_dataset_full.loc[ i, \"race\"]\n",
    "    key = race + \"_\" + sex\n",
    "    \n",
    "    if( sea_dataset_full.loc[i, phenotype] > thresholds[ key ] ):\n",
    "        sea_dataset_full.loc[i, \"Case\"] = True\n",
    "\n",
    "# Print summary of Case/Controls for each subgroup\n",
    "print( sea_dataset_full.groupby([\"race\",\"sex\",\"Case\"])[\"Case\"].count() )\n",
    "sea_dataset_full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 6)** **Standardize** raw scores and calculate **Odds Ratios** between case/controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_mapped</th>\n",
       "      <th>model_ref</th>\n",
       "      <th>race</th>\n",
       "      <th>reference_panel</th>\n",
       "      <th>threshold</th>\n",
       "      <th>GRS</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>score_pvalue</th>\n",
       "      <th>conf_interval_lower</th>\n",
       "      <th>conf_interval_upper</th>\n",
       "      <th>...</th>\n",
       "      <th>pgs_name</th>\n",
       "      <th>trait_reported</th>\n",
       "      <th>trait_efo</th>\n",
       "      <th>weight_type</th>\n",
       "      <th>genome_build</th>\n",
       "      <th>variants_number</th>\n",
       "      <th>pgp_id</th>\n",
       "      <th>citation</th>\n",
       "      <th>pgs_catalog_hyperlink</th>\n",
       "      <th>license</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>black_topmed_r0_logTG20201014Shoa</th>\n",
       "      <td>triglyceride measurement</td>\n",
       "      <td>topmed_r0_logTG20201014Shoa</td>\n",
       "      <td>black</td>\n",
       "      <td>topmed</td>\n",
       "      <td>r0</td>\n",
       "      <td>logTG20201014Shoa</td>\n",
       "      <td>1.53941</td>\n",
       "      <td>0.021245</td>\n",
       "      <td>1.06647</td>\n",
       "      <td>2.22207</td>\n",
       "      <td>...</td>\n",
       "      <td>logTG20201014Shoa</td>\n",
       "      <td>Triglycerides</td>\n",
       "      <td>EFO_0004530</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>30071</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Ask Shoa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_topmed_r03_logTG20201014Shoa</th>\n",
       "      <td>triglyceride measurement</td>\n",
       "      <td>topmed_r03_logTG20201014Shoa</td>\n",
       "      <td>black</td>\n",
       "      <td>topmed</td>\n",
       "      <td>r03</td>\n",
       "      <td>logTG20201014Shoa</td>\n",
       "      <td>1.51839</td>\n",
       "      <td>0.0269178</td>\n",
       "      <td>1.04886</td>\n",
       "      <td>2.1981</td>\n",
       "      <td>...</td>\n",
       "      <td>logTG20201014Shoa</td>\n",
       "      <td>Triglycerides</td>\n",
       "      <td>EFO_0004530</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>30071</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Ask Shoa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>white_topmed_r0_PGS000018</th>\n",
       "      <td>coronary artery disease</td>\n",
       "      <td>topmed_r0_PGS000018</td>\n",
       "      <td>white</td>\n",
       "      <td>topmed</td>\n",
       "      <td>r0</td>\n",
       "      <td>PGS000018</td>\n",
       "      <td>1.48183</td>\n",
       "      <td>0.00113608</td>\n",
       "      <td>1.16933</td>\n",
       "      <td>1.87785</td>\n",
       "      <td>...</td>\n",
       "      <td>metaGRS_CAD</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>1745180</td>\n",
       "      <td>PGP000007</td>\n",
       "      <td>Inouye M et al. J Am Coll Cardiol (2018). doi:...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS000018/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>white_hrc_r0_TEMprsCatherine</th>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>hrc_r0_TEMprsCatherine</td>\n",
       "      <td>white</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r0</td>\n",
       "      <td>TEMprsCatherine</td>\n",
       "      <td>1.43058</td>\n",
       "      <td>0.0156741</td>\n",
       "      <td>1.06998</td>\n",
       "      <td>1.91271</td>\n",
       "      <td>...</td>\n",
       "      <td>TEMprsCatherine</td>\n",
       "      <td>Coronary artery disease</td>\n",
       "      <td>EFO_0001645</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>538096</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Ask Shoa or Catherine</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_topmed_r05_logTG20201014Shoa</th>\n",
       "      <td>triglyceride measurement</td>\n",
       "      <td>topmed_r05_logTG20201014Shoa</td>\n",
       "      <td>black</td>\n",
       "      <td>topmed</td>\n",
       "      <td>r05</td>\n",
       "      <td>logTG20201014Shoa</td>\n",
       "      <td>1.42706</td>\n",
       "      <td>0.0545909</td>\n",
       "      <td>0.993022</td>\n",
       "      <td>2.05081</td>\n",
       "      <td>...</td>\n",
       "      <td>logTG20201014Shoa</td>\n",
       "      <td>Triglycerides</td>\n",
       "      <td>EFO_0004530</td>\n",
       "      <td>NR</td>\n",
       "      <td>hg19</td>\n",
       "      <td>30071</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Ask Shoa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS001979</th>\n",
       "      <td>triglyceride measurement</td>\n",
       "      <td>hrc_r08_PGS001979</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS001979</td>\n",
       "      <td>0.834608</td>\n",
       "      <td>0.128296</td>\n",
       "      <td>0.661142</td>\n",
       "      <td>1.05359</td>\n",
       "      <td>...</td>\n",
       "      <td>portability-PLR_log_triglycerides</td>\n",
       "      <td>Triglycerides</td>\n",
       "      <td>EFO_0004530</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>71203</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS001979/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r05_PGS001357</th>\n",
       "      <td>type 2 diabetes mellitus</td>\n",
       "      <td>hrc_r05_PGS001357</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r05</td>\n",
       "      <td>PGS001357</td>\n",
       "      <td>0.834141</td>\n",
       "      <td>0.134904</td>\n",
       "      <td>0.657636</td>\n",
       "      <td>1.05802</td>\n",
       "      <td>...</td>\n",
       "      <td>T2D_AnnoPred_PRS</td>\n",
       "      <td>Type 2 diabetes</td>\n",
       "      <td>MONDO_0005148</td>\n",
       "      <td>beta</td>\n",
       "      <td>hg19</td>\n",
       "      <td>2996761</td>\n",
       "      <td>PGP000252</td>\n",
       "      <td>Ye Y et al. Circ Genom Precis Med (2021). doi:...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS001357/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS001105</th>\n",
       "      <td>body fat percentage</td>\n",
       "      <td>hrc_r08_PGS001105</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS001105</td>\n",
       "      <td>0.820428</td>\n",
       "      <td>0.101502</td>\n",
       "      <td>0.647382</td>\n",
       "      <td>1.03973</td>\n",
       "      <td>...</td>\n",
       "      <td>GBE_INI23127</td>\n",
       "      <td>Body fat percentage (trunk fat)</td>\n",
       "      <td>EFO_0007800</td>\n",
       "      <td>NR</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>25651</td>\n",
       "      <td>PGP000244</td>\n",
       "      <td>Tanigawa Y et al. PLoS Genet (2022). doi:10.13...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS001105/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS001900</th>\n",
       "      <td>diastolic blood pressure</td>\n",
       "      <td>hrc_r08_PGS001900</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS001900</td>\n",
       "      <td>0.820027</td>\n",
       "      <td>0.101062</td>\n",
       "      <td>0.646885</td>\n",
       "      <td>1.03951</td>\n",
       "      <td>...</td>\n",
       "      <td>portability-PLR_diastolic_BP</td>\n",
       "      <td>Diastolic blood pressure, automated reading</td>\n",
       "      <td>EFO_0006336</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>66335</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS001900/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black_hrc_r08_PGS002037</th>\n",
       "      <td>nicotine dependence</td>\n",
       "      <td>hrc_r08_PGS002037</td>\n",
       "      <td>black</td>\n",
       "      <td>hrc</td>\n",
       "      <td>r08</td>\n",
       "      <td>PGS002037</td>\n",
       "      <td>0.736007</td>\n",
       "      <td>0.0118908</td>\n",
       "      <td>0.579637</td>\n",
       "      <td>0.934559</td>\n",
       "      <td>...</td>\n",
       "      <td>portability-ldpred2_318</td>\n",
       "      <td>Tobacco use disorder</td>\n",
       "      <td>EFO_0003768</td>\n",
       "      <td>beta</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>847691</td>\n",
       "      <td>PGP000263</td>\n",
       "      <td>Privé F et al. Am J Hum Genet (2022). doi:10.1...</td>\n",
       "      <td>https://www.pgscatalog.org/score/PGS002037/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>512 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                trait_mapped  \\\n",
       "black_topmed_r0_logTG20201014Shoa   triglyceride measurement   \n",
       "black_topmed_r03_logTG20201014Shoa  triglyceride measurement   \n",
       "white_topmed_r0_PGS000018            coronary artery disease   \n",
       "white_hrc_r0_TEMprsCatherine         Coronary artery disease   \n",
       "black_topmed_r05_logTG20201014Shoa  triglyceride measurement   \n",
       "...                                                      ...   \n",
       "black_hrc_r08_PGS001979             triglyceride measurement   \n",
       "black_hrc_r05_PGS001357             type 2 diabetes mellitus   \n",
       "black_hrc_r08_PGS001105                  body fat percentage   \n",
       "black_hrc_r08_PGS001900             diastolic blood pressure   \n",
       "black_hrc_r08_PGS002037                  nicotine dependence   \n",
       "\n",
       "                                                       model_ref   race  \\\n",
       "black_topmed_r0_logTG20201014Shoa    topmed_r0_logTG20201014Shoa  black   \n",
       "black_topmed_r03_logTG20201014Shoa  topmed_r03_logTG20201014Shoa  black   \n",
       "white_topmed_r0_PGS000018                    topmed_r0_PGS000018  white   \n",
       "white_hrc_r0_TEMprsCatherine              hrc_r0_TEMprsCatherine  white   \n",
       "black_topmed_r05_logTG20201014Shoa  topmed_r05_logTG20201014Shoa  black   \n",
       "...                                                          ...    ...   \n",
       "black_hrc_r08_PGS001979                        hrc_r08_PGS001979  black   \n",
       "black_hrc_r05_PGS001357                        hrc_r05_PGS001357  black   \n",
       "black_hrc_r08_PGS001105                        hrc_r08_PGS001105  black   \n",
       "black_hrc_r08_PGS001900                        hrc_r08_PGS001900  black   \n",
       "black_hrc_r08_PGS002037                        hrc_r08_PGS002037  black   \n",
       "\n",
       "                                   reference_panel threshold  \\\n",
       "black_topmed_r0_logTG20201014Shoa           topmed        r0   \n",
       "black_topmed_r03_logTG20201014Shoa          topmed       r03   \n",
       "white_topmed_r0_PGS000018                   topmed        r0   \n",
       "white_hrc_r0_TEMprsCatherine                   hrc        r0   \n",
       "black_topmed_r05_logTG20201014Shoa          topmed       r05   \n",
       "...                                            ...       ...   \n",
       "black_hrc_r08_PGS001979                        hrc       r08   \n",
       "black_hrc_r05_PGS001357                        hrc       r05   \n",
       "black_hrc_r08_PGS001105                        hrc       r08   \n",
       "black_hrc_r08_PGS001900                        hrc       r08   \n",
       "black_hrc_r08_PGS002037                        hrc       r08   \n",
       "\n",
       "                                                  GRS odds_ratio score_pvalue  \\\n",
       "black_topmed_r0_logTG20201014Shoa   logTG20201014Shoa    1.53941     0.021245   \n",
       "black_topmed_r03_logTG20201014Shoa  logTG20201014Shoa    1.51839    0.0269178   \n",
       "white_topmed_r0_PGS000018                   PGS000018    1.48183   0.00113608   \n",
       "white_hrc_r0_TEMprsCatherine          TEMprsCatherine    1.43058    0.0156741   \n",
       "black_topmed_r05_logTG20201014Shoa  logTG20201014Shoa    1.42706    0.0545909   \n",
       "...                                               ...        ...          ...   \n",
       "black_hrc_r08_PGS001979                     PGS001979   0.834608     0.128296   \n",
       "black_hrc_r05_PGS001357                     PGS001357   0.834141     0.134904   \n",
       "black_hrc_r08_PGS001105                     PGS001105   0.820428     0.101502   \n",
       "black_hrc_r08_PGS001900                     PGS001900   0.820027     0.101062   \n",
       "black_hrc_r08_PGS002037                     PGS002037   0.736007    0.0118908   \n",
       "\n",
       "                                   conf_interval_lower conf_interval_upper  \\\n",
       "black_topmed_r0_logTG20201014Shoa              1.06647             2.22207   \n",
       "black_topmed_r03_logTG20201014Shoa             1.04886              2.1981   \n",
       "white_topmed_r0_PGS000018                      1.16933             1.87785   \n",
       "white_hrc_r0_TEMprsCatherine                   1.06998             1.91271   \n",
       "black_topmed_r05_logTG20201014Shoa            0.993022             2.05081   \n",
       "...                                                ...                 ...   \n",
       "black_hrc_r08_PGS001979                       0.661142             1.05359   \n",
       "black_hrc_r05_PGS001357                       0.657636             1.05802   \n",
       "black_hrc_r08_PGS001105                       0.647382             1.03973   \n",
       "black_hrc_r08_PGS001900                       0.646885             1.03951   \n",
       "black_hrc_r08_PGS002037                       0.579637            0.934559   \n",
       "\n",
       "                                    ...                           pgs_name  \\\n",
       "black_topmed_r0_logTG20201014Shoa   ...                  logTG20201014Shoa   \n",
       "black_topmed_r03_logTG20201014Shoa  ...                  logTG20201014Shoa   \n",
       "white_topmed_r0_PGS000018           ...                        metaGRS_CAD   \n",
       "white_hrc_r0_TEMprsCatherine        ...                    TEMprsCatherine   \n",
       "black_topmed_r05_logTG20201014Shoa  ...                  logTG20201014Shoa   \n",
       "...                                 ...                                ...   \n",
       "black_hrc_r08_PGS001979             ...  portability-PLR_log_triglycerides   \n",
       "black_hrc_r05_PGS001357             ...                   T2D_AnnoPred_PRS   \n",
       "black_hrc_r08_PGS001105             ...                       GBE_INI23127   \n",
       "black_hrc_r08_PGS001900             ...       portability-PLR_diastolic_BP   \n",
       "black_hrc_r08_PGS002037             ...            portability-ldpred2_318   \n",
       "\n",
       "                                                                 trait_reported  \\\n",
       "black_topmed_r0_logTG20201014Shoa                                 Triglycerides   \n",
       "black_topmed_r03_logTG20201014Shoa                                Triglycerides   \n",
       "white_topmed_r0_PGS000018                               Coronary artery disease   \n",
       "white_hrc_r0_TEMprsCatherine                            Coronary artery disease   \n",
       "black_topmed_r05_logTG20201014Shoa                                Triglycerides   \n",
       "...                                                                         ...   \n",
       "black_hrc_r08_PGS001979                                           Triglycerides   \n",
       "black_hrc_r05_PGS001357                                         Type 2 diabetes   \n",
       "black_hrc_r08_PGS001105                         Body fat percentage (trunk fat)   \n",
       "black_hrc_r08_PGS001900             Diastolic blood pressure, automated reading   \n",
       "black_hrc_r08_PGS002037                                    Tobacco use disorder   \n",
       "\n",
       "                                        trait_efo weight_type genome_build  \\\n",
       "black_topmed_r0_logTG20201014Shoa     EFO_0004530          NR         hg19   \n",
       "black_topmed_r03_logTG20201014Shoa    EFO_0004530          NR         hg19   \n",
       "white_topmed_r0_PGS000018             EFO_0001645          NR         hg19   \n",
       "white_hrc_r0_TEMprsCatherine          EFO_0001645          NR         hg19   \n",
       "black_topmed_r05_logTG20201014Shoa    EFO_0004530          NR         hg19   \n",
       "...                                           ...         ...          ...   \n",
       "black_hrc_r08_PGS001979               EFO_0004530        beta       GRCh37   \n",
       "black_hrc_r05_PGS001357             MONDO_0005148        beta         hg19   \n",
       "black_hrc_r08_PGS001105               EFO_0007800          NR       GRCh37   \n",
       "black_hrc_r08_PGS001900               EFO_0006336        beta       GRCh37   \n",
       "black_hrc_r08_PGS002037               EFO_0003768        beta       GRCh37   \n",
       "\n",
       "                                   variants_number     pgp_id  \\\n",
       "black_topmed_r0_logTG20201014Shoa            30071        N/A   \n",
       "black_topmed_r03_logTG20201014Shoa           30071        N/A   \n",
       "white_topmed_r0_PGS000018                  1745180  PGP000007   \n",
       "white_hrc_r0_TEMprsCatherine                538096        N/A   \n",
       "black_topmed_r05_logTG20201014Shoa           30071        N/A   \n",
       "...                                            ...        ...   \n",
       "black_hrc_r08_PGS001979                      71203  PGP000263   \n",
       "black_hrc_r05_PGS001357                    2996761  PGP000252   \n",
       "black_hrc_r08_PGS001105                      25651  PGP000244   \n",
       "black_hrc_r08_PGS001900                      66335  PGP000263   \n",
       "black_hrc_r08_PGS002037                     847691  PGP000263   \n",
       "\n",
       "                                                                             citation  \\\n",
       "black_topmed_r0_logTG20201014Shoa                                            Ask Shoa   \n",
       "black_topmed_r03_logTG20201014Shoa                                           Ask Shoa   \n",
       "white_topmed_r0_PGS000018           Inouye M et al. J Am Coll Cardiol (2018). doi:...   \n",
       "white_hrc_r0_TEMprsCatherine                                    Ask Shoa or Catherine   \n",
       "black_topmed_r05_logTG20201014Shoa                                           Ask Shoa   \n",
       "...                                                                               ...   \n",
       "black_hrc_r08_PGS001979             Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "black_hrc_r05_PGS001357             Ye Y et al. Circ Genom Precis Med (2021). doi:...   \n",
       "black_hrc_r08_PGS001105             Tanigawa Y et al. PLoS Genet (2022). doi:10.13...   \n",
       "black_hrc_r08_PGS001900             Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "black_hrc_r08_PGS002037             Privé F et al. Am J Hum Genet (2022). doi:10.1...   \n",
       "\n",
       "                                                          pgs_catalog_hyperlink  \\\n",
       "black_topmed_r0_logTG20201014Shoa                                                 \n",
       "black_topmed_r03_logTG20201014Shoa                                                \n",
       "white_topmed_r0_PGS000018           https://www.pgscatalog.org/score/PGS000018/   \n",
       "white_hrc_r0_TEMprsCatherine                                                      \n",
       "black_topmed_r05_logTG20201014Shoa                                                \n",
       "...                                                                         ...   \n",
       "black_hrc_r08_PGS001979             https://www.pgscatalog.org/score/PGS001979/   \n",
       "black_hrc_r05_PGS001357             https://www.pgscatalog.org/score/PGS001357/   \n",
       "black_hrc_r08_PGS001105             https://www.pgscatalog.org/score/PGS001105/   \n",
       "black_hrc_r08_PGS001900             https://www.pgscatalog.org/score/PGS001900/   \n",
       "black_hrc_r08_PGS002037             https://www.pgscatalog.org/score/PGS002037/   \n",
       "\n",
       "                                   license  \n",
       "black_topmed_r0_logTG20201014Shoa      NaN  \n",
       "black_topmed_r03_logTG20201014Shoa     NaN  \n",
       "white_topmed_r0_PGS000018              NaN  \n",
       "white_hrc_r0_TEMprsCatherine           NaN  \n",
       "black_topmed_r05_logTG20201014Shoa     NaN  \n",
       "...                                    ...  \n",
       "black_hrc_r08_PGS001979                NaN  \n",
       "black_hrc_r05_PGS001357                NaN  \n",
       "black_hrc_r08_PGS001105                NaN  \n",
       "black_hrc_r08_PGS001900                NaN  \n",
       "black_hrc_r08_PGS002037                NaN  \n",
       "\n",
       "[512 rows x 34 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize raw scores and calculate OR\n",
    "\n",
    "li = []\n",
    "standardize_scores = []\n",
    "logit_models_dict = {}\n",
    "models_summaries = pd.DataFrame()\n",
    "\n",
    "# Get list of panels, thresholds and GRSs used from info dictonary\n",
    "reference_panels = set( [ key.split('_')[0] for key in scores_info_dict.keys() ] )\n",
    "thresholds = set( [ key.split('_')[1] for key in scores_info_dict.keys() ] )\n",
    "grss = set( [ key.split('_')[2] for key in scores_info_dict.keys() ] )\n",
    "\n",
    "# Normalize all scores with mean = 0 and SD = 1\n",
    "for race in [\"white\", \"black\"]:\n",
    "    \n",
    "    # Subset cohort between whites and blacks to run logistic regression individually\n",
    "    sea_dataset_subset = sea_dataset_full[ sea_dataset_full[\"race\"] == race ]\n",
    "\n",
    "    # Create a dependent variable named \"Case_recoded\" based on column \"Case\", conding it as 0 and 1 to fit glm\n",
    "    sea_dataset_subset = sea_dataset_subset.assign( Case_recoded=sea_dataset_subset[\"Case\"].replace(True, 1).replace(False, 0) )\n",
    "    \n",
    "    for reference_panel_id in reference_panels:\n",
    "        \n",
    "        for threshold in thresholds:\n",
    "\n",
    "            for grs in grss:\n",
    "\n",
    "                raw_score_name = \"_\".join( (reference_panel_id, threshold, grs) )\n",
    "                standardize_score_name = \"_\".join( (race, reference_panel_id, threshold, grs) )\n",
    "                standardize_scores.append(standardize_score_name)\n",
    "        \n",
    "                # Standardize raw scores using method scale\n",
    "                sea_dataset_subset[ standardize_score_name ] = preprocessing.scale( sea_dataset_subset[ raw_score_name ] )\n",
    "        \n",
    "                # Fit a logistic model using standardize scores and save it on dictionary \n",
    "                logit_models_dict[ standardize_score_name ] = sm.formula.glm(\n",
    "                    \"Case_recoded ~ \" + standardize_score_name + \" + age + sex + PC1 + PC2 + PC3 + PC4 + PC5\",\n",
    "                    family=sm.families.Binomial(),\n",
    "                    data=sea_dataset_subset).fit()\n",
    "\n",
    "                # Create dictonary with scores performances\n",
    "                model_summary_series = {\n",
    "                        \"model_ref\": raw_score_name,\n",
    "                        \"race\": race,\n",
    "                        \"reference_panel\": reference_panel_id,\n",
    "                        \"threshold\": threshold,\n",
    "                        \"GRS\": grs,\n",
    "                        \"odds_ratio\": np.exp( logit_models_dict[ standardize_score_name ].params )[ standardize_score_name ],\n",
    "                        \"score_pvalue\": logit_models_dict[ standardize_score_name ].pvalues[ standardize_score_name ],\n",
    "                        \"conf_interval_lower\": np.exp( logit_models_dict[ standardize_score_name ].conf_int()[0][ standardize_score_name ] ),\n",
    "                        \"conf_interval_upper\": np.exp( logit_models_dict[ standardize_score_name ].conf_int()[1][ standardize_score_name ] )\n",
    "                    }\n",
    "                # Add additional info from info files (coverage, etc)\n",
    "                model_summary_series.update( scores_info_dict[raw_score_name][race + \"s\"] )\n",
    "                \n",
    "                # Add metadata about the GRSs to the output\n",
    "                model_summary_series.update( pgs_metadata[grs] )\n",
    "                \n",
    "                # Transform dictionary on a pandas Series and name it as <standardize_score_name>\n",
    "                model_summary_series = pd.Series(\n",
    "                    data=model_summary_series,\n",
    "                    name=standardize_score_name\n",
    "                )\n",
    "                                \n",
    "                li.append(model_summary_series)\n",
    "                     \n",
    "# Concatenate results for all models tested and transpose object to make visualization easier \n",
    "models_summaries = pd.concat( li, axis=1 ).transpose().sort_values(\"odds_ratio\", ascending=False)\n",
    "\n",
    "# Move trait_mapped column to first position to facilitate analysis of excel results\n",
    "columns_order = models_summaries.columns.drop(\"trait_mapped\").tolist()\n",
    "columns_order.insert(0,\"trait_mapped\")\n",
    "models_summaries = models_summaries.reindex(columns=columns_order, copy=False)\n",
    "models_summaries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Step 7)** Export dataframe with summaries as a **xls** file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export dataframe with summaries on a xls file on the exports folder\n",
    "\n",
    "writer = pd.ExcelWriter(\n",
    "    path = \"/labs/tassimes/rodrigoguarischi/projects/sea/exports/SEA_models_summary.xls\",\n",
    "    engine = 'xlsxwriter'\n",
    "    )\n",
    "\n",
    "models_summaries.to_excel(writer, sheet_name='models_summary')\n",
    "\n",
    "# Get access to the workbook and sheet\n",
    "workbook = writer.book\n",
    "worksheet = writer.sheets['models_summary']\n",
    "\n",
    "# Automatically add filters to columns\n",
    "worksheet.autofilter( \"A1:AH{0}\".format(len( models_summaries.index ) + 1) )\n",
    "\n",
    "# Add a percent format with 2 decimal points\n",
    "percent_fmt = workbook.add_format({'num_format': '0.00%', 'align': 'left'})\n",
    "\n",
    "# Add a left-alignment format\n",
    "left_alignment_fmt = workbook.add_format({'align': 'left'})\n",
    "\n",
    "# Format the columns by width and include number formats\n",
    "default_column_width = 11\n",
    "worksheet.set_column('A:A', 35, left_alignment_fmt)                         # Model full name\n",
    "worksheet.set_column('B:B', 45, left_alignment_fmt)                         # Trait mapped\n",
    "worksheet.set_column('C:C', 30, left_alignment_fmt)                         # Model ref\n",
    "worksheet.set_column('D:O', default_column_width, left_alignment_fmt)     \n",
    "worksheet.set_column('P:P', default_column_width, percent_fmt)              # Coverage\n",
    "worksheet.set_column('Q:AF', default_column_width, left_alignment_fmt)\n",
    "worksheet.set_column('AG:AG', 70, left_alignment_fmt)                       # Citation\n",
    "worksheet.set_column('AH:AH', 40, left_alignment_fmt)                       # PGS catalog link\n",
    "worksheet.set_column('AI:AI', default_column_width, left_alignment_fmt)     # License\n",
    "\n",
    "# Transform cells on column AH on hyperlinks\n",
    "for model_index in range(0, len(models_summaries)):\n",
    "\n",
    "    # Skip one line for the header +1 because excel location is 1-based\n",
    "    row_number = model_index + 2\n",
    "    model_index_name = models_summaries.index[model_index]\n",
    "    \n",
    "    # Get text to transform it to URL\n",
    "    hyperlink = models_summaries.loc[model_index_name,\"pgs_catalog_hyperlink\"]\n",
    "    \n",
    "    # write text as hyperlink\n",
    "    worksheet.write_url( \"AH\" + str(row_number), hyperlink)\n",
    "\n",
    "## Add colors to cells of coverage labels to easy visualization\n",
    "\n",
    "# Create a red, yellow and green fills with dark text\n",
    "red_format = workbook.add_format({'bg_color':   '#FFC7CE', 'font_color': '#9C0006'})\n",
    "yellow_format = workbook.add_format({'bg_color':   '#FFEB9C', 'font_color': '#9C6500'})\n",
    "green_format = workbook.add_format({'bg_color':   '#C6EFCE', 'font_color': '#006100'})\n",
    "\n",
    "# Get location of coverage label cells\n",
    "cells_location = 'W1:W' + str( len(models_summaries) + 1)\n",
    "\n",
    "worksheet.conditional_format( cells_location, {'type': 'cell', 'criteria': 'equal to', 'value': '\"high\"', 'format': green_format} )\n",
    "worksheet.conditional_format( cells_location, {'type': 'cell', 'criteria': 'equal to', 'value': '\"medium\"', 'format': yellow_format} )\n",
    "worksheet.conditional_format( cells_location, {'type': 'cell', 'criteria': 'equal to', 'value': '\"low\"', 'format': red_format} )\n",
    "worksheet.conditional_format( cells_location, {'type': 'cell', 'criteria': 'equal to', 'value': '\"zero\"', 'format': red_format} )\n",
    "\n",
    "# Save modifications\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export dataframe with summaries on a xls file on the exports folder\n",
    "\n",
    "models_summaries.to_excel( \n",
    "  \"/labs/tassimes/rodrigoguarischi/projects/sea/exports/SEA_models_summary.xls\",\n",
    "  sheet_name=\"models_summary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_name = \"White_hrc_r03_wGRS49\"\n",
    "print( logit_models_dict[model_name].summary() )\n",
    "# print( logit_models_dict[model_name].pvalues[model_name] )\n",
    "# print( np.exp( logit_models_dict[model_name].params )[ model_name ] )\n",
    "# print( np.exp( logit_models_dict[model_name].conf_int().loc[[ model_name ]] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot boxplots and histograms and boxplots of raw scores spliting by sex\n",
    "\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "score_list = [\"wGRS49_r0\", \"wGRS49_r03\", \"wGRS49_r05\", \"wGRS49_r08\"]\n",
    "# score_list = [\"PGS000349_r0\", \"PGS000349_r03\", \"PGS000349_r05\", \"PGS000349_r08\"]\n",
    "# score_list = [\"PGS000018_r0\", \"PGS000018_r03\", \"PGS000018_r05\", \"PGS000018_r08\"]\n",
    "\n",
    "fig, axs = plt.subplots(nrows=3, ncols=4, figsize=(5*len(score_list),12))\n",
    "for i in range(0, len(score_list)):\n",
    "    sns.boxplot(\n",
    "        x = \"Case\",\n",
    "        y = score_list[i],\n",
    "        data = sea_merged_whites,\n",
    "        palette = reversed(sns.color_palette(n_colors=2)),\n",
    "        width=0.4,\n",
    "        fliersize=2,\n",
    "        ax=axs[0,i],\n",
    "    )\n",
    "    sns.histplot(\n",
    "        x = score_list[i],\n",
    "        data = sea_merged_whites[ sea_merged_whites[\"sex\"] == \"Male\" ],\n",
    "        hue=\"Case\",\n",
    "        hue_order=[True, False],\n",
    "        kde=True,\n",
    "        ax=axs[1,i]\n",
    "        ).set(title='Males (n={0})'.format( sum( sea_merged_whites[\"sex\"] == \"Male\" ) ) )\n",
    "    sns.histplot(\n",
    "        x = score_list[i],\n",
    "        data = sea_merged_whites[ sea_merged_whites[\"sex\"] == \"Female\" ],\n",
    "        hue=\"Case\",\n",
    "        hue_order=[True, False],\n",
    "        kde=True,\n",
    "        ax=axs[2,i]\n",
    "    ).set(title='Females (n={0})'.format( sum( sea_merged_whites[\"sex\"] == \"Female\" ) ) )\n",
    "\n",
    "plt.subplots_adjust(top=1.25)\n",
    "\n",
    "# # Plot only boxplots\n",
    "\n",
    "# fig, axs = plt.subplots(ncols=len(score_list), figsize=(5*len(score_list),4))\n",
    "# for i in range(0, len(score_list)):\n",
    "#     sns.boxplot(\n",
    "#         x = \"Case\",\n",
    "#         y = score_list[i],\n",
    "#         data = sea_merged_whites,\n",
    "#         hue=\"sex\",\n",
    "#         hue_order=['Male','Female'],\n",
    "#         width=0.4,\n",
    "#         fliersize=2,\n",
    "#         ax=axs[i],\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test variables\n",
    "sns.boxplot(\n",
    "    x = \"Case\",\n",
    "    y = \"age\",\n",
    "    data = sea_merged_whites,\n",
    "    width=0.4,\n",
    "    fliersize=2\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export dataset to CSV files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export full dataset\n",
    "sea_merged_whites.to_csv(\"exports/sea_whites_phenotypes_scores_full.csv\")\n",
    "\n",
    "# Export only subset of columns Tim asked\n",
    "select_columns = [\"dbGaP_SubjID\",\"sex\",\"race\",\"age\",\"agex2\",\"bmi\",\"cr\",\"wGRS49_r05\",\"PGS000349_r05\",\"PGS000018_r05\",\"wGRS49_r03\",\"PGS000349_r03\",\"PGS000018_r03\",\"wGRS49_r0\",\"PGS000349_r0\",\"PGS000018_r0\",\"wGRS49_r08\",\"PGS000349_r08\",\"PGS000018_r08\", \"Case\"]\n",
    "sea_merged_whites[select_columns].to_csv(\"exports/sea_whites_phenotypes_scores_column_subset.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "sns.scatterplot(\n",
    "    x = \"wGRS49_r03_scaled\",\n",
    "    y = \"Case\",\n",
    "    data = sea_merged_whites\n",
    "    # ,\n",
    "    # alpha=0.4,\n",
    "    # s=6\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test logistic regression in <code>R</code>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Read CSV file with raw scores\n",
    "input_table <- read.csv(\"exports/sea_whites_phenotypes_scores_full.csv\", row.names=1)\n",
    "\n",
    "# Normalize all scores with mean = 0 and SD = 1\n",
    "for ( grs in c(\"wGRS49\",\"PGS000349\",\"PGS000018\", \"PGS000667\", \"PGS000889\") ){\n",
    "    for ( threshold in c(\"r0\", \"r03\", \"r05\", \"r08\") ){\n",
    "        current_score <- paste0( grs, \"_\", threshold);\n",
    "        score_values <- input_table[[ current_score ]]\n",
    "        score_mean <- mean( score_values )\n",
    "        score_sd <- sd( score_values )\n",
    "        input_table[ paste0(current_score, \"_scaled\") ] <- (score_values-score_mean)/score_sd\n",
    "        # input_table[ paste0(current_score, \"_scaled\") ] <- scale( score_values )\n",
    "    }\n",
    "}\n",
    "\n",
    "# Test Standardization (check values of min, mean, max and sd)\n",
    "normalization_summary <- function(current_scaled_score){\n",
    "  return (\n",
    "    list(\n",
    "      \"min\"=min(current_scaled_score),\n",
    "      \"mean\"=mean(current_scaled_score),\n",
    "      \"max\"=max(current_scaled_score),\n",
    "      \"sd\"=sd(current_scaled_score)\n",
    "    )\n",
    "  )\n",
    "}\n",
    "t( sapply( input_table[ , grepl(\"_scaled\", names(input_table)) ], normalization_summary ) )\n",
    "\n",
    "# Test number of case/controls for each sex and age\n",
    "xtabs( ~ Case + sex, data=input_table)\n",
    "xtabs( ~ Case + age, data=input_table)\n",
    "\n",
    "# Apply logistic regression to scaled scores\n",
    "models <- list()\n",
    "for ( score_scaled_name in grep(\"_scaled\", names(input_table), value = T) ){\n",
    "  \n",
    "  # Skip PGSs with 0 SNPs evaluated (scores equals to NaN)  \n",
    "  if( is.na( median( input_table[, score_scaled_name ] ) ) ) next;\n",
    "  \n",
    "  input_table_subset <- input_table[, c(score_scaled_name, \"age\", \"sex\", \"Case\") ]\n",
    "  colnames( input_table_subset ) <- c(\"score_scaled\", \"age\", \"sex\", \"Case\")\n",
    "  models[[ score_scaled_name ]] <- glm( Case ~ score_scaled + age + sex, data = input_table_subset, family = \"binomial\")\n",
    "}\n",
    "\n",
    "# Print Odds Ratios for each score at 4 levels of imputation Rsqrt thresholding (0, 0.3, 0.5 and 0.8)\n",
    "t( sapply( models, function(current_model) exp( current_model$coefficients ) ) )\n",
    "\n",
    "# Get p-values\n",
    "t( sapply( models, function(current_model) coef(summary(current_model))[,4] ) )\n",
    "\n",
    "summary( models$wGRS49_r0_scaled )\n",
    "\n",
    "---\n",
    "apply_logit <- function(current_scaled_score, input_table) {\n",
    "    logit_formula <- as.formula( paste0(\"Case ~ \" , current_scaled_score ,\" + age + sex\") )\n",
    "    logistic_model <- glm( logit_formula, data = input_table, family = \"binomial\")\n",
    "    return( logistic_model )\n",
    "} \n",
    "scaled_scores <- grep(\"_scaled\", names(input_table), value = T)\n",
    "models <- lapply( scaled_scores, function(current_scaled_score) apply_logit(current_scaled_score, input_table) )\n",
    "names(models) <- scaled_scores\n",
    "\n",
    "----\n",
    "\n",
    "summary( models$wGRS49_r03_scaled )\n",
    "str( models$wGRS49_r03_scaled )\n",
    "\n",
    "table( ifelse( models$wGRS49_r03_scaled$fitted.values > 0.5, \"assigned_case\", \"assigned_control\") )\n",
    "\n",
    "summary( models[[2]] )\n",
    "\n",
    "create_pdf_plot <- function(model, input_data) {\n",
    "\n",
    "    library(ggplot2)\n",
    "    library(cowplot)\n",
    "\n",
    "    predicted_data <- data.frame(\n",
    "        probability_of_hd = model$fitted.values,\n",
    "        cr = input_table$Case\n",
    "        )\n",
    "\n",
    "    predicted_data <- predicted_data[ order(predicted_data$probability_of_hd, decreasing=FALSE), ]\n",
    "    predicted_data$rank <- 1:nrow(predicted_data)\n",
    "\n",
    "    ggplot( data=predicted_data, aes(x=rank, y=probability_of_hd) ) +\n",
    "        geom_point(aes(color=cr), alpha=1, shape=4, stroke=2) +\n",
    "        xlab(\"Index\") +\n",
    "        ylab(\"Predicted probability of getting heart disease\")\n",
    "    ggsave(\"heart_disease_probabilities.pdf\")\n",
    "}\n",
    "\n",
    "create_pdf_plot(models[[1]], input_table)\n",
    "\n",
    "-----\n",
    "\n",
    "# Test model based on multiple scores\n",
    "logistic_model <- glm( Case ~ ., data = input_table[,c(scaled_scores, \"age\", \"sex\", \"Case\")], family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "-----\n",
    "\n",
    "summary( models[, \"wGRS49_r0_scaled\"] )\n",
    "\n",
    "t( sapply( scaled_scores, function(x) apply_logit(x, input_table) ) )\n",
    "\n",
    "t( sapply( grep(\"_scaled\", names(input_table), value = T), function(x) apply_logit(x, input_table) ) )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "sapply( grep(\"_scaled\", names(input_table), value = T), function(x) logit_formula <- print0(\"Case ~ \" , x ,\" + age + sex\"; print(logit_formula) )\n",
    "\n",
    "sapply( 1:5, function(x) x*2)\n",
    "\n",
    "logistic_model <- glm(Case ~ wGRS49_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "logistic_model <- glm(Case ~ PGS000349_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "logistic_model <- glm(Case ~ PGS000018_r03_scaled + age + sex, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "exp( logistic_model$coefficients )\n",
    "\n",
    "input_table_subset <- input_table[ , c( grep(\"_scaled\", names(input_table), value = T), \"age\", \"sex\", \"Case\") ]\n",
    "\n",
    "logistic_model <- glm(Case~., data = input_table_subset, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "// input_table[ paste0(current_score, \"_scaled\") ] <- scale( input_table[[ current_score ]], center = 0, scale = 1)\n",
    "\n",
    "range( input_table$PGS000018_r03_scaled )\n",
    "input_table$wGRS49_r03_scaled <- scale(input_table$wGRS49_r03, center = 0, scale = 1)\n",
    "input_table$PGS000018_r03_scaled <- scale(input_table$PGS000018_r03, center = 0, scale = 1)\n",
    "\n",
    "logistic_model <- glm(Case ~ wGRS49_r03_scaled, data = input_table, family = \"binomial\")\n",
    "summary(logistic_model)\n",
    "\n",
    "# Predict test data based on model\n",
    "predict_reg <- predict(logistic_model, input_table, type = \"response\")\n",
    "range(predict_reg)\n",
    "predict_reg <- ifelse( predict_reg > 0.5, 1, 0)\n",
    "\n",
    "range(predict_reg)\n",
    "logistic_model$\n",
    "\n",
    "predict_reg <- <- predict(logistic_model, logistic_model, type = \"response\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# table = df.groupby(level=\"Cancer\").sum().values\n",
    "\n",
    "table = [[840,51663],[32,5053]]\n",
    "print(table)\n",
    "oddsratio, pvalue = stats.fisher_exact(table)\n",
    "print(\"OddsR: \", oddsratio, \"p-Value:\", pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sea_merged_whites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "for score in [\"wGRS49_r03\", \"wGRS49_r05\", \"wGRS49_r08\", \"PGS000018_r03\", \"PGS000018_r05\", \"PGS000018_r08\"]:\n",
    "    for traits in [\"tf_quartile4\", \"tr_quartile4\", \"af_quartile4\", \"ar_quartile4\", \"cf_quartile4\", \"cr_quartile4\"]:\n",
    "        # score = \"PGS000018_r08\"\n",
    "        cases = sea_merged_whites[traits]\n",
    "        print( score , \"\\t\", traits, \"\\t\", stats.ttest_ind( sea_merged_whites[cases][score], sea_merged_whites[np.invert(cases)][score] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "\n",
    "current_score_file = \"./apply_grs/wGRS49_PGS000349_PGS000018_r0.scores.txt\"\n",
    "\n",
    "pgs_results = pd.read_table( current_score_file, sep = \",\" )\n",
    "pgs_results[\"sample\"] = pgs_results[\"sample\"].str.split(\"_\", expand = True)[0]\n",
    "pgs_results = pgs_results.set_index(\"sample\")\n",
    "min_r2_used = os.path.basename( current_score_file ).split(\".\")[0].split(\"_\")[-1]\n",
    "pgs_results = pgs_results.add_suffix(\"_\" + min_r2_used)\n",
    "pgs_results.head()\n",
    "    # score_name = os.path.basename( current_score_file ).split(\".\")[0]\n",
    "    # pgs_results.rename( columns={ pgs_results.columns[0]:score_name }, inplace=True )\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "kernelspec": {
   "display_name": "Python 3.6.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
