{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook used to QC SEA imputed data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Allocate computing resources on Slurm cluster for imputation data QC\n",
    "\n",
    "\n",
    "```bash\n",
    "$ sdev -m 120 -J jNotebook\n",
    "$ jupyter-notebook --no-browser --ip=0.0.0.0 --port=8010\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "IMPORTANT: Tested only with kernel version <code>Python 3.6.8</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tkr\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os, glob, psutil\n",
    "from zipfile import ZipFile\n",
    "import subprocess\n",
    "import dotenv\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create functions used during script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to sort chromosomes as expected by humans (used during plotting)\n",
    "def natural_sort(l):\n",
    "    convert = lambda text: int(text) if text.isdigit() else text.lower()\n",
    "    alphanum_key = lambda key: [convert(c) for c in re.split('([0-9]+)', key)]\n",
    "    return sorted(l, key=alphanum_key)\n",
    "\n",
    "# Functions used to add vertical and horizontal mean/median lines during plotting\n",
    "def add_mean_horizontal_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding mean line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate mean for each group\n",
    "    m = np.mean(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group mean\n",
    "    ax.axhline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        0.7, \n",
    "        m+0.1 if (m+0.1) < 0.9 else m-0.125,\n",
    "        f'mean={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "\n",
    "def add_median_horizontal_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding median line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate median for each group\n",
    "    m = np.median(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group median\n",
    "    ax.axhline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        0.6,\n",
    "        m+0.1 if (m+0.1) < 0.9 else m-0.125,\n",
    "        f'median={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "    \n",
    "def add_mean_vertical_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding mean line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate mean for each group\n",
    "    m = np.mean(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group mean\n",
    "    ax.axvline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        m+0.05 if (m+0.1) < 0.9 else m-0.4,\n",
    "        0.65,\n",
    "        f'mean={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "    \n",
    "def add_median_vertical_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding median line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate median for each group\n",
    "    m = np.median(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group median\n",
    "    ax.axvline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        m+0.05 if (m+0.1) < 0.9 else m-0.4,\n",
    "        0.65,\n",
    "        f'median={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract all imputed zip files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop over topmed folder with zip files extracting only info.gz files\n",
    "# ZipFile function is too slow to extract large files so we are forced\n",
    "# to make a system call to unzip directly from OS\n",
    "dotenv.load_dotenv(\"./imputed_genotypes/imputation_passwords.env\")\n",
    "\n",
    "for imputation_zip_file in glob.glob( \"./imputed_genotypes/*/*/*.zip\"):\n",
    "    print(\"Extracting file: {0}\".format(imputation_zip_file) )\n",
    "        \n",
    "    imputation_zip_file_info = {\n",
    "        \"root_folder\": imputation_zip_file.split( \"/\" )[0],\n",
    "        \"imputation_root_folder\": imputation_zip_file.split( \"/\" )[1],\n",
    "        \"imputation_server\": imputation_zip_file.split( \"/\" )[2],\n",
    "        \"genome_build\": imputation_zip_file.split( \"/\" )[3],\n",
    "        \"filename\": imputation_zip_file.split( \"/\" )[4]\n",
    "    }\n",
    "    \n",
    "    password_key = imputation_zip_file_info[\"imputation_server\"] + \"_\" + imputation_zip_file_info[\"genome_build\"]\n",
    "    zip_password = os.getenv( password_key )\n",
    "\n",
    "    zip_files_folder = os.path.dirname( imputation_zip_file )\n",
    "\n",
    "    # print( \"unzip -P \" + zip_password + \" -d \" + zip_files_folder + \" \" + imputation_zip_file)    \n",
    "    subprocess.run(\n",
    "        [\"unzip\", \"-P\", zip_password, \"-d\", zip_files_folder, imputation_zip_file],\n",
    "        stdout=subprocess.PIPE\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read <code>info.gz</code> files with metrics from imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1 of 46; 142.29 MB] Reading file: ./blacks/chr19.info.gz\n",
      "[2 of 46; 142.29 MB] Reading file: ./blacks/chr18.info.gz\n",
      "[3 of 46; 142.29 MB] Reading file: ./blacks/chr9.info.gz\n",
      "[4 of 46; 142.29 MB] Reading file: ./blacks/chr17.info.gz\n",
      "[5 of 46; 142.29 MB] Reading file: ./blacks/chr5.info.gz\n",
      "[6 of 46; 142.29 MB] Reading file: ./blacks/chr4.info.gz\n",
      "[7 of 46; 142.29 MB] Reading file: ./blacks/chr16.info.gz\n",
      "[8 of 46; 142.29 MB] Reading file: ./blacks/chr1.info.gz\n",
      "[9 of 46; 142.29 MB] Reading file: ./blacks/chr12.info.gz\n",
      "[10 of 46; 142.29 MB] Reading file: ./blacks/chr15.info.gz\n",
      "[11 of 46; 142.29 MB] Reading file: ./blacks/chr8.info.gz\n",
      "[12 of 46; 142.29 MB] Reading file: ./blacks/chr3.info.gz\n",
      "[13 of 46; 142.29 MB] Reading file: ./blacks/chr7.info.gz\n",
      "[14 of 46; 142.29 MB] Reading file: ./blacks/chr2.info.gz\n",
      "[15 of 46; 142.29 MB] Reading file: ./blacks/chr14.info.gz\n",
      "[16 of 46; 142.29 MB] Reading file: ./blacks/chr6.info.gz\n",
      "[17 of 46; 142.29 MB] Reading file: ./blacks/chr21.info.gz\n",
      "[18 of 46; 663.58 MB] Reading file: ./blacks/chrX.info.gz\n",
      "[19 of 46; 663.59 MB] Reading file: ./blacks/chr10.info.gz\n",
      "[20 of 46; 663.59 MB] Reading file: ./blacks/chr11.info.gz\n",
      "[21 of 46; 663.59 MB] Reading file: ./blacks/chr13.info.gz\n",
      "[22 of 46; 663.59 MB] Reading file: ./blacks/chr22.info.gz\n",
      "[23 of 46; 704.79 MB] Reading file: ./blacks/chr20.info.gz\n",
      "[24 of 46; 704.79 MB] Reading file: ./whites/chr19.info.gz\n",
      "[25 of 46; 704.79 MB] Reading file: ./whites/chr18.info.gz\n",
      "[26 of 46; 704.79 MB] Reading file: ./whites/chr9.info.gz\n",
      "[27 of 46; 704.79 MB] Reading file: ./whites/chr17.info.gz\n",
      "[28 of 46; 704.79 MB] Reading file: ./whites/chr5.info.gz\n",
      "[29 of 46; 704.79 MB] Reading file: ./whites/chr4.info.gz\n",
      "[30 of 46; 704.79 MB] Reading file: ./whites/chr16.info.gz\n",
      "[31 of 46; 704.79 MB] Reading file: ./whites/chr1.info.gz\n",
      "[32 of 46; 704.79 MB] Reading file: ./whites/chr12.info.gz\n",
      "[33 of 46; 704.79 MB] Reading file: ./whites/chr15.info.gz\n",
      "[34 of 46; 704.79 MB] Reading file: ./whites/chr8.info.gz\n",
      "[35 of 46; 704.79 MB] Reading file: ./whites/chr3.info.gz\n",
      "[36 of 46; 704.79 MB] Reading file: ./whites/chr7.info.gz\n",
      "[37 of 46; 704.79 MB] Reading file: ./whites/chr2.info.gz\n",
      "[38 of 46; 704.79 MB] Reading file: ./whites/chr14.info.gz\n",
      "[39 of 46; 704.79 MB] Reading file: ./whites/chr6.info.gz\n",
      "[40 of 46; 704.79 MB] Reading file: ./whites/chr21.info.gz\n",
      "[41 of 46; 704.44 MB] Reading file: ./whites/chrX.info.gz\n",
      "[42 of 46; 704.44 MB] Reading file: ./whites/chr10.info.gz\n",
      "[43 of 46; 704.44 MB] Reading file: ./whites/chr11.info.gz\n",
      "[44 of 46; 704.44 MB] Reading file: ./whites/chr13.info.gz\n",
      "[45 of 46; 704.44 MB] Reading file: ./whites/chr22.info.gz\n",
      "[46 of 46; 733.27 MB] Reading file: ./whites/chr20.info.gz\n",
      "Total memory used during loading: 775.83 MB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>REF(0)</th>\n",
       "      <th>ALT(1)</th>\n",
       "      <th>ALT_Frq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>AvgCall</th>\n",
       "      <th>Rsq</th>\n",
       "      <th>Genotyped</th>\n",
       "      <th>LooRsq</th>\n",
       "      <th>EmpR</th>\n",
       "      <th>EmpRsq</th>\n",
       "      <th>Dose0</th>\n",
       "      <th>Dose1</th>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr21:10035213:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00630</td>\n",
       "      <td>0.00630</td>\n",
       "      <td>0.99435</td>\n",
       "      <td>0.22704</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blacks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr21:10039257:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00579</td>\n",
       "      <td>0.00579</td>\n",
       "      <td>0.99479</td>\n",
       "      <td>0.26245</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blacks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr21:10041326:AT:A</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00716</td>\n",
       "      <td>0.00716</td>\n",
       "      <td>0.99365</td>\n",
       "      <td>0.18958</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blacks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr21:10043906:A:AT</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.02261</td>\n",
       "      <td>0.02261</td>\n",
       "      <td>0.97908</td>\n",
       "      <td>0.14589</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blacks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21:10055730:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.00892</td>\n",
       "      <td>0.00892</td>\n",
       "      <td>0.99186</td>\n",
       "      <td>0.20103</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blacks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795407</th>\n",
       "      <td>chr22:50803709:T:A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.01478</td>\n",
       "      <td>0.01478</td>\n",
       "      <td>0.98522</td>\n",
       "      <td>0.00295</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>whites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795408</th>\n",
       "      <td>chr22:50803843:C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.06840</td>\n",
       "      <td>0.06840</td>\n",
       "      <td>0.93160</td>\n",
       "      <td>0.08471</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>whites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795409</th>\n",
       "      <td>chr22:50804129:A:T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.04074</td>\n",
       "      <td>0.04074</td>\n",
       "      <td>0.95926</td>\n",
       "      <td>0.02850</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>whites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795410</th>\n",
       "      <td>chr22:50805809:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00542</td>\n",
       "      <td>0.00542</td>\n",
       "      <td>0.99458</td>\n",
       "      <td>0.00573</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>whites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795411</th>\n",
       "      <td>chr22:50806802:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00841</td>\n",
       "      <td>0.00841</td>\n",
       "      <td>0.99159</td>\n",
       "      <td>0.02464</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>whites</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>795412 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        SNP REF(0) ALT(1)  ALT_Frq      MAF  AvgCall      Rsq  \\\n",
       "0        chr21:10035213:G:A      G      A  0.00630  0.00630  0.99435  0.22704   \n",
       "1        chr21:10039257:T:C      T      C  0.00579  0.00579  0.99479  0.26245   \n",
       "2       chr21:10041326:AT:A     AT      A  0.00716  0.00716  0.99365  0.18958   \n",
       "3       chr21:10043906:A:AT      A     AT  0.02261  0.02261  0.97908  0.14589   \n",
       "4        chr21:10055730:A:G      A      G  0.00892  0.00892  0.99186  0.20103   \n",
       "...                     ...    ...    ...      ...      ...      ...      ...   \n",
       "795407   chr22:50803709:T:A      T      A  0.01478  0.01478  0.98522  0.00295   \n",
       "795408   chr22:50803843:C:G      C      G  0.06840  0.06840  0.93160  0.08471   \n",
       "795409   chr22:50804129:A:T      A      T  0.04074  0.04074  0.95926  0.02850   \n",
       "795410   chr22:50805809:C:T      C      T  0.00542  0.00542  0.99458  0.00573   \n",
       "795411   chr22:50806802:G:A      G      A  0.00841  0.00841  0.99159  0.02464   \n",
       "\n",
       "       Genotyped LooRsq EmpR EmpRsq Dose0 Dose1    race  \n",
       "0        Imputed      -    -      -     -     -  blacks  \n",
       "1        Imputed      -    -      -     -     -  blacks  \n",
       "2        Imputed      -    -      -     -     -  blacks  \n",
       "3        Imputed      -    -      -     -     -  blacks  \n",
       "4        Imputed      -    -      -     -     -  blacks  \n",
       "...          ...    ...  ...    ...   ...   ...     ...  \n",
       "795407   Imputed      -    -      -     -     -  whites  \n",
       "795408   Imputed      -    -      -     -     -  whites  \n",
       "795409   Imputed      -    -      -     -     -  whites  \n",
       "795410   Imputed      -    -      -     -     -  whites  \n",
       "795411   Imputed      -    -      -     -     -  whites  \n",
       "\n",
       "[795412 rows x 14 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tkr\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os, glob, psutil\n",
    "from zipfile import ZipFile\n",
    "import subprocess\n",
    "import dotenv\n",
    "import re\n",
    "\n",
    "os.chdir(\"/labs/tassimes/rodrigoguarischi/projects/sea/imputed_data/topmed/\")\n",
    "\n",
    "li = []\n",
    "n_file = 0\n",
    "\n",
    "file_list = glob.glob( \"./**/*.info.gz\", recursive = True )\n",
    "\n",
    "# Read all info.gz files printing memory stats during loading\n",
    "for current_info_file in file_list:\n",
    "    n_file += 1\n",
    "    memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "    print( \"[{0} of {1}; {2:,.2f} MB] Reading file: {3}\".format(n_file, len(file_list), memory_used, current_info_file ) )\n",
    "    file_info = current_info_file.split( sep = \"/\" )\n",
    "    current_race = file_info[1]\n",
    "    file_name = file_info[2]\n",
    "    current_imputation_info = pd.read_table( current_info_file )\n",
    "    info_snps_above_005 = current_imputation_info[ current_imputation_info[\"MAF\"] >= 0.005 ]\n",
    "    info_snps_above_005 = info_snps_above_005.assign( race=current_race )\n",
    "    # Delete info table to save memory\n",
    "    del current_imputation_info\n",
    "    li.append( info_snps_above_005 )\n",
    "\n",
    "# Concatenate everything on pandas dataframe named imputation_info\n",
    "imputation_info = pd.concat(li, axis=0, ignore_index=True)\n",
    "\n",
    "# Delete li to save memory\n",
    "del li\n",
    "\n",
    "# Print memory status\n",
    "memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "print( \"Total memory used during loading: {0:,.2f} MB\".format(memory_used) )\n",
    "\n",
    "imputation_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAF threshold >= 0.01\n",
      "         count     min    Q_25    mean  median    Q_75  max\n",
      "race                                                       \n",
      "blacks  403044  0.0040  0.4777  0.6237  0.6670  0.8058  1.0\n",
      "whites  251991  0.0018  0.2894  0.5196  0.5338  0.7467  1.0\n",
      "\n",
      "\n",
      "MAF threshold >= 0.05\n",
      "         count     min    Q_25    mean  median    Q_75  max\n",
      "race                                                       \n",
      "blacks  226229  0.0213  0.4649  0.6201  0.6613  0.8064  1.0\n",
      "whites  170944  0.0100  0.3250  0.5457  0.5705  0.7760  1.0\n",
      "\n",
      "\n",
      "MAF threshold >= 0.1\n",
      "         count     min    Q_25    mean  median    Q_75  max\n",
      "race                                                       \n",
      "blacks  161300  0.0213  0.4524  0.6128  0.6532  0.8025  1.0\n",
      "whites  134202  0.0150  0.3330  0.5476  0.5751  0.7747  1.0\n",
      "\n",
      "\n",
      "MAF threshold >= 0.25\n",
      "        count     min    Q_25    mean  median    Q_75  max\n",
      "race                                                      \n",
      "blacks  77576  0.0213  0.4200  0.5947  0.6366  0.7920  1.0\n",
      "whites  74125  0.0166  0.3337  0.5442  0.5796  0.7675  1.0\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def percentile(n):\n",
    "    def percentile_(x):\n",
    "        return np.percentile(x, n)\n",
    "    percentile_.__name__ = 'Q_%s' % n\n",
    "    return percentile_\n",
    "\n",
    "for MAF_threshold in [0.01, 0.05]:\n",
    "    print( \"MAF threshold >= \" + str(MAF_threshold) )\n",
    "    # print( imputation_info[ imputation_info[\"MAF\"] >= MAF_threshold ].groupby( [ \"race\" ] )[\"Rsq\"].agg( [ \"count\", \"min\", percentile(25), \"mean\", \"median\", percentile(75), \"max\",  ] ).round(4) )\n",
    "    print( imputation_info[ ( imputation_info[\"MAF\"] >= MAF_threshold ) & ( imputation_info[\"MAF\"] <= MAF_threshold ) ].groupby( [ \"race\" ] )[\"Rsq\"].agg( [ \"count\", \"min\", percentile(25), \"mean\", \"median\", percentile(75), \"max\",  ] ).round(4) )\n",
    "    print( \"\\n\" )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_root_folder = \"./imputed_genotypes/\"\n",
    "\n",
    "li = []\n",
    "n_file = 0\n",
    "\n",
    "file_list = glob.glob( imputation_root_folder + \"/**/\" +  \"/*.info.gz\", recursive = True )\n",
    "\n",
    "selected_chrs = [\"chr19.info.gz\", \"chr20.info.gz\", \"chr21.info.gz\", \"chr22.info.gz\", \"chrX.info.gz\"]\n",
    "\n",
    "# Read all info.gz files printing memory stats during loading\n",
    "for current_imputation_file in file_list:\n",
    "\n",
    "    n_file += 1\n",
    "    memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "    print( \"[{0} of {1}; {2:,.2f} MB] Reading file: {3}\".format(n_file, len(file_list), memory_used, current_imputation_file ) )\n",
    "    \n",
    "    file_info = current_imputation_file.split( sep = \"/\" )\n",
    "    imputation_ref = file_info[2]\n",
    "    genome_build = file_info[3]\n",
    "    file_name = file_info[4]\n",
    "    \n",
    "    # Skip some chromosomes for testing purposes\n",
    "    if(file_name not in selected_chrs):\n",
    "        continue\n",
    "    \n",
    "    current_imputation_info = pd.read_table( current_imputation_file )\n",
    "    current_imputation_info[\"Reference_Panel\"] = imputation_ref\n",
    "    current_imputation_info[\"Genome_Build\"] = genome_build\n",
    "\n",
    "    li.append( current_imputation_info )\n",
    "\n",
    "# Concatenate everything on pandas dataframe named imputation_info\n",
    "imputation_info = pd.concat(li, axis=0, ignore_index=True)\n",
    "\n",
    "# Delete li to save memory\n",
    "del li\n",
    "\n",
    "# Print memory status\n",
    "memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "print( \"Total memory used during loading: {0:,.2f} MB\".format(memory_used) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create <code>chr</code> and <code>pos</code> atributes in the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_coords = imputation_info.SNP.str.split(\":\", expand = True)[[0,1]]\n",
    "imputation_info_coords.columns = [\"chr\", \"pos\"]\n",
    "\n",
    "imputation_info = imputation_info.join( imputation_info_coords )\n",
    "\n",
    "# Print the first 5 lines\n",
    "imputation_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "print( \"Memory used {0:,.2f} MB\".format(memory_used) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Small adjustments on pandas dataframes to enable proper interpretation/plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standardize chr names to downstream analysis (Info files are different if from hg19 or hg38)\n",
    "imputation_info[\"chr\"] = imputation_info[\"chr\"].str.replace(\"^chr\",\"\", regex=True)\n",
    "\n",
    "# Subset only to SNPs that were genotyped and save it on another dataframe\n",
    "imputation_info_genotyped = imputation_info[ imputation_info.Genotyped == \"Genotyped\" ]\n",
    "\n",
    "# Cast numeric variables to enable plotting\n",
    "imputation_info = imputation_info.astype( { \n",
    "        'pos': 'int'\n",
    "    }\n",
    ")\n",
    "\n",
    "print( \"Object imputation_info:\\n\", imputation_info.dtypes , \"\\n\" )\n",
    "\n",
    "imputation_info_genotyped = imputation_info_genotyped.astype( { \n",
    "        'LooRsq': 'float', \n",
    "        'EmpR': 'float', \n",
    "        'EmpRsq': 'float',\n",
    "        'Dose0': 'float',\n",
    "        'Dose1': 'float',\n",
    "        'pos': 'int'\n",
    "    }\n",
    ")\n",
    "\n",
    "print( \"Object imputation_info_genotyped:\\n\", imputation_info_genotyped.dtypes , \"\\n\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_categories = [0, 0.0001, 0.001, 0.05, 0.5]\n",
    "\n",
    "imputation_info[\"MAF_categories\"] = pd.cut(\n",
    "    imputation_info.MAF,\n",
    "    maf_categories,\n",
    "    include_lowest=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    width=0.7,\n",
    "    fliersize=2\n",
    "    )\n",
    "\n",
    "# Put the legend out of the figure\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.stripplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    size = 1,\n",
    "    dodge=True\n",
    "    )\n",
    "\n",
    "# Put the legend out of the figure\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ax = sns.boxplot(x=\"tip\", y=\"day\", data=tips, whis=np.inf)\n",
    "\n",
    "ax = sns.boxplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    size = 1,\n",
    "    dodge=True\n",
    "    )\n",
    "\n",
    "ax = sns.stripplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    size = 1,\n",
    "    dodge=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.relplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info_subset,\n",
    "    kind='line',\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed']\n",
    "    )\n",
    "\n",
    "# Put the legend out of the figure\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_subset = imputation_info.sample(100000)\n",
    "imputation_info_subset[\"MAF_categories\"] = imputation_info_subset[\"MAF_categories\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_subset.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_subset[\"MAF_categories\"].head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.histplot(\n",
    "    x = \"Rsq\",\n",
    "    binwidth=0.1,\n",
    "    binrange=(0,1),\n",
    "    data = imputation_info,\n",
    "#     data = imputation_info.sample(1000),\n",
    "    hue = \"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    multiple='dodge',\n",
    "    shrink=0.8\n",
    "    )\n",
    "\n",
    "g.set_yscale(\"log\")\n",
    "\n",
    "# # Put the legend out of the figure\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "# Define which attribute to use on y axis: Rsq\n",
    "y_atribute = \"Rsq\"\n",
    "\n",
    "g = sns.FacetGrid(\n",
    "    imputation_info,\n",
    "    col=\"chr\",\n",
    "    col_wrap=4,\n",
    "    col_order=natural_sort( imputation_info.chr.unique() ),\n",
    "    hue=\"Genotyped\",\n",
    "    sharex=False\n",
    "    )\n",
    "g.map_dataframe(\n",
    "    sns.scatterplot,\n",
    "    x=\"pos\",\n",
    "    y=y_atribute,\n",
    "    alpha=0.4,\n",
    "    s=6\n",
    "    )\n",
    "\n",
    "g.map_dataframe(\n",
    "    add_median_horizontal_line, \n",
    "    var=y_atribute\n",
    "    )\n",
    "g.set_axis_labels(\n",
    "    \"chromosome position\",\n",
    "    y_atribute\n",
    "    )\n",
    "\n",
    "# Add a general title to the figure\n",
    "g.fig.subplots_adjust(top=0.90)\n",
    "g.fig.suptitle( \"Genotyped and imputed SNPs from Michigan HRC hg19 (N={0:,})\".format( len(imputation_info) ) )\n",
    "\n",
    "# Edit title for the subplots\n",
    "imputation_info_counts_by_chr = imputation_info[\"chr\"].value_counts()\n",
    "g.set_titles( col_template=\"{col_name}\" )\n",
    "for subfig in g.axes:\n",
    "    chr_id = str( subfig.get_title() )\n",
    "    number_of_snps = imputation_info_counts_by_chr[ chr_id ]\n",
    "    subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "\n",
    "# format the labels with f-strings\n",
    "for ax in g.axes.flat:\n",
    "    ax.xaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000000)}M'))\n",
    "    \n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.set_style('white')\n",
    "\n",
    "# # Define which attribute to use on y axis: Rsq, LooRsq, EmpR, EmpRsq, Dose0, Dose1\n",
    "# y_atribute = \"LooRsq\"\n",
    "\n",
    "# g = sns.FacetGrid(\n",
    "#     imputation_info_genotyped,\n",
    "# #   row=\"Reference_Panel\",\n",
    "#     col=\"chr\",\n",
    "#     col_wrap=4,\n",
    "#     col_order=natural_sort( imputation_info_genotyped.chr.unique() ),\n",
    "#     sharex=False\n",
    "#     )\n",
    "# g.map_dataframe(\n",
    "#     sns.scatterplot,\n",
    "#     x=\"pos\",\n",
    "#     y=y_atribute,\n",
    "#     alpha=0.5,\n",
    "#     s=10\n",
    "#     )\n",
    "\n",
    "# g.map_dataframe(\n",
    "#     add_median_horizontal_line, \n",
    "#     var=y_atribute\n",
    "#     )\n",
    "# g.set_axis_labels(\n",
    "#     \"chromosome position\",\n",
    "#     y_atribute\n",
    "#     )\n",
    "\n",
    "# # Add a general title to the figure\n",
    "# g.fig.subplots_adjust(top=0.90)\n",
    "# g.fig.suptitle( \"Genotyped SNPs (N={0:,})\".format( len(imputation_info_genotyped) ) )\n",
    "\n",
    "# # Edit title for the subplots\n",
    "# imputation_info_genotyped_counts_by_chr = imputation_info_genotyped[\"chr\"].value_counts()\n",
    "# g.set_titles( col_template=\"{col_name}\" )\n",
    "# for subfig in g.axes:\n",
    "#     chr_id = str( subfig.get_title() )\n",
    "#     number_of_snps = imputation_info_genotyped_counts_by_chr[ chr_id ]\n",
    "#     subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "\n",
    "# # format the labels with f-strings\n",
    "# for ax in g.axes.flat:\n",
    "#     ax.xaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000000)}M'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "# Define which attribute to use on y axis: Rsq, LooRsq, EmpR, EmpRsq, Dose0, Dose1\n",
    "y_atribute = \"LooRsq\"\n",
    "\n",
    "g = sns.FacetGrid(\n",
    "    imputation_info_genotyped,\n",
    "    row=\"chr\",\n",
    "    col=\"Reference_Panel\",\n",
    "#     col_wrap=4,\n",
    "#     col_order=natural_sort( imputation_info_genotyped.chr.unique() ),\n",
    "    sharex=False\n",
    "    )\n",
    "g.map_dataframe(\n",
    "    sns.scatterplot,\n",
    "    x=\"pos\",\n",
    "    y=y_atribute,\n",
    "    alpha=0.5,\n",
    "    s=10\n",
    "    )\n",
    "\n",
    "g.map_dataframe(\n",
    "    add_median_horizontal_line, \n",
    "    var=y_atribute\n",
    "    )\n",
    "g.set_axis_labels(\n",
    "    \"chromosome position\",\n",
    "    y_atribute\n",
    "    )\n",
    "\n",
    "# Add a general title to the figure\n",
    "g.fig.subplots_adjust(top=0.90)\n",
    "g.fig.suptitle( \"Genotyped SNPs (N={0:,})\".format( len(imputation_info_genotyped) ) )\n",
    "\n",
    "# Edit title for the subplots\n",
    "imputation_info_genotyped_counts_by_chr = imputation_info_genotyped[\"chr\"].value_counts()\n",
    "g.set_titles( col_template=\"{col_name}\" )\n",
    "for subfig in g.axes:\n",
    "    chr_id = str( subfig.get_title() )\n",
    "    number_of_snps = imputation_info_genotyped_counts_by_chr[ chr_id ]\n",
    "    subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "\n",
    "# format the labels with f-strings\n",
    "for ax in g.axes.flat:\n",
    "    ax.xaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000000)}M'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_categories = [0, 0.0001, 0.01, 0.05, 0.5]\n",
    "\n",
    "imputation_info[\"MAF_categories\"] = pd.cut(\n",
    "    imputation_info.MAF,\n",
    "    maf_categories\n",
    "    ).astype(str)\n",
    "\n",
    "# median_rsq_by_maf_tmp = imputation_info.groupby(\"MAF_categories\")[\"Rsq\"].median()\n",
    "\n",
    "# median_rsq_by_maf = pd.DataFrame(\n",
    "#     {\n",
    "#     \"MAF_categories\": median_rsq_by_maf_tmp.index.astype(str),\n",
    "#     \"Median_observed_imput_rsq\": median_rsq_by_maf_tmp.values\n",
    "#     }\n",
    "# )\n",
    "\n",
    "# median_rsq_by_maf.head()\n",
    "\n",
    "sns.lineplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    color = \"maroon\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "g = sns.FacetGrid(\n",
    "    imputation_info,\n",
    "    col=\"chr\",\n",
    "    col_wrap=4,\n",
    "    col_order=natural_sort( imputation_info.chr.unique() ),\n",
    "    hue=\"Genotyped\",\n",
    "    sharex=False\n",
    "    )\n",
    "g.map_dataframe(\n",
    "    sns.histplot,\n",
    "    x=\"Rsq\",\n",
    "    binwidth=0.1,\n",
    "    binrange=(0,1)\n",
    "    )\n",
    "\n",
    "g.map_dataframe(\n",
    "    add_median_vertical_line, \n",
    "    var=\"Rsq\"\n",
    "    )\n",
    "\n",
    "# Add a general title to the figure\n",
    "g.fig.subplots_adjust(top=0.90)\n",
    "g.fig.suptitle( \"Genotyped and imputed SNPs from Michigan HRC hg19 (N={0:,})\".format( len(imputation_info) ) )\n",
    "\n",
    "# Edit title for the subplots\n",
    "imputation_info_counts_by_chr = imputation_info[\"chr\"].value_counts()\n",
    "g.set_titles( col_template=\"{col_name}\" )\n",
    "for subfig in g.axes:\n",
    "    chr_id = str( subfig.get_title() )\n",
    "    number_of_snps = imputation_info_counts_by_chr[ chr_id ]\n",
    "    subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "    \n",
    "# format the labels with f-strings\n",
    "for ax in g.axes.flat:\n",
    "    ax.yaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000)}k'))\n",
    "    # ax.set_yscale(\"log\")\n",
    "    \n",
    "# g.set_yscale(\"log\")\n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_categories = [0, 0.0001, 0.01, 0.05, 0.5]\n",
    "\n",
    "imputation_info[\"MAF_categories\"] = pd.cut(\n",
    "    imputation_info.MAF,\n",
    "    maf_categories\n",
    "    )\n",
    "\n",
    "median_rsq_by_maf_tmp = imputation_info.groupby(\"MAF_categories\")[\"Rsq\"].median()\n",
    "\n",
    "median_rsq_by_maf = pd.DataFrame(\n",
    "    {\n",
    "    \"MAF_categories\": median_rsq_by_maf_tmp.index.astype(str),\n",
    "    \"Median_observed_imput_rsq\": median_rsq_by_maf_tmp.values\n",
    "    }\n",
    ")\n",
    "\n",
    "g = sns.lineplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Median_observed_imput_rsq\",\n",
    "    data = median_rsq_by_maf,\n",
    "    color = \"maroon\"\n",
    "    )\n",
    "\n",
    "g.set_title( \"Genotyped and imputed SNPs from Michigan HRC hg19\\n(N={0:,})\".format( len(imputation_info) ) )\n",
    "g.set(\n",
    "    xlabel = \"Reference panel Minor Allele Frequency\",\n",
    "    ylabel = \"Median Observed Imput Rsq\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    # sns.histplot,\n",
    "    # x=\"Rsq\",\n",
    "    # bins=10\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.histplot(\n",
    "    x = \"MAF\",\n",
    "    y = \"Rsq\",\n",
    "    # hue=\"maf_categories\",\n",
    "    data = imputation_info,\n",
    "    bins = [0,0.0001,0.01,0.05,0.5]\n",
    "    # stat='probability',\n",
    "    # cumulative=True,\n",
    "    # fill=False,\n",
    "    # element='step'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imputation_info[\"maf_categories\"] = pd.cut(\n",
    "#     imputation_info,\n",
    "#     [-np.inf, 0.001, 0.1, 0.2, 0.3, np.inf]\n",
    "#     )\n",
    "\n",
    "print( imputation_info.groupby(\"maf_categories\")[\"Rsq\"].mean() )\n",
    "\n",
    "# sns.histplot(\n",
    "#     x = \"Rsq\",\n",
    "#     # y = \"Rsq\",\n",
    "#     hue=\"maf_categories\",\n",
    "#     data = imputation_info\n",
    "#     # cumulative=True,\n",
    "#     # kde=True\n",
    "# )\n",
    "#     x = \"Rsq\",\n",
    "#     data = imputation_info,\n",
    "#     bins = 10,\n",
    "#     element='step',\n",
    "#     col=\n",
    "#     )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a plot of **Rsq** along with the chromosome genomic positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create scatter plot of genomic positions by estimated r2 (from minimac4 leave one out column)\n",
    "plt.scatter( \n",
    "    x = chr21_genotyped.pos.astype(int), \n",
    "    y = chr21_genotyped.LooRsq.astype(float), \n",
    "    marker = 'o',\n",
    "    color='gray',\n",
    "    s=3,\n",
    "    alpha=0.75\n",
    "    )\n",
    "\n",
    "# Add horizontal line with mean LooRsq\n",
    "mean_rsq = statistics.mean( chr21_genotyped.LooRsq.astype(float) )\n",
    "plt.axhline(\n",
    "    y=mean_rsq, \n",
    "    color='red',\n",
    "    linestyle='--'\n",
    "    )\n",
    "plt.text( \n",
    "    x = max(chr21_genotyped.pos.astype(int))*0.825,\n",
    "    y = mean_rsq+0.025,\n",
    "    s = \"Mean R\\u00b2 = {:.2f}\".format(mean_rsq),\n",
    "    color='red'\n",
    "    )\n",
    "\n",
    "# Adjust labels and axis\n",
    "plt.title( \"chr21 (N = \" + str(len(chr21_genotyped)) + \")\" )\n",
    "plt.xlabel(\"chr21 genomic position\")\n",
    "plt.ylabel(\"Empirical (Leave-One-Out) R\\u00b2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a histogram of LooRsq\n",
    "plt.hist(\n",
    "  x = chr21_genotyped.LooRsq.astype(float),\n",
    "  bins = 10,\n",
    "  color='black',\n",
    "  alpha = 0.6\n",
    "  )\n",
    "\n",
    "# Adjust labels and axis\n",
    "plt.title( \"chr21 (N = \" + str(len(chr21_genotyped)) + \")\" )\n",
    "plt.xlabel(\"Empirical (Leave-One-Out) R\\u00b2\")\n",
    "plt.ylabel(\"Frequency\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get number of passing variants at different RSq levels \n",
    "rsq_filter = [0, 0.001, 0.1, 0.2, 0.3]\n",
    "\n",
    "total = len(chr21.Rsq.astype(float))\n",
    "\n",
    "for c_rsq in rsq_filter:\n",
    "    passing = sum ( chr21.Rsq.astype(float) >= c_rsq)\n",
    "    print (\"Passing = {0:,}, filtered= {1:,} \\t (RSq threshold = {2:,})\".format(passing, total-passing, c_rsq) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a plot of **MAF** by **Mean imputation Rsq** "
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "kernelspec": {
   "display_name": "Python 3.6.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
