{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook used to QC SEA imputed data from Michigan imputation server (MIS) and TopMed imputation server"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Allocate computing resources on Slurm cluster for imputation data QC\n",
    "\n",
    "\n",
    "```bash\n",
    "$ sdev -m 120 -J jNotebook\n",
    "$ jupyter-notebook --no-browser --ip=0.0.0.0 --port=8010\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "IMPORTANT: Tested only with kernel version <code>Python 3.6.8</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tkr\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os, glob, psutil\n",
    "from zipfile import ZipFile\n",
    "import subprocess\n",
    "import dotenv\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create functions used during script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to sort chromosomes as expected by humans (used during plotting)\n",
    "def natural_sort(l):\n",
    "    convert = lambda text: int(text) if text.isdigit() else text.lower()\n",
    "    alphanum_key = lambda key: [convert(c) for c in re.split('([0-9]+)', key)]\n",
    "    return sorted(l, key=alphanum_key)\n",
    "\n",
    "# Functions used to add vertical and horizontal mean/median lines during plotting\n",
    "def add_mean_horizontal_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding mean line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate mean for each group\n",
    "    m = np.mean(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group mean\n",
    "    ax.axhline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        0.7, \n",
    "        m+0.1 if (m+0.1) < 0.9 else m-0.125,\n",
    "        f'mean={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "\n",
    "def add_median_horizontal_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding median line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate median for each group\n",
    "    m = np.median(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group median\n",
    "    ax.axhline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        0.6,\n",
    "        m+0.1 if (m+0.1) < 0.9 else m-0.125,\n",
    "        f'median={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "    \n",
    "def add_mean_vertical_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding mean line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate mean for each group\n",
    "    m = np.mean(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group mean\n",
    "    ax.axvline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        m+0.05 if (m+0.1) < 0.9 else m-0.4,\n",
    "        0.65,\n",
    "        f'mean={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "    \n",
    "def add_median_vertical_line(data, var=None, **kws):\n",
    "    \n",
    "    # If no variable provided skip adding median line\n",
    "    if not var: return\n",
    "    \n",
    "    #Calculate median for each group\n",
    "    m = np.median(data[var])\n",
    "    \n",
    "    #Get current axis\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    #add line at group median\n",
    "    ax.axvline(\n",
    "        m,\n",
    "        color='maroon',\n",
    "        lw=2,\n",
    "        ls='--'\n",
    "        )\n",
    "        \n",
    "    ax.text(\n",
    "        m+0.05 if (m+0.1) < 0.9 else m-0.4,\n",
    "        0.65,\n",
    "        f'median={m:.2f}',\n",
    "        # transforms positions to range from (0,0) to (1,1)\n",
    "        transform=ax.transAxes,\n",
    "        color='maroon',\n",
    "        fontweight='bold',\n",
    "        fontsize=8\n",
    "        )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract all imputed zip files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop over topmed folder with zip files extracting only info.gz files\n",
    "# ZipFile function is too slow to extract large files so we are forced\n",
    "# to make a system call to unzip directly from OS\n",
    "dotenv.load_dotenv(\"./imputed_genotypes/imputation_passwords.env\")\n",
    "\n",
    "for imputation_zip_file in glob.glob( \"./imputed_genotypes/*/*/*.zip\"):\n",
    "    print(\"Extracting file: {0}\".format(imputation_zip_file) )\n",
    "        \n",
    "    imputation_zip_file_info = {\n",
    "        \"root_folder\": imputation_zip_file.split( \"/\" )[0],\n",
    "        \"imputation_root_folder\": imputation_zip_file.split( \"/\" )[1],\n",
    "        \"imputation_server\": imputation_zip_file.split( \"/\" )[2],\n",
    "        \"genome_build\": imputation_zip_file.split( \"/\" )[3],\n",
    "        \"filename\": imputation_zip_file.split( \"/\" )[4]\n",
    "    }\n",
    "    \n",
    "    password_key = imputation_zip_file_info[\"imputation_server\"] + \"_\" + imputation_zip_file_info[\"genome_build\"]\n",
    "    zip_password = os.getenv( password_key )\n",
    "\n",
    "    zip_files_folder = os.path.dirname( imputation_zip_file )\n",
    "\n",
    "    # print( \"unzip -P \" + zip_password + \" -d \" + zip_files_folder + \" \" + imputation_zip_file)    \n",
    "    subprocess.run(\n",
    "        [\"unzip\", \"-P\", zip_password, \"-d\", zip_files_folder, imputation_zip_file],\n",
    "        stdout=subprocess.PIPE\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read <code>info.gz</code> files with metrics from imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1 of 46; 117.45 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr19.info.gz\n",
      "[2 of 46; 273.07 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr18.info.gz\n",
      "[3 of 46; 273.07 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr9.info.gz\n",
      "[4 of 46; 273.07 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr17.info.gz\n",
      "[5 of 46; 273.07 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr5.info.gz\n",
      "[6 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr4.info.gz\n",
      "[7 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr16.info.gz\n",
      "[8 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr1.info.gz\n",
      "[9 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr12.info.gz\n",
      "[10 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr15.info.gz\n",
      "[11 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr8.info.gz\n",
      "[12 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr3.info.gz\n",
      "[13 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr7.info.gz\n",
      "[14 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr2.info.gz\n",
      "[15 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr14.info.gz\n",
      "[16 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr6.info.gz\n",
      "[17 of 46; 273.32 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr21.info.gz\n",
      "[18 of 46; 389.69 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chrX.info.gz\n",
      "[19 of 46; 604.33 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr10.info.gz\n",
      "[20 of 46; 604.36 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr11.info.gz\n",
      "[21 of 46; 604.36 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr13.info.gz\n",
      "[22 of 46; 604.36 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr22.info.gz\n",
      "[23 of 46; 694.05 MB] Reading file: ./imputed_genotypes/michigan_hrc/hg19/chr20.info.gz\n",
      "[24 of 46; 852.26 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr19.info.gz\n",
      "[25 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr18.info.gz\n",
      "[26 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr9.info.gz\n",
      "[27 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr17.info.gz\n",
      "[28 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr5.info.gz\n",
      "[29 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr4.info.gz\n",
      "[30 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr16.info.gz\n",
      "[31 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr1.info.gz\n",
      "[32 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr12.info.gz\n",
      "[33 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr15.info.gz\n",
      "[34 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr8.info.gz\n",
      "[35 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr3.info.gz\n",
      "[36 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr7.info.gz\n",
      "[37 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr2.info.gz\n",
      "[38 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr14.info.gz\n",
      "[39 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr6.info.gz\n",
      "[40 of 46; 2,063.32 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr21.info.gz\n",
      "[41 of 46; 2,955.25 MB] Reading file: ./imputed_genotypes/topmed/hg38/chrX.info.gz\n",
      "[42 of 46; 6,442.14 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr10.info.gz\n",
      "[43 of 46; 6,442.14 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr11.info.gz\n",
      "[44 of 46; 6,442.14 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr13.info.gz\n",
      "[45 of 46; 6,442.14 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr22.info.gz\n",
      "[46 of 46; 7,020.57 MB] Reading file: ./imputed_genotypes/topmed/hg38/chr20.info.gz\n",
      "Total memory used during loading: 9,992.62 MB\n"
     ]
    }
   ],
   "source": [
    "imputation_root_folder = \"./imputed_genotypes/\"\n",
    "\n",
    "li = []\n",
    "n_file = 0\n",
    "\n",
    "file_list = glob.glob( imputation_root_folder + \"/**/\" +  \"/*.info.gz\", recursive = True )\n",
    "\n",
    "selected_chrs = [\"chr19.info.gz\", \"chr20.info.gz\", \"chr21.info.gz\", \"chr22.info.gz\", \"chrX.info.gz\"]\n",
    "\n",
    "# Read all info.gz files printing memory stats during loading\n",
    "for current_imputation_file in file_list:\n",
    "\n",
    "    n_file += 1\n",
    "    memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "    print( \"[{0} of {1}; {2:,.2f} MB] Reading file: {3}\".format(n_file, len(file_list), memory_used, current_imputation_file ) )\n",
    "    \n",
    "    file_info = current_imputation_file.split( sep = \"/\" )\n",
    "    imputation_ref = file_info[2]\n",
    "    genome_build = file_info[3]\n",
    "    file_name = file_info[4]\n",
    "    \n",
    "    # Skip some chromosomes for testing purposes\n",
    "    if(file_name not in selected_chrs):\n",
    "        continue\n",
    "    \n",
    "    current_imputation_info = pd.read_table( current_imputation_file )\n",
    "    current_imputation_info[\"Reference_Panel\"] = imputation_ref\n",
    "    current_imputation_info[\"Genome_Build\"] = genome_build\n",
    "\n",
    "    li.append( current_imputation_info )\n",
    "\n",
    "# Concatenate everything on pandas dataframe named imputation_info\n",
    "imputation_info = pd.concat(li, axis=0, ignore_index=True)\n",
    "\n",
    "# Delete li to save memory\n",
    "del li\n",
    "\n",
    "# Print memory status\n",
    "memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "print( \"Total memory used during loading: {0:,.2f} MB\".format(memory_used) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create <code>chr</code> and <code>pos</code> atributes in the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>REF(0)</th>\n",
       "      <th>ALT(1)</th>\n",
       "      <th>ALT_Frq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>AvgCall</th>\n",
       "      <th>Rsq</th>\n",
       "      <th>Genotyped</th>\n",
       "      <th>LooRsq</th>\n",
       "      <th>EmpR</th>\n",
       "      <th>EmpRsq</th>\n",
       "      <th>Dose0</th>\n",
       "      <th>Dose1</th>\n",
       "      <th>Reference_Panel</th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19:89282:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.10673</td>\n",
       "      <td>0.10673</td>\n",
       "      <td>0.89327</td>\n",
       "      <td>0.00155</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>michigan_hrc</td>\n",
       "      <td>hg19</td>\n",
       "      <td>19</td>\n",
       "      <td>89282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19:89304:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>michigan_hrc</td>\n",
       "      <td>hg19</td>\n",
       "      <td>19</td>\n",
       "      <td>89304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19:89309:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.00031</td>\n",
       "      <td>0.00031</td>\n",
       "      <td>0.99969</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>michigan_hrc</td>\n",
       "      <td>hg19</td>\n",
       "      <td>19</td>\n",
       "      <td>89309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19:90678:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>michigan_hrc</td>\n",
       "      <td>hg19</td>\n",
       "      <td>19</td>\n",
       "      <td>90678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19:90974:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00051</td>\n",
       "      <td>0.00051</td>\n",
       "      <td>0.99949</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Imputed</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>michigan_hrc</td>\n",
       "      <td>hg19</td>\n",
       "      <td>19</td>\n",
       "      <td>90974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            SNP REF(0) ALT(1)  ALT_Frq      MAF  AvgCall      Rsq Genotyped  \\\n",
       "0  19:89282:C:T      C      T  0.10673  0.10673  0.89327  0.00155   Imputed   \n",
       "1  19:89304:G:A      G      A  0.00000  0.00000  1.00000  0.00000   Imputed   \n",
       "2  19:89309:A:G      A      G  0.00031  0.00031  0.99969  0.00001   Imputed   \n",
       "3  19:90678:G:A      G      A  0.00000  0.00000  1.00000  0.00000   Imputed   \n",
       "4  19:90974:T:C      T      C  0.00051  0.00051  0.99949  0.00001   Imputed   \n",
       "\n",
       "  LooRsq EmpR EmpRsq Dose0 Dose1 Reference_Panel Genome_Build chr    pos  \n",
       "0      -    -      -     -     -    michigan_hrc         hg19  19  89282  \n",
       "1      -    -      -     -     -    michigan_hrc         hg19  19  89304  \n",
       "2      -    -      -     -     -    michigan_hrc         hg19  19  89309  \n",
       "3      -    -      -     -     -    michigan_hrc         hg19  19  90678  \n",
       "4      -    -      -     -     -    michigan_hrc         hg19  19  90974  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imputation_info_coords = imputation_info.SNP.str.split(\":\", expand = True)[[0,1]]\n",
    "imputation_info_coords.columns = [\"chr\", \"pos\"]\n",
    "\n",
    "imputation_info = imputation_info.join( imputation_info_coords )\n",
    "\n",
    "# Print the first 5 lines\n",
    "imputation_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Memory used 22,704.49 MB\n"
     ]
    }
   ],
   "source": [
    "memory_used = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2;\n",
    "print( \"Memory used {0:,.2f} MB\".format(memory_used) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Small adjustments on pandas dataframes to enable proper interpretation/plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Object imputation_info:\n",
      " SNP                 object\n",
      "REF(0)              object\n",
      "ALT(1)              object\n",
      "ALT_Frq            float64\n",
      "MAF                float64\n",
      "AvgCall            float64\n",
      "Rsq                float64\n",
      "Genotyped           object\n",
      "LooRsq              object\n",
      "EmpR                object\n",
      "EmpRsq              object\n",
      "Dose0               object\n",
      "Dose1               object\n",
      "Reference_Panel     object\n",
      "Genome_Build        object\n",
      "chr                 object\n",
      "pos                  int64\n",
      "dtype: object \n",
      "\n",
      "Object imputation_info_genotyped:\n",
      " SNP                 object\n",
      "REF(0)              object\n",
      "ALT(1)              object\n",
      "ALT_Frq            float64\n",
      "MAF                float64\n",
      "AvgCall            float64\n",
      "Rsq                float64\n",
      "Genotyped           object\n",
      "LooRsq             float64\n",
      "EmpR               float64\n",
      "EmpRsq             float64\n",
      "Dose0              float64\n",
      "Dose1              float64\n",
      "Reference_Panel     object\n",
      "Genome_Build        object\n",
      "chr                 object\n",
      "pos                  int64\n",
      "dtype: object \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Standardize chr names to downstream analysis (Info files are different if from hg19 or hg38)\n",
    "imputation_info[\"chr\"] = imputation_info[\"chr\"].str.replace(\"^chr\",\"\", regex=True)\n",
    "\n",
    "# Subset only to SNPs that were genotyped and save it on another dataframe\n",
    "imputation_info_genotyped = imputation_info[ imputation_info.Genotyped == \"Genotyped\" ]\n",
    "\n",
    "# Cast numeric variables to enable plotting\n",
    "imputation_info = imputation_info.astype( { \n",
    "        'pos': 'int'\n",
    "    }\n",
    ")\n",
    "\n",
    "print( \"Object imputation_info:\\n\", imputation_info.dtypes , \"\\n\" )\n",
    "\n",
    "imputation_info_genotyped = imputation_info_genotyped.astype( { \n",
    "        'LooRsq': 'float', \n",
    "        'EmpR': 'float', \n",
    "        'EmpRsq': 'float',\n",
    "        'Dose0': 'float',\n",
    "        'Dose1': 'float',\n",
    "        'pos': 'int'\n",
    "    }\n",
    ")\n",
    "\n",
    "print( \"Object imputation_info_genotyped:\\n\", imputation_info_genotyped.dtypes , \"\\n\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_categories = [0, 0.0001, 0.001, 0.05, 0.5]\n",
    "\n",
    "imputation_info[\"MAF_categories\"] = pd.cut(\n",
    "    imputation_info.MAF,\n",
    "    maf_categories,\n",
    "    include_lowest=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x2b03bb60f668>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfkAAAEHCAYAAACtLj+NAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAmhElEQVR4nO3de3RU5b3/8fc3kxAIN7kdrSEQPOEiIIhQrFiU46Wl1GJP4SjaytFfvdBWyrG1VWsXtrrqqbX1KGhtrb3hvdr181B/9LCsxR49PSiIioCgo4IQRQG5JwJJvr8/ZgeHmElmktkzmZ3Pa61Zmdn72c/znSeTfGfv/exnm7sjIiIi0VOU7wBEREQkHEryIiIiEaUkLyIiElFK8iIiIhGlJC8iIhJRxfkOIFP9+/f3ysrKfIchIlJQXnjhhe3uPqCddfxDcXHxvcBotJPYETQAa+rq6i4dP378+80VKLgkX1lZycqVK/MdhohIQTGzTe2to7i4+N5jjjnm+AEDBuwsKirS9dd51tDQYNu2bRu5devWe4HpzZXRNzEREUnX6AEDBuxRgu8YioqKfMCAAbtJHFlpvkwO4xERkcJWpATfsQS/j5S5XEleREQkogrunLyIiHQMl33jm8O379zTJVv19e/T6+Cv7lqwIVv1iZK8iIi00fade7q8XX5G1pI81X9tdxUPPPBA77Vr13a7+eabtza3fsOGDV3OOeecoa+//vrapuv+7d/+7dgpU6bs/eIXv7i33YG0oqysbFxNTc2LYbcTWpI3s98A5wDvu/vHBgWYmQF3ANOAGuBid18VVjwiIhJ9X/7yl3cDu9uy7e233/5OlsNpl0OHDlFSUtKuOsI8J/87YGoL6z8HDA0elwN3hxiL5NmhQ4f46U9/yqFDh/IdiogUqA0bNnQZMmTIqBkzZlRWVlaOnj59+pDHH3+850knnTRi8ODBo5ctW1a2YMGCfrNnzx4EsHnz5uKzzz77H4cPHz5y+PDhI5988snuAPX19cyaNWtwVVXVqFNPPXXovn37DGDGjBmVv/3tb/sAPPLII72HDBkyatSoUcdffPHFFf/0T/9UBbBs2bKyE088ccTxxx8/cty4cSNefvnlUoAFCxb0+8xnPvOPkydPHjp48ODRc+bMGdja+5k7d2758OHDR44dO3bE5s2bixtjuPDCCweNGTNmxNe+9rWBa9asKZ00adKw4cOHjxw5cuTxa9euLc2kz0Lbk3f3/zazyhaKnAss8sS9bpeb2VFm9gl3fzesmCR/Zs6cye7du3n22Wd5/PHHs1bvwoULicfjza6rrq4GoLy8vNn1VVVVzJ07N2uxZFs+3ltU+7M97wva9t5aajOddjtyf+bT5s2buz7yyCNvjh8/fuOYMWOOf+CBB/qtXLly/YMPPnjUj370o09Mnz59V2PZOXPmDJo8efLe+fPnv1FXV8fu3btj27dvj7399ttd77///jcnTZq0adq0acctWrSoz9e//vUPGrerqamxefPmDX766afXjxgx4uAXvvCFIY3rxo4d++GKFSvWl5SU8Pjjj/f87ne/O3Dp0qVvAKxbt67s5ZdfXtetW7eGqqqq0VdfffV7VVVVze7Z1NbWFp1yyin7Fi5cWD1nzpyBCxcuHPCTn/zkXYB33323y6pVq9YXFxczZsyYEVdfffXW2bNn76qpqbH6+nrLpL/yObq+HNic9HpLsOxjzOxyM1tpZiu3bduWk+Aku3bvThw927VrV87arK2tpba2Nmft5VI+3ltU+zNf7yuq/Rm28vLyAxMnTqyNxWIMGzas9owzzthTVFTESSedVLNly5Yj9nL//ve/9/zOd76zDaC4uJh+/frVN9YxadKkWoBx48bVbNy48YjtXnrppa4VFRUHRowYcRBg1qxZh78AfPDBB7Fp06b949ChQ0d997vfrXjttde6Nq779Kc/vadfv371ZWVlXlVV9eEbb7yRcq+7pKTEZ82atRtg/Pjx+zdt2nR4bMOXvvSlncXFxezcubPovffe6zJ79uxdAGVlZd6zZ8+GTPqrIAbeufs9wD0AEyZM0DWaclhLezrz5s0D4I477shVOFmVj/cW1f7saH0ZZrtR16VLl8M5oKioiK5duzpALBYj3b3c5DpisZjX1tamvcN7zTXXlJ9++ul7n3zyyTc2bNjQ5Ywzzhieqt5Dhw6ljKe4uNiLiooan1NXV3e4bI8ePTJK5C3JZ5KvBiqSXg8MlomISAHo36fXwWyMiD+iviw69dRT9956660D5s+f/37j4fp0thszZsyHmzdvLt2wYUOX4cOHH3zkkUf6Nq7bs2dPbODAgQcBfvnLX/bPZrxN9enTp+GYY445eN999x110UUX7aqtrbW6ujrLZG8+n0l+MXClmT0MnAzs1vl4EZHC0dGvab/77rvfvvjiiwcPGzasf1FREXfeeeemioqKVkf/9ujRw2+77bZNU6dOHVpWVtYwduzY/Y3rrrnmmq2XXnrpkFtuueXYs88+e1eobwC4//7737rssssG33TTTceWlJT4o48++sbIkSPT/jIU5iV0DwFTgP5mtgW4ASgBcPdfAEtIXD4XJ3EJ3SVhxSIiIoVv+PDhB5Ovb//jH/+4McW6HQAVFRV1Tz311BtN60mu48Ybb3yvufo+//nP773wwgvXNjQ0MHv27EHjx4/fD3DWWWft37hx45rGcgsWLHgH4Jvf/OaOxnYBli1blnrUJZB8jfwll1yy85JLLtnZNAaAE0444cDy5ctfa6muloQ5uv6CVtY78I2w2hcREWmr22+/vf9DDz3U/9ChQzZq1Kiab33rW9vzHVNbFMTAOxERkVy64YYb3r/hhhuavUd7JsaMGTPi4MGDRwzsW7Ro0VsTJ07MyaUVSvIiIiIhWb169fp8tq+70ImIiESUkryIiEhEKcmLiIhElM7Ji4hIm1z9jUuH79u1I2u3mu1xVL+DP73r3pTX3m/fvj1277339r322mvzPr/5jBkzKs8555zdjZe+dVRK8iIi0ib7du3o8r3h8awl+ZtbmVpnx44dsV//+tf/0BGSfKHQ4XoRESkI3/72twdu3ry5dMSIESOvuOKKgVdcccXAoUOHjho2bNjIX/3qV30AnnjiiZ4TJkwYPmXKlKrKysrRF1544aD6+noAysrKxl1xxRUDq6qqRk2aNGnYsmXLyiZOnDh84MCBJzzwwAO9Aerq6rjiiisGjh49+vhhw4aNvPXWW/sDNE6KU1lZOXrSpEnDtm/fXhA7yUryIiJSEH72s59tqaioOLB+/fp1p5xyyr5XXnml26uvvrr2qaeeem3+/PkDN23aVALwyiuvdP/5z3/+djweX7Nx48bSRYsW9YHE7V3PPPPMPfF4fG337t3rv//975c/88wzrz366KPxm266qRwSk+D07t27fs2aNa++/PLLr/7+978fsH79+i733XffUfF4vDQej6958MEH31q1alWPfPZFugrim4iIiEiyZ555pud55533QXFxMRUVFXUnn3zyvmeffbasd+/eDSeccML+xvndzzvvvA+eeeaZHpdccsnOkpISnzlz5h6AUaNG1ZaWljaUlpb6xIkTa6urq7sA/OUvf+m1fv36ssWLF/cB2Lt3b2zdunVd//a3vx1ur7Ky8tApp5yyN3/vPn1K8iIiEilm1uzr5Nu7FhUVUVpa+rHb1Lq7/exnP3t7xowZe5LreOKJJ3rnIPSs0+F6EREpCL17967fv39/EcBpp52297HHHutbV1fHO++8U/z888/3mDx58n5IHK5fv359l/r6eh577LG+kydPTnuv++yzz9599913Dzhw4IABrF69unTPnj1Fp59++uH2Nm3aVLJ8+fKe4bzL7NKevIiItEmPo/odbG1EfKb1tbT+mGOOqR8/fvy+oUOHjjrjjDN2jxo1qvb4448fZWb+wx/+cMugQYPqVq9ezejRo/fPmTNn0MaNG7tOmjRpz0UXXbQr3Riuuuqq7Rs3biw94YQTjnd369u376ElS5a8cdFFF+166qmnelVVVY0+9thjD4wbN25fu99wDijJi4hIm7R0TXtY/vSnP73VZNGWpmV69uxZ39ytXpNv73rbbbe909y6WCzGnXfeWQ1UN91+0aJFb7c17nzR4XoREZGIUpIXEZHIOOecc/Y2txffWSnJi4hIuhoaGhqs9WKSK8HvoyHVeiV5ERFJ15pt27b1VqLvGBoaGmzbtm29gTWpymjgnYiIpKWuru7SrVu33rt169bRaCexI2gA1tTV1V2aqoCSvIiIpGX8+PHvA9PzHYekT9/EREREIkpJXkREJKKU5EVERCJKSV5ERCSilORFREQiSkleREQkopTkRUREIkpJXkREJKKU5EVERCJKSV5ERCSilORFREQiKtQkb2ZTzWyDmcXN7Npm1g8ys2Vm9qKZrTazaWHGIyIi0pmEluTNLAbcBXwOGAlcYGYjmxT7PvAHdx8HzAJ+HlY8IiIinU2Ye/ITgbi7v+nuB4GHgXOblHGgV/C8N/BOiPGIiIh0KmEm+XJgc9LrLcGyZD8AvmJmW4AlwNzmKjKzy81spZmt3LZtWxixioiIRE6+B95dAPzO3QcC04D7zOxjMbn7Pe4+wd0nDBgwIOdBioiIFKIwk3w1UJH0emCwLNlXgT8AuPv/Al2B/iHGJCIi0mmEmeRXAEPNbIiZdSExsG5xkzJvA2cCmNnxJJK8jseLiIhkQWhJ3t3rgCuBpcCrJEbRrzWzG81selDs28BlZvYy8BBwsbt7WDGJiIh0JsVhVu7uS0gMqEteNj/p+Trg1DBjEBER6azyPfBOREREQqIkLyIiElFK8iIiIhGlJC8iIhJRSvIiIiIRpSQvIiISUUryIiIiEaUkLyIiElFK8iIiIhGlJC8iIhJRSvIiIiIRpSQvIiISUaHeoEZEOp6FCxcSj8cz3q5xm3nz5mW8bVVVFXPnzs14OxFpHyV5kU4mHo/z+toXGdSjPqPtuhxKHPg7sGllRtu9vS+WUXkRyR4leZFOaFCPer530p6ctHXzql45aUdEPk5JXkSkjdp66gN0+kNyQ0leRKSN2nrqA3T6Q3JDSV5EpB1yeeoDdPpDMqNL6ERERCJKSV5ERCSilORFREQiSkleREQkopTkRUREIkpJXkREJKKU5EVERCJKSV5ERCSilORFREQiSkleREQkopTkRUREIkpJXkREJKJCTfJmNtXMNphZ3MyuTVHmPDNbZ2ZrzezBMOMRERHpTEK7C52ZxYC7gLOBLcAKM1vs7uuSygwFrgNOdfedZvYPYcUjIiLS2YS5Jz8RiLv7m+5+EHgYOLdJmcuAu9x9J4C7vx9iPCIiIp1KmPeTLwc2J73eApzcpMwwADP7HyAG/MDd/6tpRWZ2OXA5wKBBg0IJVkQK28KFC4nH4xlv17jNvHnz2rRtRUnGm4nkTJhJPt32hwJTgIHAf5vZCe6+K7mQu98D3AMwYcIEz3GMIlIA4vE4L615lfqyvhltV3Qw8S/lhTffy7jN2P4aOCrjzURyJswkXw1UJL0eGCxLtgV4zt0PAW+Z2Wskkv6KEOMSkYiqL+tL7YhpOWuvx6r7gIM5a08kU2Gek18BDDWzIWbWBZgFLG5S5nESe/GYWX8Sh+/fDDEmERGRTiO0JO/udcCVwFLgVeAP7r7WzG40s+lBsaXADjNbBywDvuPuO8KKSUREpDMJ9Zy8uy8BljRZNj/puQPfCh4iIiKSRZrxTkREJKKU5EVERCJKSV5ERCSilORFREQiSkleREQkopTkRUREIkpJXkREJKKU5EVERCIqrclwzGxBS+vd/ZvZCUdERESyJd09+a7AScDrweNEoAvwQvAQERGRDibdaW3HAJ8O5qPHzH4BPOPuc0KLTERERNol3STfB+gFfBC87hEsE5E2WrhwIfF4vE3bNm43b968jLetrq6mf5taFZFCk26S/zHwopktAww4DfhBWEGJdAbxeJyX1rxKfVnfjLctOugAvPDmexltF6v5gB5dS6Ak4yZFpAClleTd/bdm9mfg5GDRNe6+NbywRDqH+rK+1I6YlrP2uq1fAg17c9aeiORXWgPvzOxUYK+7/yfQE/iumQ0ONTIRERFpl3RH198N1JjZWBL3fn8DWBRaVCIiItJu6Sb5Ond34FzgLne/i8QevYiIiHRQ6Q6822tm1wFfAU4zsyI0dEekIB04cIBNH8a4eVWvnLS3aW+M7tXVOWlLRI6U7p78+cAB4KvBgLuBwK2hRSUiIiLtlu7o+q3AbUmLtgD1oUQkIqEqLS2loqSW7520Jyft3byqF6Xl5TlpS0SO1GKSN7NewDeAcuA/gb8AVwLfBl4GHgg7QBGRDquhnk17c3fqA3T6QzLT2p78fcBO4H+By4DrSUyG80V3fync0ERERKQ9Wkvyx7n7CQBmdi/wLjDI3T8MPTIRkY6uKMbgngdyduoDdPpDMtPawLtDjU/cvR7YogQvIiJSGFrbkx9rZo1fUQ3oFrw2wN09dyeipNNq641c2nMTF4Cqqirmzp3bpm1FRDqCFpO8u8dyFYhIKvF4nNfXvsigHpld0NHlUOJA1YFNKzNu8+19+uiLSOFLdzIckbwa1KM+5+c9RUQKXbqT4YiIiEiBUZIXERGJKCV5ERGRiFKSFxERiahQk7yZTTWzDWYWN7NrWyg3w8zczCaEGY+IiEhnElqSN7MYcBfwOWAkcIGZjWymXE9gHvBcWLGIiIh0RmHuyU8E4u7+prsfBB4Gzm2m3E3ALYBm0hMREcmiMJN8ObA56fWWYNlhZnYSUOHu/6+liszscjNbaWYrt23blv1IRUREIihvA+/MrIjEPeq/3VpZd7/H3Se4+4QBAwaEH5yIiEgEhJnkq4GKpNcDg2WNegKjgafNbCPwKWCxBt+JiIhkR5hJfgUw1MyGmFkXYBawuHGlu+929/7uXunulcByYLq7Zz7RuIiIiHxMaEne3euAK4GlwKvAH9x9rZndaGbTw2pXREREEkK9QY27LwGWNFk2P0XZKWHGIiIi0tloxjsREZGIUpIXERGJKCV5ERGRiFKSFxERiahQB96JSGrV1dXEanbTbf2S1gtnSaxmBwfMoSRnTYpIHmlPXkREJKK0Jy+SJ+Xl5Ww9UEztiGk5a7Pb+iX0aNgL1OaszVzJx5ER6ut4r0b7StJx6dMpIiISUdqTF5FIyMeRkR6r7uPosoM5a08kU9qTFxERiSgleRERkYhSkhcREYkoJXkREZGIUpIXERGJKCV5ERGRiFKSFxERiSgleRERkYhSkhcREYkoJXkREZGIUpIXERGJKCV5ERGRiFKSFxERiSgleRERkYhSkhcREYkoJXkREZGIUpIXERGJKCV5ERGRiCrOdwAdVX19PYsXL2b69OnEYrGctTtz5ky2b99O//79eeyxx3LWroi0zdv7Yty8qlfG271Xk9jHOrqsIeP2hmbcmnRWSvIpXHfddTz//PMsX76cW265JWftbt++/YifItJxeVEJ1qULpYOrMt72YDwOkPG2Q4Gqqszbk85JST6F559/HoDnnnsuz5GISEfV0LUXVccdzR133JHxtvPmzQNo07Yi6Qr1nLyZTTWzDWYWN7Nrm1n/LTNbZ2arzewpMxscZjyZKC4uPuKniIhIoQktyZtZDLgL+BwwErjAzEY2KfYiMMHdxwCPAT8JK55M1dfXH/FTRESk0IS5Jz8RiLv7m+5+EHgYODe5gLsvc/ea4OVyYGCI8WSke/fuAPTo0SPPkYiIiLRNmEm+HNic9HpLsCyVrwJ/bm6FmV1uZivNbOW2bduyGGJq+/btA2Dv3r05aU9ERCTbOsR18mb2FWACcGtz6939Hnef4O4TBgwYkNvgREREClSYo8qqgYqk1wODZUcws7OA64HT3f1AiPGISKAt13brum6RwhNmkl8BDDWzISSS+yzgwuQCZjYO+CUw1d3fDzEWEQl069aN8jZcZ63rukUKT2hJ3t3rzOxKYCkQA37j7mvN7EZgpbsvJnF4vgfwqJkBvO3u08OKSUSgvLxc13WLdBKhXgTu7kuAJU2WzU96flaY7YuIiHRmHWLgnYiIiGSfkryIiEhEac5WyZqFCxcSDwZntaTx3G6yqqoq5s6dG0ZYIiKdlpK8iERGrOYDuq1f0nrBJEUf7gESN5tpS3twdMbbieSKkrxkTUt74lOmTDn8XKOzJQxtvUwvHk/Mall1XFuS9dG6PFA6NCV5EYmEtp7u0aWBEmUaeCc5MXbsWMaOHcvTTz+d71BERDoNJXkREZGIUpIXERGJKCV5ERGRiFKSFxERiSgleRERkYjSJXQiedSWyVug7RO4aPIWkc5FSV4kT9oziUrbJ3DR5C0inYmSvEietGeufk3gIiLp0Dl5ERGRiFKSFxERiSgleRERkYhSkhcREYkoJXkREZGI0uh6ycjChQuJx+MZb9e4TeOo8ExUV1fTP+OtRERESV4yEo/HeWnNq9SX9c1ou6KDDsALb76X0Xaxmg/o0bUESjLaTEREUJKXNqgv60vtiGk5aavb+iXQsDcnbYmIRI3OyYuIiESUkryIiEhE6XC9dHgHDhxg04cxbl6V2c1Y2mPT3hjdq6tz1p6ISBi0Jy8iIhJR2pOXDq+0tJSKklq+d9KenLV586pelJaX56w9EZEwaE9eREQkopTkRUREIkpJXkREJKJCPSdvZlOBO4AYcK+7/7jJ+lJgETAe2AGc7+4bw4wpWbpTtDY3FWtVVRVz584NI6wOrbq6mljN7sQkNTkQq9nBAXPNeCci0gahJXkziwF3AWcDW4AVZrbY3dclFfsqsNPdq8xsFnALcH4242gpkVdXV1NbW9tqHc1tX11d3eIXhFRfAjKZ+11fLj7y9r7ML6F7ryZxoOrosoY2tTc0461ERDoWc/dwKjY7BfiBu382eH0dgLv/e1KZpUGZ/zWzYmArMMBbCGrChAm+cuXKtOOYOXMm27dvb+O7aJ/u3bt/bFlNTQ3t6XMzo6ysrNl1U6dODf0LwLx581LOXV/04R6s4VCb6vWiEhq6fjyJx2o+YECfXpSnGOne0he1xuXdunVrdn23bt1S1gv5/0LV0hfCxuVVVVXNrm9r7PloMxfa874a12X63lr7Qp/r/jSzF9x9QtYqlIIQ5uH6cmBz0ustwMmpyrh7nZntBvoBR2RlM7scuBxg0KBBGQVx1FFHpUwCBw4coKGh+b285OVFRR8fulBUVERpaWnKeuvr65tdF4vFUraZbrv51NI/wurqurSOjDQnkXCPbmbN0S3+s2vtSA2QMpF35KTUmlRfXKLWZi7k631FtT+lYwlzT34mMNXdLw1eXwSc7O5XJpVZE5TZErx+IyiTctc70z35QjNlypTDz59++um8xSEi0aI9+c4pzD35aqAi6fXAYFlzZbYEh+t7kxiA12kpsYuISLaEeex3BTDUzIaYWRdgFrC4SZnFwL8Gz2cCf23pfLyIiIikL7Q9+eAc+5XAUhKX0P3G3dea2Y3ASndfDPwauM/M4sAHJL4IiIiISBaEep28uy8BljRZNj/p+YfAv4QZg4iISGelGe9EREQiSkleREQkopTkRUREIkpJXkREJKJCmwwnLGa2DdiU7zjS0J8mM/dJu6g/s0d9mV2F0p+D3X1AvoOQ3Cq4JF8ozGylZpfKHvVn9qgvs0v9KR2ZDteLiIhElJK8iIhIRCnJh+eefAcQMerP7FFfZpf6UzosnZMXERGJKO3Ji4iIRJSSvIiISEQpyYuIiERUQSd5M+tmZn8zs1gz60rN7BEzi5vZc2ZWmaKOqWa2ISh3bdLyIcF28aCeLsHy08xslZnVmdnMNOMcb2avBHUtMDNrpowF6+JmttrMTkpa969m9nrw+NfW6jWzfzGztWbWYGYTkspPNrN1ZramhViP6NNUbTfZpq+ZPRmUedLM+rTjPf3IzDab2b50+jbY5rqgjQ1m9tkUZVL9Ppv9nJhZPzNbZmb7zOzOJnU1Lm/ztdE57Of/MrNdZvZEmnG19+/md2b2lpm9FDxODJafH5RNK45syXI/TzGz3UnvbX5z2zepq9nPXZMylWZWm1TvL5LWtfuzJp2cuxfsA/gGMC/Fuq8DvwiezwIeaaZMDHgDOA7oArwMjAzW/QGYFTz/BfC14HklMAZYBMxMM87ngU8BBvwZ+FwzZaYF6ywo+1ywvC/wZvCzT/C8T0v1AscDw4GngQlN2qkE1qTTpy213WSbnwDXBs+vBW5px3v6FPAJYF+afTsy+L2VAkOC32esmXKpfp/Nfk6A7sCngTnAnc3U97G+betnN6x+DtadCXwBeCLNuNr7d/M7UvxdAFPSjSNbjyz3c8bxp/rcNSlTSct/k+36rOnRuR8FvScPfBn4zxTrzgV+Hzx/DDjT7GN70BOBuLu/6e4HgYeBc4NyZwTbEdTzRQB33+juq4GGdAI0s08Avdx9ubs7iS8HX0wR7yJPWA4cFWz7WeBJd//A3XcCTwJTW6rX3V919w3pxNeM5D5ttu0UsTf29eG+yvQ9BbEvd/d3M4j3XOBhdz/g7m8BcRK/18Na+n2S4nPi7vvd/VngwwxiyUQu+hl3fwrYm0Fcbf67yaCNXMpmP2eklc+dSE4UbJIPDnsd5+4bUxQpBzYDuHsdsBvol6pMYEuwrB+wK9gueXlblAfbN20jZbxNyrW0PJ1609ZMn6Zqu6mjkxLzVuDoVrZPt950pFNXS7/PdD4nWZXDfm6L9vzdNPpRcNrgP8ystI1xtFsI/Qxwipm9bGZ/NrNRrYSQyf+RIWb2YnBqYXIr9YqkrWCTPImbQuzKdxAR0+4+DY4qaPKFlkW5n68DRgCfJHFY/Jo8xpLtfl5F4iYvY4GFwOPtqTvJu8Agdx8HfAt40Mx6Zalu6eQKOcnXAl0bX1hiwNZLZvZSsKgaqAjWFQO9gR1N6jhcJjAwWLaDxCHP4ibL26I62L5pG82Vay6WlpanU28mjujTFtpu6r3Gw8PBz/db2T7detORTl0t/T7T+ZxkW676uS3a83eDu78bnDY4APyWJqdOciyr/ezue9x9X/B8CVBiZv1baD+t/yPBqaYdwfMXSIx3GNb62xNpXcEm+eCcWszMugavr3f3E939xKDIYqBx9OxM4K/Bt/JkK4ChwQjYLiQGGi0Oyi0LtiOoJ9W5/8PMbH0zcb4L7DGzTwXn6GanqGsxMDsYKf0pYHew7VLgM2bWJxjl+xlgaQb1pq1pn6ZqO0XsjX2d3FcZvaeWYjOzfzazf0/R9ixLjAofAgwlMSAx+X219PtM53OSVTns55TM7N/N7J9baSOjv5ug3sbkaCTOP6e8kiNs2e5nMzumcXyCmU0k8f9zR/D6KTM74lB8uv9HzGyAfTT6/zgSn+E32/SmRZoKe2RfmA/g18BZKdZ1BR4lMRDreRLn5gCOBZYklZsGvEbi2/P1ScuPC7aLB/WUBss/SeLc2n4Sf+Brg+X9gQ0pYplA4p/dG8CdfDSd8BxgTvDcgLuCMq+QNJoW+D9BHHHgkjTq/ecgxgPAeyS+FDRuU0nLI3mP6NMW2r63MUYS5x6fAl4H/gL0bcd7+kkQe0Pw8wfB8quB61LEfH3QxgaSrlwAlgDHtvL7bPZzEqzbCHwA7AtiGZm07mnaN7o+V/38DLCNxF7tFuCzwfIngFNC+Lv5axDDGuB+oEfSuinkfnR9Nvv5SmAtiasJlgOTguVFwCagWzPtp/rcTQduDJ7PCOp9icQpgS80qaNdnzU9OvejoOeut8T1wFe5+0UdIJZzSPxDXJDvWFpiieuen3D30SnWd5g+TWZm95OIa1u+YwEws6eBq919ZRu3z2s/m9lSd292ToEQ25xCos/OyWGbofezmY0G/o+7fyuk+p+mHZ816dyKWy/Scbn7KktMFhFz9/o8x5LTST7aIhi1+3Nge6oyHalPk7n7V/IdQyMzW0ZiD+1QW+vIdz/nIcGfD9wAvJDLdnPRz+6+hsSAuazLxmdNOreC3pMXERGR1Ap24J2IiIi0TEleREQkopTkRUREIkpJXgqOmXkw2r7xdbGZbbMmdzgzs8fNbHmTZT8ws2r76I5fP85STJVmdmE26kqjrRvN7KxctCUiha2gR9dLp7UfGG1m3dy9FjibJjOJmdlRwHhgn5kd5+7Jk4v8h7v/NMsxVQIXAg9mud4jBKPEW73FqYgIaE9eCtcS4PPB8wuAh5qs/xLwJxJ3SJvVlgbM7JNm9ndL3JDkeTPrGeyxP2Nmq4LHpKD4j4HJwdGBq8wsZma3mtkKS9ys5YqgziIz+7mZrbfEvcqXmNnMYN2ZlrhJyStm9hsLbu5iZhvN7BYzWwX8iyXu2d64zXhL3NTkBTNbmjTj3DfNbF3Q9sNtef8iUviU5KVQPUxiOtuuwBjguSbrGxP/Q8HzZFclHa5v9nrxYLrWR0jci3wscBaJWePeB85295OA84HGyY+uBZ7xxNTK/wF8lcT0sp8kMUviZcG0u18isdc/ErgIOCVoryuJe7Gf7+4nkDjK9rWkkHa4+0nufjhhm1kJiRulzHT38cBvgB8lxTPO3ceQmFlRRDohHa6XguTuq4PZ+y4gsVd/mJkdTWL+72fd3c3skJmNDiYtgfQO1w8H3nX3FUF7e4K6uwN3mtmJQD2pbyTyGWBM4x43iRu9DAU+DTzq7g3A1mCyk8b23nL314LXvwe+AdwevH4kRYyjgSeDKdVjJO5oBrAaeMDMHid7d0sTkQKjJC+FbDHwUxJzoiff8/w8oA/wVpD8epH4MnB9Ftq8isT9AMaSOBL2YYpyBsx19yNugGJm09rY7v4Ubax191OaWfd54DTgC8D1ZnaCf3RfcxHpJHS4XgrZb4AfuvsrTZZfAEx190p3ryQxAC/T8/IbgE+Y2ScBgvPxjbdefTfYE7+IxN4zwF6gZ9L2S4GvBYfUMbNhwVGA/wFmBOfmjybxBaWxvUozqwpeXwT8LY0YB5hZ4yH/EjMbZWZFQIW7LyNxP/feQI8M37+IRID25KVgufsWPjonDhy+Ac9gEncJayz3lpntNrOTM6j7YDDf+kIz60bifPxZJOb+/6OZzQb+i4/2sFcD9Wb2Molz63eQOPe+yhKHE7aRuPXqH4EzgXXAZhJ3Hdvt7h+a2SXAo8GXiRXAL9KIcSawwMx6k/h7vp3E3eHuD5YZsMDdd6X73kUkOjR3vUiOmVkPd99nZv1I3Ib0VHffmu+4RCR6tCcvkntPBNfxdwFuUoIXkbBoT146PTP7v8CQJouvaTpoTkSk0CjJi4iIRJRG14uIiESUkryIiEhEKcmLiIhElJK8iIhIRP1/ZsmAB6ZGCgkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.boxplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    width=0.7,\n",
    "    fliersize=2\n",
    "    )\n",
    "\n",
    "# Put the legend out of the figure\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.stripplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    size = 1,\n",
    "    dodge=True\n",
    "    )\n",
    "\n",
    "# Put the legend out of the figure\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ax = sns.boxplot(x=\"tip\", y=\"day\", data=tips, whis=np.inf)\n",
    "\n",
    "ax = sns.boxplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    size = 1,\n",
    "    dodge=True\n",
    "    )\n",
    "\n",
    "ax = sns.stripplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    size = 1,\n",
    "    dodge=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.relplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info_subset,\n",
    "    kind='line',\n",
    "    hue=\"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed']\n",
    "    )\n",
    "\n",
    "# Put the legend out of the figure\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_subset = imputation_info.sample(100000)\n",
    "imputation_info_subset[\"MAF_categories\"] = imputation_info_subset[\"MAF_categories\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_subset.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_info_subset[\"MAF_categories\"].head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No handles with labels found to put in legend.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x2b02e6857c88>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZ0AAAEGCAYAAAC+fkgiAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAARVklEQVR4nO3da5CkV13H8e9vswawxEXNcHGzmwmVSJEKirAGDOINtQLUkvICJIBAKhBUgqVSlFi+AOWV5aW8kJIsMQSoIhcptDYmmvISCGUSyCIIGzDUGgg7WdYsAiOICsv+fdEdHYbsTs/s9Hnm9Hw/VV3pPv08Pf+T6e3fnPOc5+lUFZIktbBl6AIkSZuHoSNJasbQkSQ1Y+hIkpoxdCRJzWwduoCTcdppp9X8/PzQZUhSVz70oQ99rqrmTvI1Hr1169argHP5xgHMMWD/0aNHX/HUpz71geX7dR068/Pz7Nu3b+gyJKkrSe472dfYunXrVY997GOfODc394UtW7b837k3x44dy5EjR845fPjwVcDzlu/n9JokaS3OnZub+4+lgQOwZcuWmpubW2Q0Avomho4kaS22LA+cJU8Ux8kXQ0eS1EyXoZNkd5I9i4uLQ5ciSVqFLkOnqm6sqsu2bds2dCmStFkdO3bsWI7zRBitYvsmXYaOJGlw+48cObJtefCMV69tA/Y/1E5dL5mWJA3j6NGjrzh8+PBVhw8fPu55Og+1n6EjSVq18Ymf33QezkqcXluF+R3bSTLxbX7H9qFLlqQNxZHOKty3cIh623Mn3j6X3DTFaiSpP450JEnNdBk6nqcjSX3qMnQ8T0eS+tRl6EiS+mToSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNdNl6HgZHEnqU5eh42VwJKlPXYaOJKlPho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzho4kqRlDR5LUjKEjSWrG0JEkNWPoSJKa2Tp0AQ9K8kzgxYxqOqeqzh+4JEnSOpvqSCfJ1UkeSLJ/WfsFSe5JciDJ6wGq6v1V9QvAXwFvn2ZdkqRhTHt67RrggqUNSU4BrgCeDZwDXJzknCWbvAh415TrkiQNYKqhU1W3AZ9f1nwecKCq7q2qrwLXARcCJNkJLFbVl6ZZlyRpGEMsJNgOHFzyeGHcBnAp8LYT7ZzksiT7kuw7cuTIlEqUJE3Dhlq9VlVvqKrbV9hmT1Xtqqpdc3NzrUqTJK2DIULnfmDHksenj9skSTNuiNC5Czg7yZlJTgUuAvau5gWS7E6yZ3FxcSoFSpKmY9pLpq8F7gCekGQhyaVVdRS4HLgF+ARwQ1XdvZrXraobq+qybdu2rX/RkqSpmerJoVV18XHabwZunubPliRtPBtqIcGknF6TpD51GTpOr0lSn7oMHUlSnwwdSVIzXYaOx3QkqU9dho7HdCSpT12GjiSpT4aOJKmZLkPHYzqS1KcuQ8djOpLUpy5DR5LUJ0NHktSMoSNJaqbL0HEhgST1qcvQcSGBJPWpy9CRJPXJ0JEkNWPoSJKaMXQkSc10GTquXpOkPnUZOq5ek6Q+dRk6kqQ+GTqSpGYMHUlSM4aOJKkZQ0eS1IyhI0lqpsvQ8TwdSepTl6HjeTqS1KcuQ0eS1CdDR5LUjKEjSWrG0JEkNWPoSJKaMXQkSc1s2tDZvmMnSVZ1Y8vWocuWpK5t2k/RQwsHeeGVt69qn+tfdf6UqpGkzWHTjnQkSe11GTpeBkeS+tRl6HgZHEnqU5ehI0nqk6EjSWrG0JEkNWPoSJKaMXQkSc0YOpKkZgwdSVIzE4VOkmdM0iZJ0olMOtL5kwnbJEk6rhNe8DPJDwLnA3NJfm3JU98OnDLNwiRJs2elq0yfCnzbeLtHLmn/D+DnplWUJGk2nTB0qup9wPuSXFNV9zWqSZI0oyb9Pp2HJdkDzC/dp6p+fBpFSZJm06Sh8+fAW4CrgK9PrxxJ0iybNHSOVtWfTrOQJFuANzFapLCvqt4+zZ8nSWpv0iXTNyb5pSSPS/KdD95W2inJ1UkeSLJ/WfsFSe5JciDJ68fNFwKnA18DFlbVC0lSFyYd6bxs/N/XLWkr4PEr7HcN8GbgHQ82JDkFuAL4SUbhcleSvcATgNur6sok7wb+fsLaJEmdmCh0qurMtbx4Vd2WZH5Z83nAgaq6FyDJdYxGOQeBr4638biRJM2giUInyUsfqr2q3vFQ7SvYzihgHrQAPA34I+BPkjwTuO0EtVwGXAawc+fONfx4SdJQJp1e+4El9x8OPAv4J5ZMm52sqvoKcOkE2+0B9gDs2rWr1uvnS5Kmb9LptdcsfZzkUcB1a/yZ9wM7ljw+fdwmSZpxa/1qg/8E1nScB7gLODvJmUlOBS4C9q7mBZLsTrJncXFxjSVIkoYw6Vcb3Jhk7/h2E3AP8BcT7HctcAfwhCQLSS6tqqPA5cAtwCeAG6rq7tUUXVU3VtVl27ZtW81ukqSBTXpM5/eW3D8K3FdVK55LU1UXH6f9ZuDmCX+2JGlGTDTSGV/4818YXWn6O/j/pc2DcHpNkvo06fTaC4APAs8HXgB8IMlgX23g9Jok9WnS6bXfBH6gqh4ASDIH/B3w7mkVJkmaPZOuXtvyYOCM/fsq9pUkCZh8pPM3SW4Brh0/fiEDLgRIshvYfdZZZw1VgiRpDU44WklyVpJnVNXrgCuB7x3f7mB8VYAheExHkvq00kjnD4HfAKiq9wDvAUjypPFzu6dYmyRpxqx0XOYxVfWx5Y3jtvmpVCRJmlkrhc6jTvDcI9axjlXxPB1J6tNKobMvySuXNyZ5BfCh6ZS0Mo/pSFKfVjqm8yvAXyR5Mf8fMruAU4GfnmJdkqQZdMLQqap/A85P8mPAuePmm6rqH6Zemb7B/I7t3LdwaOLtzzj9u/n0Qb8xQtLGMun36dwK3DrlWnQC9y0cot723Im3zyU3TbEaSVqbLq8q4EICSepTl6HjQgJJ6lOXoSNJ6pOhI0lqxtCR9A3md2wnycS3+R3bZ+Jnq41JrzItqaEhl8gPuVLSVZqzr8vQ8asNNOv88NWs6nJ6zdVrbTnloc1gte9z3+tr0+VIR21t1r+6vQrE5rLa9zms33t9te816Pf9ZuhoQ9usxza0uQwZeK0ZOtrQ/OCXZkuXx3R6t33HzlXPHbPFvw8k9c9PsgEcWjjIC6+8fVX7XP+q86dUjSS140hHm4KjS2lj6PJflefpaLUcXUobQ5cjHc/T6ZOjDUn+i95ktu/YyaGFg6vbaZ0++B1tSDJ0Nhk/+CUNqcvpNUlSnwwdqYHVHs/yWJZmle9sqYHVTms6palZ5UhHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzXQZOkl2J9mzuLg4dCmSpFXoMnS89pok9anL0JEk9cnQkSQ1Y+hIM85L8Ggj8d0lzTgvwaONxJGOJKkZQ0eS1IyhI0lqxtCRJDVj6EiSmjF0JEnNGDqSpGYMHUlSM4aOJKkZQ0fS1HgJHi23YX7DSX4UeBNwN3BdVb13yHoknTwvwaPlpjrSSXJ1kgeS7F/WfkGSe5IcSPL6cXMBXwYeDixMsy5J0jCmPb12DXDB0oYkpwBXAM8GzgEuTnIO8P6qejbw68BvTbkuSdIApho6VXUb8PllzecBB6rq3qr6KnAdcGFVHRs//wXgYcd7zSSXJdmXZN+RI0emUrckaTqGWEiwHTi45PECsD3JzyS5Engn8Obj7VxVe6pqV1Xtmpubm3KpkqT1tGEWElTVe4D3DF2HJGl6hhjp3A/sWPL49HHbxJLsTrJncXFxXQuTJE3XEKFzF3B2kjOTnApcBOxdzQtU1Y1Vddm2bdumUqAkaTqmvWT6WuAO4AlJFpJcWlVHgcuBW4BPADdU1d3TrEOStDFM9ZhOVV18nPabgZvX+rpJdgO7zzrrrLW+hCRpAF1eBsfpNUnqU5ehI0nqk6EjaSZ5sdGNqcv/yx7TkbQSLza6MXU50vGYjiT1qcvQkST1ydCRJDXTZeh4GRxJ6lOXoeMxHUnqU5ehI0nqk6EjSWrG0JEkNdNl6LiQQJL61GXouJBAkvrUZehIkvpk6EiSmjF0JEnNGDqSpGa6DB1Xr0lSn7oMHVevSVKfugwdSVKfDB1JUjOGjiSpGUNHktSMoSNJasbQkSQ102XoeJ6OJPWpy9DxPB1J6lOXoSNJ6pOhI0nrbPuOnSSZ+MaWrUOX3Mzm6akkNXJo4SAvvPL2ibe//lXnT7GajcWRjiSpGUNHktSMoSNJasbQkSQ1Y+hIkpoxdCRJzRg6kqRmugwdr70mSX1KVQ1dw5olOQLct4pdTgM+N6VyNjL7vbls1n7D5u37avt9RlXNTauYE+k6dFYryb6q2jV0Ha3Z781ls/YbNm/fe+p3l9NrkqQ+GTqSpGY2W+jsGbqAgdjvzWWz9hs2b9+76femOqYjSRrWZhvpSJIGZOhIkpqZydBJckGSe5IcSPL6h3j+YUmuHz//gSTzA5S57ibo968l+XiSjyb5+yRnDFHnelup30u2+9kklaSLpaUrmaTfSV4w/p3fneRdrWuchgne5zuT3Jrkw+P3+nOGqHO9Jbk6yQNJ9h/n+ST54/H/l48meUrrGidSVTN1A04B/hV4PHAq8M/AOcu2+SXgLeP7FwHXD113o37/GPCt4/u/uFn6Pd7ukcBtwJ3ArqHrbvT7Phv4MPAd48ePHrruRv3eA/zi+P45wKeHrnud+v7DwFOA/cd5/jnAXwMBng58YOiaH+o2iyOd84ADVXVvVX0VuA64cNk2FwJvH99/N/CsJGlY4zSs2O+qurWqvjJ+eCdweuMap2GS3zfAm4DfAf67ZXFTNEm/XwlcUVVfAKiqBxrXOA2T9LuAbx/f3wYcaljf1FTVbcDnT7DJhcA7auRO4FFJHtemusnNYuhsBw4uebwwbnvIbarqKLAIfFeT6qZnkn4vdSmjv4p6t2K/x9MMO6rqppaFTdkkv+/vAb4nyT8muTPJBc2qm55J+v1G4CVJFoCbgde0KW1wq/0MGMTWoQtQe0leAuwCfmToWqYtyRbgD4CXD1zKELYymmL7UUaj2tuSPKmqvjhkUQ1cDFxTVb+f5AeBdyY5t6qODV2YZnOkcz+wY8nj08dtD7lNkq2MhuD/3qS66Zmk3yT5CeA3gedV1f80qm2aVur3I4Fzgfcm+TSjue69M7CYYJLf9wKwt6q+VlWfAj7JKIR6Nkm/LwVuAKiqO4CHM7og5qyb6DNgaLMYOncBZyc5M8mpjBYK7F22zV7gZeP7Pwf8Q42PxHVsxX4n+X7gSkaBMwvz+7BCv6tqsapOq6r5qppndCzreVW1b5hy180k7/O/ZDTKIclpjKbb7m1Y4zRM0u/PAM8CSPJERqFzpGmVw9gLvHS8iu3pwGJVfXboopabuem1qjqa5HLgFkYrXa6uqruT/Dawr6r2An/GaMh9gNGBuYuGq3h9TNjv3wW+Dfjz8bqJz1TV8wYreh1M2O+ZM2G/bwF+KsnHga8Dr6uqrkf0E/b7tcBbk/wqo0UFL5+BPypJci2jPyJOGx+vegPwLQBV9RZGx6+eAxwAvgJcMkylJ+ZlcCRJzczi9JokaYMydCRJzRg6kqRmDB1JUjOGjiSpmZlbMi21kOTrwMcY/Rv6FPDzm+BMf+mkOdKR1ua/qurJVXUuo3O9Xj10QVIPDB3p5N3B+MKKSR6X5LYkH0myP8kzx+2XJPlkkg8meWuSNw9asTQQQ0c6CUlOYXTJlQevfPAi4JaqejLwfcBHxpeX/y3gGcAPMfqOF2lTMnSktXlEko8Ah4HHAH87br8LuCTJG4EnVdWXgKcB762qI+PvgLl+gHqlDcHQkdbmv8ajmTMYfVPjq+H/vmjrhxld3feaJC8drEJpAzJ0pJMw/ibWXwZem2RrkjOAf6uqtwJXMfp64Q8AP5Lku5J8C/D84SqWhuWSaekkVdWHk3yU0ZeHbQFel+RrwJeBl1bVZ8fTbXcAXwQ+MlCp0uC8yrTUWJKXA7uq6vKha5Fac3pNktSMIx1JUjOOdCRJzRg6kqRmDB1JUjOGjiSpGUNHktTM/wL/J6XYo2Ym4wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "g = sns.histplot(\n",
    "    x = \"Rsq\",\n",
    "    binwidth=0.1,\n",
    "    binrange=(0,1),\n",
    "    data = imputation_info,\n",
    "#     data = imputation_info.sample(1000),\n",
    "    hue = \"Reference_Panel\",\n",
    "    hue_order=['michigan_hrc','topmed'],\n",
    "    multiple='dodge',\n",
    "    shrink=0.8\n",
    "    )\n",
    "\n",
    "g.set_yscale(\"log\")\n",
    "\n",
    "# # Put the legend out of the figure\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "# Define which attribute to use on y axis: Rsq\n",
    "y_atribute = \"Rsq\"\n",
    "\n",
    "g = sns.FacetGrid(\n",
    "    imputation_info,\n",
    "    col=\"chr\",\n",
    "    col_wrap=4,\n",
    "    col_order=natural_sort( imputation_info.chr.unique() ),\n",
    "    hue=\"Genotyped\",\n",
    "    sharex=False\n",
    "    )\n",
    "g.map_dataframe(\n",
    "    sns.scatterplot,\n",
    "    x=\"pos\",\n",
    "    y=y_atribute,\n",
    "    alpha=0.4,\n",
    "    s=6\n",
    "    )\n",
    "\n",
    "g.map_dataframe(\n",
    "    add_median_horizontal_line, \n",
    "    var=y_atribute\n",
    "    )\n",
    "g.set_axis_labels(\n",
    "    \"chromosome position\",\n",
    "    y_atribute\n",
    "    )\n",
    "\n",
    "# Add a general title to the figure\n",
    "g.fig.subplots_adjust(top=0.90)\n",
    "g.fig.suptitle( \"Genotyped and imputed SNPs from Michigan HRC hg19 (N={0:,})\".format( len(imputation_info) ) )\n",
    "\n",
    "# Edit title for the subplots\n",
    "imputation_info_counts_by_chr = imputation_info[\"chr\"].value_counts()\n",
    "g.set_titles( col_template=\"{col_name}\" )\n",
    "for subfig in g.axes:\n",
    "    chr_id = str( subfig.get_title() )\n",
    "    number_of_snps = imputation_info_counts_by_chr[ chr_id ]\n",
    "    subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "\n",
    "# format the labels with f-strings\n",
    "for ax in g.axes.flat:\n",
    "    ax.xaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000000)}M'))\n",
    "    \n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.set_style('white')\n",
    "\n",
    "# # Define which attribute to use on y axis: Rsq, LooRsq, EmpR, EmpRsq, Dose0, Dose1\n",
    "# y_atribute = \"LooRsq\"\n",
    "\n",
    "# g = sns.FacetGrid(\n",
    "#     imputation_info_genotyped,\n",
    "# #   row=\"Reference_Panel\",\n",
    "#     col=\"chr\",\n",
    "#     col_wrap=4,\n",
    "#     col_order=natural_sort( imputation_info_genotyped.chr.unique() ),\n",
    "#     sharex=False\n",
    "#     )\n",
    "# g.map_dataframe(\n",
    "#     sns.scatterplot,\n",
    "#     x=\"pos\",\n",
    "#     y=y_atribute,\n",
    "#     alpha=0.5,\n",
    "#     s=10\n",
    "#     )\n",
    "\n",
    "# g.map_dataframe(\n",
    "#     add_median_horizontal_line, \n",
    "#     var=y_atribute\n",
    "#     )\n",
    "# g.set_axis_labels(\n",
    "#     \"chromosome position\",\n",
    "#     y_atribute\n",
    "#     )\n",
    "\n",
    "# # Add a general title to the figure\n",
    "# g.fig.subplots_adjust(top=0.90)\n",
    "# g.fig.suptitle( \"Genotyped SNPs (N={0:,})\".format( len(imputation_info_genotyped) ) )\n",
    "\n",
    "# # Edit title for the subplots\n",
    "# imputation_info_genotyped_counts_by_chr = imputation_info_genotyped[\"chr\"].value_counts()\n",
    "# g.set_titles( col_template=\"{col_name}\" )\n",
    "# for subfig in g.axes:\n",
    "#     chr_id = str( subfig.get_title() )\n",
    "#     number_of_snps = imputation_info_genotyped_counts_by_chr[ chr_id ]\n",
    "#     subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "\n",
    "# # format the labels with f-strings\n",
    "# for ax in g.axes.flat:\n",
    "#     ax.xaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000000)}M'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "# Define which attribute to use on y axis: Rsq, LooRsq, EmpR, EmpRsq, Dose0, Dose1\n",
    "y_atribute = \"LooRsq\"\n",
    "\n",
    "g = sns.FacetGrid(\n",
    "    imputation_info_genotyped,\n",
    "    row=\"chr\",\n",
    "    col=\"Reference_Panel\",\n",
    "#     col_wrap=4,\n",
    "#     col_order=natural_sort( imputation_info_genotyped.chr.unique() ),\n",
    "    sharex=False\n",
    "    )\n",
    "g.map_dataframe(\n",
    "    sns.scatterplot,\n",
    "    x=\"pos\",\n",
    "    y=y_atribute,\n",
    "    alpha=0.5,\n",
    "    s=10\n",
    "    )\n",
    "\n",
    "g.map_dataframe(\n",
    "    add_median_horizontal_line, \n",
    "    var=y_atribute\n",
    "    )\n",
    "g.set_axis_labels(\n",
    "    \"chromosome position\",\n",
    "    y_atribute\n",
    "    )\n",
    "\n",
    "# Add a general title to the figure\n",
    "g.fig.subplots_adjust(top=0.90)\n",
    "g.fig.suptitle( \"Genotyped SNPs (N={0:,})\".format( len(imputation_info_genotyped) ) )\n",
    "\n",
    "# Edit title for the subplots\n",
    "imputation_info_genotyped_counts_by_chr = imputation_info_genotyped[\"chr\"].value_counts()\n",
    "g.set_titles( col_template=\"{col_name}\" )\n",
    "for subfig in g.axes:\n",
    "    chr_id = str( subfig.get_title() )\n",
    "    number_of_snps = imputation_info_genotyped_counts_by_chr[ chr_id ]\n",
    "    subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "\n",
    "# format the labels with f-strings\n",
    "for ax in g.axes.flat:\n",
    "    ax.xaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000000)}M'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_categories = [0, 0.0001, 0.01, 0.05, 0.5]\n",
    "\n",
    "imputation_info[\"MAF_categories\"] = pd.cut(\n",
    "    imputation_info.MAF,\n",
    "    maf_categories\n",
    "    ).astype(str)\n",
    "\n",
    "# median_rsq_by_maf_tmp = imputation_info.groupby(\"MAF_categories\")[\"Rsq\"].median()\n",
    "\n",
    "# median_rsq_by_maf = pd.DataFrame(\n",
    "#     {\n",
    "#     \"MAF_categories\": median_rsq_by_maf_tmp.index.astype(str),\n",
    "#     \"Median_observed_imput_rsq\": median_rsq_by_maf_tmp.values\n",
    "#     }\n",
    "# )\n",
    "\n",
    "# median_rsq_by_maf.head()\n",
    "\n",
    "sns.lineplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Rsq\",\n",
    "    data = imputation_info,\n",
    "    color = \"maroon\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "g = sns.FacetGrid(\n",
    "    imputation_info,\n",
    "    col=\"chr\",\n",
    "    col_wrap=4,\n",
    "    col_order=natural_sort( imputation_info.chr.unique() ),\n",
    "    hue=\"Genotyped\",\n",
    "    sharex=False\n",
    "    )\n",
    "g.map_dataframe(\n",
    "    sns.histplot,\n",
    "    x=\"Rsq\",\n",
    "    binwidth=0.1,\n",
    "    binrange=(0,1)\n",
    "    )\n",
    "\n",
    "g.map_dataframe(\n",
    "    add_median_vertical_line, \n",
    "    var=\"Rsq\"\n",
    "    )\n",
    "\n",
    "# Add a general title to the figure\n",
    "g.fig.subplots_adjust(top=0.90)\n",
    "g.fig.suptitle( \"Genotyped and imputed SNPs from Michigan HRC hg19 (N={0:,})\".format( len(imputation_info) ) )\n",
    "\n",
    "# Edit title for the subplots\n",
    "imputation_info_counts_by_chr = imputation_info[\"chr\"].value_counts()\n",
    "g.set_titles( col_template=\"{col_name}\" )\n",
    "for subfig in g.axes:\n",
    "    chr_id = str( subfig.get_title() )\n",
    "    number_of_snps = imputation_info_counts_by_chr[ chr_id ]\n",
    "    subfig.set_title(\"chr{0} (N={1:,})\".format(chr_id, number_of_snps) )\n",
    "    \n",
    "# format the labels with f-strings\n",
    "for ax in g.axes.flat:\n",
    "    ax.yaxis.set_major_formatter(tkr.FuncFormatter(lambda x, p: f'{round(x/1000)}k'))\n",
    "    # ax.set_yscale(\"log\")\n",
    "    \n",
    "# g.set_yscale(\"log\")\n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_categories = [0, 0.0001, 0.01, 0.05, 0.5]\n",
    "\n",
    "imputation_info[\"MAF_categories\"] = pd.cut(\n",
    "    imputation_info.MAF,\n",
    "    maf_categories\n",
    "    )\n",
    "\n",
    "median_rsq_by_maf_tmp = imputation_info.groupby(\"MAF_categories\")[\"Rsq\"].median()\n",
    "\n",
    "median_rsq_by_maf = pd.DataFrame(\n",
    "    {\n",
    "    \"MAF_categories\": median_rsq_by_maf_tmp.index.astype(str),\n",
    "    \"Median_observed_imput_rsq\": median_rsq_by_maf_tmp.values\n",
    "    }\n",
    ")\n",
    "\n",
    "g = sns.lineplot(\n",
    "    x = \"MAF_categories\",\n",
    "    y = \"Median_observed_imput_rsq\",\n",
    "    data = median_rsq_by_maf,\n",
    "    color = \"maroon\"\n",
    "    )\n",
    "\n",
    "g.set_title( \"Genotyped and imputed SNPs from Michigan HRC hg19\\n(N={0:,})\".format( len(imputation_info) ) )\n",
    "g.set(\n",
    "    xlabel = \"Reference panel Minor Allele Frequency\",\n",
    "    ylabel = \"Median Observed Imput Rsq\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    # sns.histplot,\n",
    "    # x=\"Rsq\",\n",
    "    # bins=10\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.histplot(\n",
    "    x = \"MAF\",\n",
    "    y = \"Rsq\",\n",
    "    # hue=\"maf_categories\",\n",
    "    data = imputation_info,\n",
    "    bins = [0,0.0001,0.01,0.05,0.5]\n",
    "    # stat='probability',\n",
    "    # cumulative=True,\n",
    "    # fill=False,\n",
    "    # element='step'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imputation_info[\"maf_categories\"] = pd.cut(\n",
    "#     imputation_info,\n",
    "#     [-np.inf, 0.001, 0.1, 0.2, 0.3, np.inf]\n",
    "#     )\n",
    "\n",
    "print( imputation_info.groupby(\"maf_categories\")[\"Rsq\"].mean() )\n",
    "\n",
    "# sns.histplot(\n",
    "#     x = \"Rsq\",\n",
    "#     # y = \"Rsq\",\n",
    "#     hue=\"maf_categories\",\n",
    "#     data = imputation_info\n",
    "#     # cumulative=True,\n",
    "#     # kde=True\n",
    "# )\n",
    "#     x = \"Rsq\",\n",
    "#     data = imputation_info,\n",
    "#     bins = 10,\n",
    "#     element='step',\n",
    "#     col=\n",
    "#     )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a plot of **Rsq** along with the chromosome genomic positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create scatter plot of genomic positions by estimated r2 (from minimac4 leave one out column)\n",
    "plt.scatter( \n",
    "    x = chr21_genotyped.pos.astype(int), \n",
    "    y = chr21_genotyped.LooRsq.astype(float), \n",
    "    marker = 'o',\n",
    "    color='gray',\n",
    "    s=3,\n",
    "    alpha=0.75\n",
    "    )\n",
    "\n",
    "# Add horizontal line with mean LooRsq\n",
    "mean_rsq = statistics.mean( chr21_genotyped.LooRsq.astype(float) )\n",
    "plt.axhline(\n",
    "    y=mean_rsq, \n",
    "    color='red',\n",
    "    linestyle='--'\n",
    "    )\n",
    "plt.text( \n",
    "    x = max(chr21_genotyped.pos.astype(int))*0.825,\n",
    "    y = mean_rsq+0.025,\n",
    "    s = \"Mean R\\u00b2 = {:.2f}\".format(mean_rsq),\n",
    "    color='red'\n",
    "    )\n",
    "\n",
    "# Adjust labels and axis\n",
    "plt.title( \"chr21 (N = \" + str(len(chr21_genotyped)) + \")\" )\n",
    "plt.xlabel(\"chr21 genomic position\")\n",
    "plt.ylabel(\"Empirical (Leave-One-Out) R\\u00b2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a histogram of LooRsq\n",
    "plt.hist(\n",
    "  x = chr21_genotyped.LooRsq.astype(float),\n",
    "  bins = 10,\n",
    "  color='black',\n",
    "  alpha = 0.6\n",
    "  )\n",
    "\n",
    "# Adjust labels and axis\n",
    "plt.title( \"chr21 (N = \" + str(len(chr21_genotyped)) + \")\" )\n",
    "plt.xlabel(\"Empirical (Leave-One-Out) R\\u00b2\")\n",
    "plt.ylabel(\"Frequency\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get number of passing variants at different RSq levels \n",
    "rsq_filter = [0, 0.001, 0.1, 0.2, 0.3]\n",
    "\n",
    "total = len(chr21.Rsq.astype(float))\n",
    "\n",
    "for c_rsq in rsq_filter:\n",
    "    passing = sum ( chr21.Rsq.astype(float) >= c_rsq)\n",
    "    print (\"Passing = {0:,}, filtered= {1:,} \\t (RSq threshold = {2:,})\".format(passing, total-passing, c_rsq) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a plot of **MAF** by **Mean imputation Rsq** "
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "kernelspec": {
   "display_name": "Python 3.6.8 64-bit",
   "language": "python",
   "name": "python368jvsc74a57bd0916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
